// @mcpfusion/core/dist/domain/Group.js
import { BaseModel } from './BaseModel.js';
import {} from './GroupItem.js';
import {} from './Tool.js';
import {} from './Prompt.js';
import {} from './Resource.js';
import { removeFromArray } from './utils.js';
/**
 * Represents a hierarchical group (tree node) in the MCP domain model.
 *
 * Groups can contain child groups, tools, prompts, and resources.
 * Supports fully qualified names via recursive parent traversal.
 *
 * @example
 * ```typescript
 * // Build a hierarchy
 * const api = new Group('api');
 * const v2 = new Group('v2');
 * api.addChildGroup(v2);
 *
 * const readFile = new Tool('read_file');
 * v2.addChildTool(readFile);
 *
 * v2.getFullyQualifiedName();  // "api.v2"
 * api.isRoot();                // true
 * v2.getRoot();                // api
 * ```
 *
 * @see {@link GroupItem} for leaf nodes (Tool, Prompt, Resource)
 * @see {@link BaseModel} for shared properties
 */
export class Group extends BaseModel {
    /** Parent group (null if this is a root node) */
    parent = null;
    /** Direct child groups in this container */
    childGroups;
    /** Tools contained in this group */
    childTools;
    /** Prompts contained in this group */
    childPrompts;
    /** Resources contained in this group */
    childResources;
    constructor(name, nameSeparator) {
        super(name, nameSeparator !== undefined ? nameSeparator : BaseModel.DEFAULT_SEPARATOR);
        this.childGroups = [];
        this.childTools = [];
        this.childPrompts = [];
        this.childResources = [];
    }
    // ── Private helpers to eliminate add/remove repetition ──
    addChild(list, child) {
        if (list.includes(child))
            return false;
        list.push(child);
        child.addParentGroup(this);
        return true;
    }
    removeChild(list, child) {
        if (!removeFromArray(list, child))
            return false;
        child.removeParentGroup(this);
        return true;
    }
    // ── Tree navigation ──
    /**
     * Traverse to the root of the tree.
     *
     * @returns The topmost ancestor group
     *
     * @example
     * ```typescript
     * const root = new Group('api');
     * const child = new Group('users');
     * root.addChildGroup(child);
     * child.getRoot(); // root
     * ```
     */
    getRoot() {
        return this.parent === null ? this : this.parent.getRoot();
    }
    /** Returns `true` if this group has no parent (is a root node). */
    isRoot() {
        return this.parent === null;
    }
    // ── Child groups (special: sets parent, not parentGroup) ──
    /**
     * Add a child group to this container.
     *
     * @param childGroup - The group to nest under this one
     * @returns `false` if already a child, `true` if added
     */
    addChildGroup(childGroup) {
        if (this.childGroups.includes(childGroup))
            return false;
        // Cycle detection — walk the parent chain to prevent infinite recursion
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let ancestor = this;
        while (ancestor !== null) {
            if (ancestor === childGroup) {
                throw new Error(`[mcpfusion] Cycle detected: adding '${childGroup.name}' as a child of '${this.name}' would create a circular reference.`);
            }
            ancestor = ancestor.parent;
        }
        // Remove from previous parent to avoid ghost references ( fix)
        if (childGroup.parent !== null) {
            childGroup.parent.removeChildGroup(childGroup);
        }
        this.childGroups.push(childGroup);
        childGroup.parent = this;
        return true;
    }
    /**
     * Remove a child group from this container.
     *
     * @param childGroup - The group to remove
     * @returns `false` if not found, `true` if removed
     */
    removeChildGroup(childGroup) {
        if (!removeFromArray(this.childGroups, childGroup))
            return false;
        childGroup.parent = null;
        return true;
    }
    // ── Child items (delegated to helpers) ──
    /** Add a tool to this group. Returns `false` if already present. */
    addChildTool(childTool) {
        return this.addChild(this.childTools, childTool);
    }
    /** Remove a tool from this group. Returns `false` if not found. */
    removeChildTool(childTool) {
        return this.removeChild(this.childTools, childTool);
    }
    /** Add a prompt to this group. Returns `false` if already present. */
    addChildPrompt(childPrompt) {
        return this.addChild(this.childPrompts, childPrompt);
    }
    /** Remove a prompt from this group. Returns `false` if not found. */
    removeChildPrompt(childPrompt) {
        return this.removeChild(this.childPrompts, childPrompt);
    }
    /** Add a resource to this group. Returns `false` if already present. */
    addChildResource(childResource) {
        return this.addChild(this.childResources, childResource);
    }
    /** Remove a resource from this group. Returns `false` if not found. */
    removeChildResource(childResource) {
        return this.removeChild(this.childResources, childResource);
    }
    // ── FQN ──
    getFullyQualifiedNameRecursive(tg) {
        const parent = tg.parent;
        if (parent !== null) {
            const parentName = this.getFullyQualifiedNameRecursive(parent);
            return parentName + this.nameSeparator + tg.name;
        }
        return tg.name;
    }
    /**
     * Returns the dot-separated fully qualified name.
     *
     * @example
     * ```typescript
     * const root = new Group('api');
     * const v2 = new Group('v2');
     * root.addChildGroup(v2);
     * v2.getFullyQualifiedName(); // "api.v2"
     * ```
     */
    getFullyQualifiedName() {
        return this.getFullyQualifiedNameRecursive(this);
    }
}
