// @mcpfusion/core/dist/core/middleware/AuditTrail.js
// ── SHA-256 Helper ───────────────────────────────────────
/**
 * Compute SHA-256 hash of a string.
 * Uses Web Crypto API (available in Node 18+).
 *
 * @internal
 */
async function sha256Hex(input) {
    try {
        const encoder = new TextEncoder();
        const data = encoder.encode(input);
        const hashBuffer = await globalThis.crypto.subtle.digest('SHA-256', data);
        const hashArray = Array.from(new Uint8Array(hashBuffer));
        return 'sha256:' + hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
    }
    catch {
        // Fallback for environments without crypto.subtle
        return 'sha256:unavailable';
    }
}
// ── Middleware Factory ───────────────────────────────────
/**
 * Create an AuditTrail middleware for SOC2/GDPR compliance.
 *
 * Logs every tool invocation with identity, args hash, status,
 * and duration. The audit event is emitted AFTER the handler
 * completes (or fails), capturing the full lifecycle.
 *
 * @param config - Audit trail configuration
 * @returns A middleware function compatible with `.use()`
 */
export function auditTrail(config) {
    const hashArgs = config.hashArgs !== false; // default: true
    const logResult = config.logResult ?? 'status';
    return async (ctx, args, next) => {
        const start = Date.now();
        const identity = config.extractIdentity ? config.extractIdentity(ctx) : {};
        // Compute args hash (async, but fast)
        const argsHash = hashArgs
            ? await sha256Hex(JSON.stringify(args))
            : 'none';
        let status = 'success';
        try {
            const result = await next();
            // Detect error/blocked status from response
            if (logResult === 'status') {
                const r = result;
                if (r != null && r['isError'] === true) {
                    const content = r['content'];
                    const text = content?.[0]?.text ?? '';
                    // extract the error code from the XML
                    // attribute instead of substring matching on full text
                    const codeMatch = text.match(/code="([^"]+)"/);
                    const errorCode = codeMatch?.[1] ?? '';
                    if (errorCode === 'RATE_LIMITED') {
                        status = 'rate_limited';
                    }
                    else if (errorCode === 'INPUT_REJECTED') {
                        status = 'firewall_blocked';
                    }
                    else {
                        status = 'error';
                    }
                }
            }
            emitAudit(config, args, identity, argsHash, status, start);
            return result;
        }
        catch (err) {
            status = 'error';
            emitAudit(config, args, identity, argsHash, status, start);
            throw err;
        }
    };
}
// ── Internal ─────────────────────────────────────────────
function emitAudit(config, args, identity, argsHash, status, start) {
    // Extract action from args using the configured field name
    const actionField = config.actionField ?? 'action';
    const action = typeof args[actionField] === 'string' ? args[actionField] : 'unknown';
    const tool = config.toolName ?? 'unknown';
    try {
        config.sink({
            type: 'security.audit',
            tool,
            action,
            identity,
            argsHash,
            status,
            durationMs: Date.now() - start,
            timestamp: Date.now(),
        });
    }
    catch {
        // Fire-and-forget — never break the handler
    }
}
