// @mcpfusion/core/dist/core/middleware/RateLimiter.js
import { toolError } from '../response.js';
/**
 * In-memory sliding window rate limit store.
 *
 * ⚠️ **Single-process only.** Each process maintains its own counters.
 * In multi-instance deployments (Kubernetes, PM2 cluster, serverless),
 * each instance has an independent counter — an attacker effectively
 * gets `max * instanceCount` requests. For distributed rate limiting,
 * implement {@link RateLimitStore} with a shared backend (Redis, Valkey).
 *
 * Automatic cleanup runs every `windowMs` to prune expired entries.
 */
export class InMemoryStore {
    _entries = new Map();
    _cleanupInterval;
    _windowMs;
    constructor(windowMs = 60_000) {
        this._windowMs = windowMs;
        this._cleanupInterval = setInterval(() => this._cleanup(), windowMs);
        // Ensure the interval doesn't prevent Node.js from exiting
        if (typeof this._cleanupInterval === 'object' && 'unref' in this._cleanupInterval) {
            this._cleanupInterval.unref();
        }
    }
    increment(key, windowMs) {
        const now = Date.now();
        const windowStart = now - windowMs;
        let entry = this._entries.get(key);
        if (!entry) {
            entry = { timestamps: [] };
            this._entries.set(key, entry);
        }
        // Prune expired timestamps (sliding window)
        entry.timestamps = entry.timestamps.filter(ts => ts > windowStart);
        // Check BEFORE pushing — rejected requests don't inflate the window
        const currentCount = entry.timestamps.length;
        // Calculate reset time: when the oldest request in the window expires
        // When the window is empty (all timestamps pruned),
        // resetMs should be 0 (window already reset), not windowMs.
        const oldestInWindow = entry.timestamps[0];
        const resetMs = oldestInWindow != null ? (oldestInWindow + windowMs) - now : 0;
        return {
            count: currentCount,
            resetMs: Math.max(0, resetMs),
            // Timestamp will be pushed by the caller only on success
        };
    }
    /**
     * Record a successful (non-rejected) request.
     * Only called when the request is under the limit.
     */
    record(key) {
        const entry = this._entries.get(key);
        if (entry) {
            entry.timestamps.push(Date.now());
        }
    }
    destroy() {
        clearInterval(this._cleanupInterval);
        this._entries.clear();
    }
    _cleanup() {
        const now = Date.now();
        for (const [key, entry] of this._entries) {
            // Remove entries with no recent timestamps
            if (entry.timestamps.length === 0 ||
                entry.timestamps[entry.timestamps.length - 1] < now - this._windowMs) {
                this._entries.delete(key);
            }
        }
    }
}
/**
 * Create a rate limiting middleware.
 *
 * Returns a self-healing `toolError('RATE_LIMITED')` with `retryAfterSeconds`
 * when the limit is exceeded, following RFC 7231 semantics.
 *
 * The returned middleware exposes a `destroy()` method to clean up the
 * underlying store (intervals, connections). Call it on server shutdown
 * or in test `afterEach` hooks to prevent timer leaks.
 *
 * @param config - Rate limit configuration
 * @returns A middleware function compatible with `.use()`, with `destroy()`
 */
export function rateLimit(config) {
    const store = config.store ?? new InMemoryStore(config.windowMs);
    const errorCode = config.errorCode ?? 'RATE_LIMITED';
    const keyFn = config.keyFn ?? (() => '__global__');
    const middleware = async (ctx, args, next) => {
        const key = keyFn(ctx);
        const entry = await store.increment(key, config.windowMs);
        if (entry.count >= config.max) {
            const retryAfterSeconds = Math.ceil(entry.resetMs / 1000);
            if (config.onRejected) {
                try {
                    config.onRejected(ctx, key);
                }
                catch { /* fire-and-forget */ }
            }
            // Emit telemetry event
            if (config.telemetry) {
                try {
                    config.telemetry({
                        type: 'security.rateLimit',
                        key,
                        count: entry.count,
                        max: config.max,
                        retryAfterSeconds,
                        timestamp: Date.now(),
                    });
                }
                catch { /* fire-and-forget */ }
            }
            return toolError(errorCode, {
                message: `Rate limit exceeded. Maximum ${config.max} requests per ${Math.ceil(config.windowMs / 1000)}s window.`,
                suggestion: `Wait ${retryAfterSeconds} seconds before retrying. Current key: "${key}".`,
                retryAfter: retryAfterSeconds,
            });
        }
        // Record the request ONLY if not rate-limited
        await store.record(key);
        return next();
    };
    middleware.destroy = () => store.destroy();
    return middleware;
}
