// @mcpfusion/core/dist/core/execution/ExecutionPipeline.js
/**
 * ExecutionPipeline — Orchestrates MCP Tool Execution Steps
 *
 * Breaks the monolithic execute() flow into discrete, testable steps
 * using the Result monad for railway-oriented error handling.
 *
 * Each step either succeeds (passes data to the next step) or fails
 * (short-circuits with an error response).
 *
 * Pipeline: ensureBuilt → parseDiscriminator → resolveAction → validateArgs → runChain
 */
import {} from 'zod';
import { error, escapeXml, toolError } from '../response.js';
import { toErrorMessage } from '../ErrorUtils.js';
import { formatValidationError } from './ValidationErrorFormatter.js';
import { succeed, fail } from '../result.js';
import {} from '../types.js';
import {} from './MiddlewareCompiler.js';
import { isProgressEvent } from './ProgressHelper.js';
import { postProcessResult } from '../../presenter/PostProcessor.js';
/** Intentional no-op — used to suppress unhandled rejection warnings on best-effort cleanup. */
const noop = () => { };
// ── Pipeline Steps (pure functions) ──────────────────────
/** Step 1: Parse discriminator value from raw args */
export function parseDiscriminator(execCtx, args) {
    const raw = args[execCtx.discriminator];
    // distinguish missing from wrong-type discriminator.
    if (raw !== undefined && typeof raw !== 'string') {
        const text = [
            `<tool_error code="INVALID_DISCRIMINATOR">`,
            `<message>Discriminator field "${escapeXml(execCtx.discriminator)}" must be a string, got ${typeof raw}.</message>`,
            `<available_actions>${escapeXml(execCtx.actionKeysString)}</available_actions>`,
            `<recovery>Set "${escapeXml(execCtx.discriminator)}" to one of the available actions as a string.</recovery>`,
            `</tool_error>`,
        ].join('\n');
        return fail({ content: [{ type: 'text', text }], isError: true });
    }
    const value = typeof raw === 'string' ? raw : undefined;
    if (!value) {
        const text = [
            `<tool_error code="MISSING_DISCRIMINATOR">`,
            `<message>The required field "${escapeXml(execCtx.discriminator)}" is missing.</message>`,
            `<available_actions>${escapeXml(execCtx.actionKeysString)}</available_actions>`,
            `<recovery>Add the "${escapeXml(execCtx.discriminator)}" field as a string and call the tool again.</recovery>`,
            `</tool_error>`,
        ].join('\n');
        return fail({ content: [{ type: 'text', text }], isError: true });
    }
    return succeed(value);
}
/** Step 2: Resolve the action by discriminator value — O(1) lookup */
export function resolveAction(execCtx, discriminatorValue) {
    const action = execCtx.actionMap.get(discriminatorValue);
    if (!action) {
        const text = [
            `<tool_error code="UNKNOWN_ACTION">`,
            `<message>The ${escapeXml(execCtx.discriminator)} "${escapeXml(discriminatorValue)}" does not exist.</message>`,
            `<available_actions>${escapeXml(execCtx.actionKeysString)}</available_actions>`,
            `<recovery>Choose a valid action from available_actions and call the tool again.</recovery>`,
            `</tool_error>`,
        ].join('\n');
        return fail({ content: [{ type: 'text', text }], isError: true });
    }
    return succeed({ action, discriminatorValue });
}
/** Step 3: Validate and strip args using pre-cached Zod schema */
export function validateArgs(execCtx, resolved, args) {
    const validationSchema = execCtx.validationSchemaCache.get(resolved.action.key);
    // Extract _select before validation — it's a framework-level field
    // that must NOT reach the .strict() Zod schema.
    const rawSelect = args['_select'];
    const selectFields = (Array.isArray(rawSelect) && rawSelect.every(v => typeof v === 'string'))
        ? rawSelect
        : undefined;
    if (!validationSchema) {
        // No schema — strip _select (framework field) but preserve discriminator
        // (handlers rely on it). Re-inject with resolved value like the with-schema path.
        const { _select: _sel, ...cleaned } = args;
        return succeed({ validated: cleaned, selectFields });
    }
    // Remove discriminator AND _select before validation
    const { [execCtx.discriminator]: _, _select: _unused, ...argsToValidate } = args;
    const result = validationSchema.safeParse(argsToValidate);
    if (!result.success) {
        const text = formatValidationError(result.error.issues, `${execCtx.toolName}/${resolved.discriminatorValue}`, argsToValidate);
        // formatValidationError already produces complete XML — bypass error() to avoid double-wrapping
        return fail({ content: [{ type: 'text', text }], isError: true });
    }
    // Mutate directly — zero-copy re-injection of discriminator
    const validated = result.data;
    // Guard against prototype-pollution via poisoned discriminator names
    const disc = execCtx.discriminator;
    if (disc === '__proto__' || disc === 'constructor' || disc === 'prototype') {
        return fail({
            content: [{ type: 'text', text: `Invalid discriminator name: "${disc}".` }],
            isError: true,
        });
    }
    validated[disc] = resolved.discriminatorValue;
    return succeed({ validated, selectFields });
}
/**
 * Step 4: Run pre-compiled middleware chain → handler.
 *
 * @param rethrow - When `true`, handler exceptions propagate to the caller
 *   instead of being caught and converted to error responses. Used by the
 *   traced execution path so that `_executeTraced` can classify system errors
 *   (`SpanStatusCode.ERROR` + `recordException`). Default: `false`.
 * @param signal - Optional AbortSignal for cooperative cancellation.
 *   Checked before handler execution. If already aborted, returns an
 *   immediate error response without invoking the handler chain.
 * @param selectFields - Optional `_select` field names extracted from the
 *   AI's input. Forwarded to `postProcessResult()` → `Presenter.make()`
 *   for Late Guillotine filtering.
 */
export async function runChain(execCtx, resolved, ctx, args, progressSink, rethrow = false, signal, selectFields, telemetry) {
    const chain = execCtx.compiledChain.get(resolved.action.key);
    if (!chain) {
        return error(`No compiled chain for action "${resolved.action.key}".`);
    }
    // Cancellation gate: abort before starting the handler chain
    if (signal?.aborted) {
        return error(`[${execCtx.toolName}/${resolved.discriminatorValue}] Request cancelled.`);
    }
    try {
        const result = await chain(ctx, args);
        // If the middleware chain returned a generator result envelope, drain it
        if (isGeneratorResultEnvelope(result)) {
            const drained = await drainGenerator(result.generator, progressSink, signal);
            return postProcessResult(drained, resolved.action.returns, ctx, selectFields, telemetry);
        }
        return postProcessResult(result, resolved.action.returns, ctx, selectFields, telemetry);
    }
    catch (err) {
        if (rethrow)
            throw err;
        const message = toErrorMessage(err);
        return toolError('INTERNAL_ERROR', {
            message: `[${execCtx.toolName}/${resolved.discriminatorValue}] ${message}`,
            suggestion: 'This may be a transient error. Retry the same call with identical parameters.',
            severity: 'error',
        });
    }
}
/** @internal */
function isGeneratorResultEnvelope(value) {
    return (typeof value === 'object' &&
        value !== null &&
        '__brand' in value &&
        value.__brand === 'GeneratorResultEnvelope');
}
/**
 * Drain an async generator, forwarding ProgressEvents to the sink
 * and returning the final ToolResponse.
 *
 * Checks the AbortSignal before each iteration. If cancelled, the
 * generator is returned (cleanup runs) and an error response is returned.
 * This prevents zombie generators from continuing to execute after
 * the user cancels the request.
 *
 * @internal
 */
async function drainGenerator(gen, progressSink, signal) {
    // build a reusable abort promise so Promise.race can cancel
    // during `await gen.next()`, preventing zombie handlers on slow I/O.
    const abortPromise = signal && !signal.aborted
        ? new Promise((_, reject) => {
            signal.addEventListener('abort', () => {
                reject(new DOMException('Request cancelled.', 'AbortError'));
            }, { once: true });
        })
        : undefined;
    // Suppress unhandled rejection if the generator finishes before abort fires
    abortPromise?.catch(noop);
    let result = await gen.next();
    while (!result.done) {
        // Cancellation check: abort generator if signal fired between iterations
        if (signal?.aborted) {
            await gen.return(error('Request cancelled.'));
            return error('Request cancelled.');
        }
        if (progressSink && isProgressEvent(result.value)) {
            progressSink(result.value);
        }
        // race next iteration against abort signal to prevent
        // zombie generators that block on slow I/O (DB queries, network, etc.)
        if (abortPromise) {
            try {
                result = await Promise.race([gen.next(), abortPromise]);
            }
            catch (err) {
                if (err instanceof DOMException && err.name === 'AbortError') {
                    // Fire-and-forget cleanup — gen.return() may also block
                    // if the generator is stuck on slow I/O, so don't await it.
                    gen.return(error('Request cancelled.')).catch(noop);
                    return error('Request cancelled.');
                }
                throw err;
            }
        }
        else {
            result = await gen.next();
        }
    }
    return result.value;
}
