// @mcpfusion/core/dist/core/execution/PipelineHooks.js
/**
 * PipelineHooks — Strategy Pattern for Execution Observability
 *
 * Defines the hook interface used to instrument the tool execution
 * pipeline without duplicating the core flow. The fast path passes
 * no hooks (zero overhead); debug and traced paths supply their
 * strategy via factory methods.
 *
 * @module
 */
import {} from '../response.js';
// ── Utilities ────────────────────────────────────────────
/**
 * Compute the UTF-8 byte size of a ToolResponse.
 *
 * Sums the byte length of all text content blocks.
 * Uses TextEncoder for UTF-8 accurate byte measurement,
 * consistent with EgressGuard's byte measurement.
 *
 * Used by tracing to record `mcp.response_size` on spans.
 */
const _sizeEncoder = new TextEncoder();
export function computeResponseSize(response) {
    let size = 0;
    for (const c of response.content) {
        if ('text' in c && typeof c.text === 'string')
            size += _sizeEncoder.encode(c.text).byteLength;
    }
    return size;
}
// ── Hook Merging ─────────────────────────────────────────
/**
 * Merge two PipelineHooks into one that calls both in sequence.
 *
 * When both `primary` and `secondary` define the same hook,
 * the merged hook calls `primary` first, then `secondary`.
 *
 * Returns `primary` if `secondary` is undefined (zero allocation).
 *
 * Used to layer telemetry emission on top of debug or traced hooks.
 *
 * @internal
 */
export function mergeHooks(primary, secondary) {
    if (!secondary)
        return primary;
    const merged = {
        rethrow: !!(primary.rethrow || secondary.rethrow),
        onRouteError: () => { primary.onRouteError?.(); secondary.onRouteError?.(); },
        onRouteOk: (a) => { primary.onRouteOk?.(a); secondary.onRouteOk?.(a); },
        onResolveError: (a) => { primary.onResolveError?.(a); secondary.onResolveError?.(a); },
        onValidateError: (a, d) => { primary.onValidateError?.(a, d); secondary.onValidateError?.(a, d); },
        onValidateOk: (a, d) => { primary.onValidateOk?.(a, d); secondary.onValidateOk?.(a, d); },
        onMiddleware: (a, c) => { primary.onMiddleware?.(a, c); secondary.onMiddleware?.(a, c); },
        onExecuteOk: (a, r) => { primary.onExecuteOk?.(a, r); secondary.onExecuteOk?.(a, r); },
        onExecuteError: (a, e) => { primary.onExecuteError?.(a, e); secondary.onExecuteError?.(a, e); },
    };
    if (primary.wrapResponse) {
        merged.wrapResponse = (r) => { const wrapped = primary.wrapResponse(r); return secondary.wrapResponse?.(wrapped) ?? wrapped; };
    }
    else if (secondary.wrapResponse) {
        merged.wrapResponse = secondary.wrapResponse;
    }
    return merged;
}
