// @mcpfusion/core/dist/sandbox/SandboxEngine.js
/**
 * SandboxEngine — Zero-Trust V8 Isolate for Computation Delegation
 *
 * Allows LLMs to send JavaScript functions as strings to be executed
 * in a sealed V8 isolate. The data stays on the client's machine,
 * only the result travels back to the LLM.
 *
 * Architecture (V8 Engineering Rules):
 *   1. ONE Isolate per SandboxEngine (boot ~5-10ms), reused across requests
 *   2. NEW Context per execute() call (~0.1ms), pristine and empty
 *   3. ExternalCopy + Script + Context are ALWAYS released in `finally`
 *   4. Execution is ALWAYS async (script.run, never runSync)
 *   5. Context is empty — no process, require, fs, globalThis injected
 *   6. AbortSignal kills isolate.dispose() instantly (Connection Watchdog)
 *
 * The `isolated-vm` package is a peerDependency (optional).
 * If not installed, the engine throws a clear error at construction time.
 *
 *   ┌─────────────────────────────────────────────────────────┐
 *   │  SandboxEngine (owns 1 Isolate)                         │
 *   │                                                         │
 *   │  execute(code, data, { signal? })                       │
 *   │    ┌──────────┐                                         │
 *   │    │ Abort?   │ pre-flight signal check                  │
 *   │    ├──────────┤                                         │
 *   │    │ Guard    │ fail-fast syntax check                   │
 *   │    ├──────────┤                                         │
 *   │    │ Context  │ new per request (empty!)                 │
 *   │    ├──────────┤                                         │
 *   │    │ Copy In  │ ExternalCopy (deep, no refs)             │
 *   │    ├──────────┤                                         │
 *   │    │ Compile  │ isolate.compileScript                    │
 *   │    ├──────────┤                                         │
 *   │    │ Run      │ script.run (ASYNC, with timeout + abort) │
 *   │    ├──────────┤                                         │
 *   │    │ Copy Out │ JSON.parse result                        │
 *   │    └──────────┘                                         │
 *   │                                                         │
 *   │  finally: signal.removeEventListener()                  │
 *   │           inputCopy.release()                           │
 *   │           script.release()                              │
 *   │           context.release()                             │
 *   └─────────────────────────────────────────────────────────┘
 *
 * @module
 */
import { validateSandboxCode } from './SandboxGuard.js';
import {} from '../observability/TelemetryEvent.js';
import { toErrorMessage } from '../core/ErrorUtils.js';
// ── Constants ────────────────────────────────────────────
const DEFAULT_TIMEOUT_MS = 5_000;
const DEFAULT_MEMORY_LIMIT_MB = 128;
const DEFAULT_MAX_OUTPUT_BYTES = 1_048_576; // 1MB
const MAX_CODE_LENGTH = 65_536; // 64KB — generous for any legitimate sandbox function
const TEXT_ENCODER = new TextEncoder(); // reuse stateless encoder
// ── Lazy Require ─────────────────────────────────────────
/**
 * Lazy-load isolated-vm to avoid hard dependency.
 * Returns `null` if the package is not installed.
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
let _ivm = undefined;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getIvm() {
    if (_ivm !== undefined)
        return _ivm;
    try {
        // Dynamic import would be cleaner but isolated-vm uses
        // native bindings that require synchronous resolution.
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        _ivm = require('isolated-vm');
    }
    catch {
        _ivm = null;
    }
    return _ivm;
}
/**
 * Reset the cached isolated-vm module reference.
 * Exported exclusively for testing — allows mock/unmock cycles
 * without process restart ().
 * @internal
 */
export function resetIvmCache() {
    _ivm = undefined;
}
// ── Engine Implementation ────────────────────────────────
/**
 * Zero-trust V8 sandbox for executing LLM-provided JavaScript.
 *
 * Creates a single V8 `Isolate` at construction time and reuses it
 * across all `execute()` calls. Each call gets a fresh, empty `Context`
 * with no dangerous globals (no `process`, `require`, `fs`, etc.).
 *
 * If the isolate dies (e.g., OOM), it is automatically recreated
 * on the next `execute()` call.
 *
 * @example
 * ```typescript
 * const sandbox = new SandboxEngine({ timeout: 3000, memoryLimit: 64 });
 *
 * const result = await sandbox.execute(
 *     '(data) => data.filter(d => d.risk > 90)',
 *     [{ name: 'A', risk: 95 }, { name: 'B', risk: 30 }],
 * );
 *
 * if (result.ok) {
 *     console.log(result.value); // [{ name: 'A', risk: 95 }]
 * }
 *
 * // IMPORTANT: dispose when no longer needed
 * sandbox.dispose();
 * ```
 */
export class SandboxEngine {
    _timeout;
    _memoryLimit;
    _maxOutputBytes;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    _isolate; // ivm.Isolate
    _disposed = false;
    _telemetry;
    /** Number of in-flight execute() calls sharing the isolate. */
    _activeExecutions = 0;
    constructor(config) {
        this._timeout = config?.timeout ?? DEFAULT_TIMEOUT_MS;
        this._memoryLimit = config?.memoryLimit ?? DEFAULT_MEMORY_LIMIT_MB;
        this._maxOutputBytes = config?.maxOutputBytes ?? DEFAULT_MAX_OUTPUT_BYTES;
        const ivm = getIvm();
        if (ivm == null) {
            throw new Error('SandboxEngine requires the "isolated-vm" package. ' +
                'Install it with: npm install isolated-vm');
        }
        this._isolate = new ivm.Isolate({ memoryLimit: this._memoryLimit });
    }
    /**
     * Set the telemetry sink for `sandbox.exec` event emission.
     * When set, every `execute()` call emits a telemetry event
     * visible in the Inspector TUI.
     */
    telemetry(sink) {
        this._telemetry = sink;
        return this;
    }
    /**
     * Execute a JavaScript function string against the provided data.
     *
     * The function is compiled and run in a sealed V8 isolate with:
     * - No `process`, `require`, `fs`, or network access
     * - Strict timeout enforcement (async, non-blocking)
     * - Memory limit enforcement
     * - Automatic C++ pointer cleanup (ExternalCopy, Script, Context)
     * - Cooperative cancellation via AbortSignal (Connection Watchdog)
     *
     * @param code - A JavaScript function expression as a string.
     *   Must be an arrow function or function expression.
     *   Example: `(data) => data.filter(d => d.value > 10)`
     *
     * @param data - The data to pass into the function.
     *   Deeply copied into the isolate (no references leak).
     *
     * @param options - Optional execution options.
     * @param options.signal - AbortSignal for cooperative cancellation.
     *   When the signal fires (e.g., MCP client disconnects), the engine
     *   calls `isolate.dispose()` to kill the V8 C++ threads instantly.
     *   The isolate is auto-recovered on the next `.execute()` call.
     *
     * @returns A `SandboxResult` with the computed value or an error.
     */
    async execute(code, data, options) {
        if (this._disposed) {
            return { ok: false, error: 'SandboxEngine has been disposed.', code: 'UNAVAILABLE' };
        }
        const signal = options?.signal;
        // ── Step 0: Pre-flight abort check ───────────────
        // If the signal is already aborted (client disconnected before
        // we even started), skip all V8 allocation entirely.
        if (signal?.aborted) {
            return {
                ok: false,
                error: 'Execution aborted: client disconnected before sandbox started.',
                code: 'ABORTED',
            };
        }
        // ── Step 0b: Code length guard ──────────────────
        // Prevent host-side OOM before V8 is even involved.
        // The V8 memoryLimit only bounds the isolate heap, not the host.
        if (code.length > MAX_CODE_LENGTH) {
            return {
                ok: false,
                error: `Code length (${code.length} chars) exceeds maximum (${MAX_CODE_LENGTH} chars). ` +
                    'Sandbox functions should be small, pure transformations.',
                code: 'INVALID_CODE',
            };
        }
        // ── Step 1: Fail-fast guard ─────────────────────
        const guard = validateSandboxCode(code);
        if (!guard.ok) {
            return { ok: false, error: guard.violation, code: 'INVALID_CODE' };
        }
        // ── Step 2: Ensure isolate is alive ─────────────
        this._ensureIsolate();
        const ivm = getIvm();
        const isolate = this._isolate;
        this._activeExecutions++;
        // ── Step 3: Wire abort kill-switch ──────────────
        // When the signal fires and this is the ONLY execution in
        // progress, we call isolate.dispose() to kill V8 instantly.
        //
        // Bug fix: `context` is assigned later via `await isolate.createContext()`,
        // so a plain closure over `let context` would capture `undefined` if the abort
        // fires before that line. We use a mutable reference object (`ctxRef`) so the
        // closure always sees the latest assigned value.
        //
        // When other executions share the same isolate
        // (activeExecutions > 1), we cannot dispose the isolate
        // without killing all concurrent work ( fix). Instead
        // we set the `aborted` flag AND release the per-request
        // context (if already created) to interrupt the running
        // script without collateral damage. If context is not yet
        // created, the script will terminate at its timeout boundary.
        let aborted = false;
        // Mutable wrapper so the abort closure always reads the latest context ref.
        const ctxRef = { current: undefined };
        const onAbort = signal ? () => {
            aborted = true;
            if (this._activeExecutions <= 1) {
                try {
                    isolate.dispose();
                }
                catch { /* may already be dead */ }
            }
            else {
                // Release this request's context to interrupt execution
                // without killing the shared isolate ().
                try {
                    ctxRef.current?.release();
                }
                catch { /* may already be released */ }
            }
        } : undefined;
        if (signal && onAbort) {
            signal.addEventListener('abort', onAbort, { once: true });
        }
        // ── Step 4: Execute in sealed context ───────────
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        let inputCopy; // ivm.ExternalCopy
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        let context; // ivm.Context
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        let script; // ivm.Script
        const startMs = Date.now();
        try {
            // Create pristine context (NO globals injected — this IS the security)
            context = await isolate.createContext();
            // Keep ctxRef in sync so the abort handler can release it if needed.
            ctxRef.current = context;
            // Deep-copy data into isolated heap (no references!)
            // catch serialization errors from ExternalCopy separately
            try {
                inputCopy = new ivm.ExternalCopy(data);
            }
            catch (copyErr) {
                const copyMsg = copyErr instanceof Error ? copyErr.message : String(copyErr);
                const result = {
                    ok: false,
                    error: `Data serialization error: ${copyMsg}. The input data contains non-serializable values (functions, Symbols, WeakRefs, etc.).`,
                    code: 'INVALID_DATA',
                };
                this._emitTelemetry(result);
                return result;
            }
            await context.global.set('__input__', inputCopy.copyInto());
            // Compile with wrapper: call the function and serialize result
            const wrappedCode = `const __fn__ = ${code};\nJSON.stringify(__fn__(__input__));`;
            script = await isolate.compileScript(wrappedCode);
            // ASYNC execution — never blocks the Node.js event loop
            const rawResult = await script.run(context, { timeout: this._timeout });
            const executionMs = Date.now() - startMs;
            // ── Step 5: Output size guard ───────────────
            if (typeof rawResult === 'string') {
                const outputByteLength = TEXT_ENCODER.encode(rawResult).byteLength;
                if (outputByteLength > this._maxOutputBytes) {
                    const oversized = {
                        ok: false,
                        error: `Output size (${outputByteLength} bytes) exceeds limit (${this._maxOutputBytes} bytes). ` +
                            'Use more selective filters to reduce the result set.',
                        code: 'OUTPUT_TOO_LARGE',
                    };
                    this._emitTelemetry(oversized);
                    return oversized;
                }
            }
            // ── Step 6: Parse result ────────────────────
            const parsed = typeof rawResult === 'string' ? JSON.parse(rawResult) : rawResult;
            const result = { ok: true, value: parsed, executionMs };
            this._emitTelemetry(result);
            return result;
        }
        catch (err) {
            // If the abort listener disposed the isolate, classify as ABORTED
            // (not MEMORY — the user disconnected, not an OOM condition)
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- false positive: `aborted` is mutated by the onAbort event listener closure (L334)
            if (aborted) {
                const result = {
                    ok: false,
                    error: 'Execution aborted: client disconnected during sandbox execution.',
                    code: 'ABORTED',
                };
                this._emitTelemetry(result);
                return result;
            }
            const result = this._classifyError(err);
            this._emitTelemetry(result);
            return result;
        }
        finally {
            // ── MANDATORY ABORT LISTENER CLEANUP ─────────
            // Remove the listener to prevent memory leaks when
            // execution completes before the signal fires.
            if (signal && onAbort) {
                signal.removeEventListener('abort', onAbort);
            }
            // ── MANDATORY C++ POINTER RELEASE ────────────
            // Order matters: release inner resources first.
            // After abort-triggered dispose, these will throw
            // but the catch blocks handle dead-isolate gracefully.
            try {
                inputCopy?.release();
            }
            catch { /* already released or isolate dead */ }
            try {
                script?.release();
            }
            catch { /* already released or isolate dead */ }
            try {
                context?.release();
            }
            catch { /* already released or isolate dead */ }
            this._activeExecutions--;
        }
    }
    /**
     * Release all resources held by this engine.
     *
     * After calling `dispose()`, any subsequent `execute()` calls
     * will return `{ ok: false, code: 'UNAVAILABLE' }`.
     */
    dispose() {
        if (this._disposed)
            return;
        this._disposed = true;
        try {
            this._isolate?.dispose();
        }
        catch {
            // Isolate may already be dead (OOM)
        }
    }
    /**
     * Check if the engine has been disposed.
     */
    get isDisposed() {
        return this._disposed;
    }
    /**
     * Emit a `sandbox.exec` telemetry event if a sink is configured.
     * @internal
     */
    _emitTelemetry(result) {
        if (!this._telemetry)
            return;
        const event = {
            type: 'sandbox.exec',
            ok: result.ok,
            executionMs: result.ok ? result.executionMs : 0,
            ...(!result.ok ? { errorCode: result.code } : {}),
            timestamp: Date.now(),
        };
        this._telemetry(event);
    }
    // ── Private ──────────────────────────────────────────
    /**
     * Ensure the isolate is alive. If it died (OOM), create a new one.
     * @internal
     */
    _ensureIsolate() {
        const ivm = getIvm();
        // Check if isolate is still usable
        try {
            if (this._isolate?.isDisposed === true) {
                this._isolate = new ivm.Isolate({ memoryLimit: this._memoryLimit });
            }
        }
        catch {
            // isDisposed threw → isolate is dead, recreate
            this._isolate = new ivm.Isolate({ memoryLimit: this._memoryLimit });
        }
    }
    /**
     * Classify an error from V8 execution into a typed SandboxResult.
     * @internal
     */
    _classifyError(err) {
        const message = toErrorMessage(err);
        // Timeout: isolated-vm throws a specific error
        if (message.includes('Script execution timed out')) {
            return {
                ok: false,
                error: `Script execution timed out (${this._timeout}ms). ` +
                    'Simplify the computation or increase the timeout limit.',
                code: 'TIMEOUT',
            };
        }
        // Memory: V8 kills the isolate
        if (message.includes('Isolate was disposed') ||
            message.includes('out of memory') ||
            message.includes('allocation failed')) {
            // Mark isolate for recreation on next call
            try {
                this._isolate?.dispose();
            }
            catch { /* ignore */ }
            return {
                ok: false,
                error: `Isolate ran out of memory (${this._memoryLimit}MB limit). ` +
                    'Reduce the data size or simplify the computation.',
                code: 'MEMORY',
            };
        }
        // Syntax: V8 compilation error
        if (message.includes('SyntaxError')) {
            return {
                ok: false,
                error: `JavaScript syntax error: ${message}`,
                code: 'SYNTAX',
            };
        }
        // Runtime: any other V8 error (ReferenceError, TypeError, etc.)
        return {
            ok: false,
            error: `Runtime error: ${message}`,
            code: 'RUNTIME',
        };
    }
}
