// @mcpfusion/core/dist/core/createGroup.js
/**
 * createGroup() — Functional Tool Group Compiler
 *
 * A lightweight, closure-based alternative to the class-heavy
 * `GroupedToolBuilder`. Instead of maintaining internal state via `this._x`,
 * it compiles a group config into a frozen object with pre-composed
 * middleware and an O(1) dispatch function.
 *
 * **Why Functional?**
 * - Closures minify 30-40% better than class methods (Terser can rename local vars)
 * - No prototype chain overhead — zero `this` binding issues
 * - Frozen return type prevents accidental mutation
 * - Compatible with Edge Runtimes (Cloudflare Workers, Deno Deploy)
 *
 * @example
 * ```typescript
 * import { createGroup, success } from '@mcpfusion/core';
 *
 * const billing = createGroup({
 *   name: 'billing',
 *   description: 'Invoice management',
 *   middleware: [requireAuth],
 *   actions: {
 *     get_invoice: {
 *       schema: z.object({ id: z.string() }),
 *       readOnly: true,
 *       handler: async (ctx, args) => success(await ctx.db.invoices.get(args.id)),
 *     },
 *     pay: {
 *       schema: z.object({ invoice_id: z.string(), amount: z.number() }),
 *       destructive: true,
 *       handler: async (ctx, args) => success(await ctx.db.payments.create(args)),
 *     },
 *   },
 * });
 *
 * // billing.execute(ctx, 'get_invoice', { id: '123' })
 * // billing.name === 'billing'
 * // billing.actionNames === ['get_invoice', 'pay']
 * ```
 *
 * @module
 */
import {} from 'zod';
import { error as toolError } from '../core/response.js';
import {} from '../core/types.js';
// ── Middleware Composition ────────────────────────────────
/**
 * Compose an array of middleware into a single function.
 * Right-to-left composition: last middleware is innermost (closest to handler).
 * @internal
 */
function composeMiddleware(globalMiddleware, actionMiddleware, handler) {
    const allMiddleware = [...globalMiddleware, ...actionMiddleware];
    if (allMiddleware.length === 0) {
        return handler;
    }
    // Build the chain from right to left
    return allMiddleware.reduceRight((next, mw) => (ctx, args) => mw(ctx, args, () => next(ctx, args)), handler);
}
// ── Factory ──────────────────────────────────────────────
/**
 * Create a compiled, frozen tool group from a declarative config.
 *
 * The returned object has O(1) action dispatch via a pre-built Map.
 * All middleware chains are pre-composed at creation time — zero
 * runtime overhead on each call.
 *
 * @typeParam TContext - Application context type
 * @param config - Group configuration with actions
 * @returns A frozen {@link CompiledGroup} ready for execution
 *
 * @example
 * ```typescript
 * const tasks = createGroup({
 *   name: 'tasks',
 *   middleware: [logMiddleware],
 *   actions: {
 *     list: {
 *       readOnly: true,
 *       handler: async (ctx) => success(await ctx.db.tasks.findMany()),
 *     },
 *     create: {
 *       schema: z.object({ title: z.string() }),
 *       handler: async (ctx, args) => success(await ctx.db.tasks.create(args)),
 *     },
 *   },
 * });
 *
 * const result = await tasks.execute(ctx, 'create', { title: 'Buy milk' });
 * ```
 */
export function createGroup(config) {
    const globalMiddleware = config.middleware ?? [];
    const actionNames = Object.keys(config.actions);
    // Pre-compose middleware chains for each action (O(1) dispatch)
    const dispatchMap = new Map();
    for (const [name, action] of Object.entries(config.actions)) {
        const compiled = composeMiddleware(globalMiddleware, action.middleware ?? [], action.handler);
        dispatchMap.set(name, compiled);
    }
    // Pre-validate action args using Zod schemas
    const schemaMap = new Map();
    for (const [name, action] of Object.entries(config.actions)) {
        if (action.schema) {
            schemaMap.set(name, action.schema);
        }
    }
    const execute = async (ctx, action, args) => {
        const chain = dispatchMap.get(action);
        if (!chain) {
            return toolError(`Unknown action "${action}" in group "${config.name}". Available: ${actionNames.join(', ')}`, 'INVALID_PARAMS');
        }
        // Validate with Zod if schema is defined — use safeParse to return
        // ToolResponse instead of throwing ZodError (contract compliance).
        // Uses the consumer's schema as-is to respect their unknown-keys policy
        // (.passthrough(), .strip(), or .strict()).
        const schema = schemaMap.get(action);
        if (schema) {
            const result = schema.safeParse(args);
            if (!result.success) {
                const issues = result.error.issues
                    .map(i => `${i.path.join('.')}: ${i.message}`)
                    .join('; ');
                return toolError(`Validation failed for action "${action}" in group "${config.name}": ${issues}`, 'INVALID_PARAMS');
            }
            // Use result.data to preserve Zod transforms, defaults, and strip behavior
            return chain(ctx, result.data);
        }
        return chain(ctx, args);
    };
    const getAction = (name) => {
        return config.actions[name];
    };
    return Object.freeze({
        name: config.name,
        description: config.description,
        tags: Object.freeze(config.tags ?? []),
        actionNames: Object.freeze(actionNames),
        execute,
        getAction,
    });
}
