// @mcpfusion/core/dist/core/index.js
/**
 * Core — Barrel Export
 *
 * Public API for the core tool consolidation engine.
 * Contains only the foundational building blocks: response helpers,
 * result monad, types, builder, registry, schema, execution, and middleware.
 */
// ── Cross-cutting ────────────────────────────────────────
export { success, error, required, toonSuccess, toolError, handoff, isHandoffResponse, TOOL_RESPONSE_BRAND } from './response.js';
export { succeed, fail } from './result.js';
// ── Builder ──────────────────────────────────────────────
export { GroupedToolBuilder, ActionGroupBuilder, createTool, defineTool } from './builder/index.js';
export { FluentToolBuilder } from './builder/index.js';
export { FluentRouter } from './builder/index.js';
// ── Registry ─────────────────────────────────────────────
export { ToolRegistry } from './registry/index.js';
// ── Schema (public strategies) ───────────────────────────
export { generateToonDescription } from './schema/index.js';
// ── Execution (progress streaming + runtime guards) ──────
export { progress } from './execution/index.js';
// ── Middleware (context derivation) ──────────────────────
export { defineMiddleware, resolveMiddleware } from './middleware/index.js';
// ── MCP mcpfusion Initializer (tRPC-style) ─────────────────────
export { initMCPFusion } from './initMCPFusion.js';
// ── Functional Core ──────────────────────────────────────
export { createGroup } from './createGroup.js';
// ── Standard Schema ──────────────────────────────────────
export { toStandardValidator, fromZodSchema, isStandardSchema, autoValidator, } from './StandardSchema.js';
