// @mcpfusion/core/dist/core/initMCPFusion.js
/**
 * initMCPFusion() — tRPC-Style Context Initialization
 *
 * Eliminates the need to pass `<AppContext>` as a generic parameter
 * everywhere. Define your context type once, and every `f.query()`,
 * `f.mutation()`, `f.action()` call automatically inherits it.
 *
 * @example
 * ```typescript
 * // src/mcpfusion.ts — defined once in the project
 * import { initMCPFusion } from '@mcpfusion/core';
 *
 * interface AppContext {
 *   db: PrismaClient;
 *   user: { id: string; role: string };
 * }
 *
 * export const f = initMCPFusion<AppContext>();
 *
 * // src/tools/billing.ts — clean fluent API
 * import { f } from '../mcpfusion';
 *
 * export const getInvoice = f.query('billing.get_invoice')
 *     .describe('Get an invoice by ID')
 *     .withString('id', 'The invoice ID')
 *     .handle(async (input, ctx) => {
 *         return await ctx.db.invoices.findUnique(input.id);
 *     });
 * ```
 *
 * @module
 */
import {} from 'zod';
import { ToolRegistry } from './registry/ToolRegistry.js';
import {} from '../presenter/Presenter.js';
import { definePresenter } from '../presenter/definePresenter.js';
import { defineMiddleware } from './middleware/index.js';
import { defineTool } from './builder/defineTool.js';
import { definePrompt } from '../prompt/definePrompt.js';
import { FluentPromptBuilder } from '../prompt/FluentPromptBuilder.js';
import {} from '../prompt/types.js';
import { FluentToolBuilder, QUERY_DEFAULTS, MUTATION_DEFAULTS, ACTION_DEFAULTS, } from './builder/FluentToolBuilder.js';
import { FluentRouter } from './builder/FluentRouter.js';
import { ErrorBuilder } from './builder/ErrorBuilder.js';
import { StateSyncBuilder } from '../state-sync/StateSyncBuilder.js';
import {} from './response.js';
import { SandboxEngine } from '../sandbox/SandboxEngine.js';
import { defaultSerializer } from './serialization/JsonSerializer.js';
import { StateMachineGate } from '../fsm/StateMachineGate.js';
// ── Factory ──────────────────────────────────────────────
/**
 * Initialize a MCP Fusion instance with a fixed context type.
 *
 * Call once per project. All factory methods on the returned instance
 * automatically inherit the context type — zero generic repetition.
 *
 * @typeParam TContext - The application-level context type
 * @returns A {@link MCPFusionInstance} with context-typed factories
 *
 * @example
 * ```typescript
 * // Single definition, typically in src/mcpfusion.ts
 * export const f = initMCPFusion<AppContext>();
 *
 * // Build tools with the Fluent API
 * const listUsers = f.query('users.list')
 *     .describe('List all users')
 *     .withOptionalNumber('limit', 'Max results (default: 50)')
 *     .handle(async (input, ctx) => {
 *         return ctx.db.users.findMany({ take: input.limit ?? 50 });
 *     });
 * ```
 */
export function initMCPFusion() {
    return {
        // ── Semantic Verbs ────────────────────────────────
        query(name) {
            return new FluentToolBuilder(name, QUERY_DEFAULTS);
        },
        mutation(name) {
            return new FluentToolBuilder(name, MUTATION_DEFAULTS);
        },
        action(name) {
            return new FluentToolBuilder(name, ACTION_DEFAULTS);
        },
        // ── MVA Presenter ────────────────────────────────
        presenter(config) {
            return definePresenter(config);
        },
        // ── Prompts ──────────────────────────────────────
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        prompt(name, config) {
            if (config == null) {
                return new FluentPromptBuilder(name);
            }
            return definePrompt(name, config);
        },
        // ── Middleware ───────────────────────────────────
        middleware(derive) {
            return defineMiddleware(derive);
        },
        // ── Registry ─────────────────────────────────────
        registry() {
            return new ToolRegistry();
        },
        // ── Router ───────────────────────────────────────
        router(prefix) {
            return new FluentRouter(prefix);
        },
        // ── Error Builder ────────────────────────────────
        error(code, message) {
            return new ErrorBuilder(code, message);
        },
        // ── State Sync ───────────────────────────────────
        stateSync() {
            return new StateSyncBuilder();
        },
        // ── Internal / Advanced ──────────────────────────
        defineTool(name, config) {
            return defineTool(name, config);
        },
        // ── Sandbox ──────────────────────────────────────
        sandbox(config) {
            const engine = new SandboxEngine(config);
            // warn when engines are garbage-collected without dispose().
            // V8 Isolates hold native memory invisible to the Node GC.
            _trackSandboxDispose(engine);
            return engine;
        },
        // ── Serialization ────────────────────────────────
        serializer: defaultSerializer,
        // ── FSM State Gate ─────────────────────────────
        fsm(config) {
            return new StateMachineGate(config);
        },
    };
}
// ── FinalizationRegistry leak-detection for SandboxEngine ──
/**
 * Track SandboxEngine instances and emit a console.warn when one is
 * garbage-collected without being disposed first. V8 Isolates hold
 * native C++ memory invisible to the Node.js GC — a forgotten
 * `.dispose()` silently leaks until process exit.
 *
 * Uses `FinalizationRegistry` (ES2021). No-op if the API is unavailable.
 * The warning is best-effort — GC timing is non-deterministic.
 * @internal
 */
const _sandboxFinalizer = typeof FinalizationRegistry !== 'undefined'
    ? new FinalizationRegistry((label) => {
        console.warn(`[mcpfusion] SandboxEngine was garbage-collected without dispose(). ` +
            `This leaks native V8 Isolate memory. Call engine.dispose() when done. (${label})`);
    })
    : undefined;
/** @internal */
function _trackSandboxDispose(engine) {
    if (!_sandboxFinalizer)
        return;
    const label = `created at ${new Date().toISOString()}`;
    _sandboxFinalizer.register(engine, label, engine);
    // When dispose() is called, unregister so no false warning fires
    const originalDispose = engine.dispose.bind(engine);
    engine.dispose = () => {
        _sandboxFinalizer.unregister(engine);
        originalDispose();
    };
}
