// @mcpfusion/core/dist/core/response.js
/**
 * Response Helpers
 *
 * Universal MCP response builders. No API coupling.
 * These produce the standard MCP ToolResponse format.
 *
 * @example
 * ```typescript
 * import { success, error, toonSuccess } from '@mcpfusion/core';
 *
 * // String response
 * return success('Project created');
 *
 * // Object response (auto JSON.stringify)
 * return success({ id: '123', name: 'My Project' });
 *
 * // Error response
 * return error('Project not found');
 *
 * // TOON-encoded response (~40% fewer tokens)
 * return toonSuccess(users);
 * ```
 *
 * @see {@link ToolResponse} for the response shape
 * @see {@link toonSuccess} for token-optimized responses
 *
 * @module
 */
import { encode } from '@toon-format/toon';
import {} from './serialization/JsonSerializer.js';
/**
 * Non-enumerable brand symbol stamped on all helper-created ToolResponse objects.
 * Used by FluentToolBuilder to reliably distinguish framework responses from
 * domain data that coincidentally matches the ToolResponse shape ().
 *
 * @internal
 */
export const TOOL_RESPONSE_BRAND = Symbol.for('mcpfusion.ToolResponse');
// ============================================================================
// XML Safety
// ============================================================================
/**
 * Escape XML structural characters for element content.
 *
 * Only `&` and `<` are mandatory escapes in XML element content.
 * `>` is preserved for LLM readability (e.g. `>= 1`, `Must be > 0`).
 * Single and double quotes are also preserved since they have no
 * special meaning outside attribute values.
 *
 * @internal
 */
export function escapeXml(str) {
    return str
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;');
}
/**
 * Escape XML special characters for use inside attribute values.
 *
 * Attribute values are delimited by `"` or `'` and must also
 * escape `<`, `>`, and `&`. All 5 XML special characters are handled.
 *
 * @internal
 */
export function escapeXmlAttr(str) {
    return str
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&apos;');
}
// ============================================================================
// Response Builders
// ============================================================================
/**
 * Create a success response from text or a JSON-serializable object.
 *
 * - Strings are returned verbatim (empty strings become `"OK"`)
 * - Objects are serialized with `JSON.stringify(data, null, 2)`
 *
 * @param data - A string message or any JSON-serializable object
 * @returns A {@link ToolResponse} with `isError` unset
 *
 * @example
 * ```typescript
 * // String response
 * return success('Task completed');
 *
 * // Object response (pretty-printed JSON)
 * const project = await db.projects.create({ name: 'Acme' });
 * return success(project);
 *
 * // Array response
 * const users = await db.users.findMany();
 * return success(users);
 * ```
 *
 * @see {@link error} for error responses
 * @see {@link toonSuccess} for token-optimized array responses
 */
export function success(data, compiledStringify) {
    const text = typeof data === 'string'
        ? (data || 'OK')
        : (compiledStringify ? compiledStringify(data) : JSON.stringify(data, null, 2));
    const resp = { content: [{ type: "text", text }] };
    Object.defineProperty(resp, TOOL_RESPONSE_BRAND, { value: true });
    return resp;
}
/**
 * Create an error response.
 *
 * Sets `isError: true` so the MCP client and LLM recognize the failure.
 * The LLM will typically retry or ask the user for clarification.
 *
 * @param message - Human-readable error description
 * @returns A {@link ToolResponse} with `isError: true`
 *
 * @example
 * ```typescript
 * // Simple error
 * return error('Project not found');
 *
 * // Contextual error
 * return error(`User "${userId}" does not have access to workspace "${wsId}"`);
 *
 * // In a handler with early return
 * handler: async (ctx, args) => {
 *     const project = await ctx.db.projects.findUnique(args.id);
 *     if (!project) return error(`Project "${args.id}" not found`);
 *     return success(project);
 * }
 * ```
 *
 * @see {@link required} for missing field errors
 * @see {@link success} for success responses
 */
export function error(message, code) {
    const codeAttr = code !== undefined ? ` code="${escapeXmlAttr(code)}"` : '';
    const resp = {
        content: [{ type: "text", text: `<tool_error${codeAttr}>\n<message>${escapeXml(message)}</message>\n</tool_error>` }],
        isError: true,
    };
    Object.defineProperty(resp, TOOL_RESPONSE_BRAND, { value: true });
    return resp;
}
/**
 * Create a validation error for a missing required field.
 *
 * Convenience shortcut for `error(\`Error: ${field} required\`)`.
 * Typically used in handlers that accept dynamic or optional schemas.
 *
 * @param field - Name of the missing required field
 * @returns A {@link ToolResponse} with `isError: true`
 *
 * @example
 * ```typescript
 * handler: async (ctx, args) => {
 *     if (!args.workspace_id) return required('workspace_id');
 *     // ...
 * }
 * ```
 *
 * @see {@link error} for general error responses
 */
export function required(field) {
    const f = escapeXml(field);
    const resp = {
        content: [{
                type: "text",
                text: `<tool_error code="MISSING_REQUIRED_FIELD">\n<message>Required field "${f}" is missing.</message>\n<recovery>Provide the "${f}" parameter and retry.</recovery>\n</tool_error>`,
            }],
        isError: true,
    };
    Object.defineProperty(resp, TOOL_RESPONSE_BRAND, { value: true });
    return resp;
}
/**
 * Create a success response with TOON-encoded payload.
 *
 * Encodes structured data using TOON (Token-Oriented Object Notation)
 * for ~40-50% token reduction compared to `JSON.stringify()`.
 * Ideal for list/tabular responses (arrays of uniform objects).
 *
 * @param data - Any JSON-serializable value (objects, arrays, primitives)
 * @param options - Optional TOON encode options (default: pipe delimiter)
 * @returns A {@link ToolResponse} with TOON-encoded text
 *
 * @example
 * ```typescript
 * // Array response — saves ~40% tokens vs JSON
 * const users = await db.users.findMany();
 * return toonSuccess(users);
 * // Output: "id|name|email\n1|Alice|alice@co.io\n2|Bob|bob@co.io"
 *
 * // With custom delimiter
 * return toonSuccess(data, { delimiter: ',' });
 *
 * // Single object (still valid, but savings are smaller)
 * return toonSuccess({ id: 1, name: 'Alice' });
 * ```
 *
 * @see {@link success} for standard JSON responses
 */
export function toonSuccess(data, options) {
    const defaults = { delimiter: '|' };
    const text = encode(data, { ...defaults, ...options });
    const resp = { content: [{ type: "text", text }] };
    Object.defineProperty(resp, TOOL_RESPONSE_BRAND, { value: true });
    return resp;
}
/**
 * Create a self-healing error response with recovery instructions.
 *
 * Unlike {@link error}, this provides structured guidance so the LLM
 * agent can self-correct instead of hallucinating or giving up.
 * The response includes an error code, message, suggestion, and
 * available actions — all formatted for optimal LLM comprehension.
 *
 * @param code - Short error code (e.g. `'ProjectNotFound'`, `'Unauthorized'`)
 * @param options - Error details and recovery instructions
 * @returns A {@link ToolResponse} with `isError: true` and recovery guidance
 *
 * @example
 * ```typescript
 * handler: async (ctx, args) => {
 *     const project = await ctx.db.get(args.project_id);
 *
 *     if (!project) {
 *         return toolError('ProjectNotFound', {
 *             message: `Project '${args.project_id}' does not exist.`,
 *             suggestion: 'Call projects.list first to get valid IDs, then retry.',
 *             availableActions: ['projects.list'],
 *         });
 *     }
 *
 *     return success(project);
 * }
 * ```
 *
 * @example
 * ```typescript
 * // Minimal usage (no suggestion)
 * return toolError('RateLimited', {
 *     message: 'Too many requests. Wait 30 seconds.',
 * });
 * ```
 *
 * @see {@link error} for simple error responses
 * @see {@link required} for missing field errors
 */
export function toolError(code, options) {
    const severity = options.severity ?? 'error';
    const parts = [
        `<tool_error code="${escapeXmlAttr(code)}" severity="${escapeXmlAttr(severity)}">`,
        `<message>${escapeXml(options.message)}</message>`,
    ];
    if (options.suggestion) {
        parts.push(`<recovery>${escapeXml(options.suggestion)}</recovery>`);
    }
    if ((options.availableActions?.length ?? 0) > 0) {
        parts.push('<available_actions>');
        for (const action of options.availableActions) {
            parts.push(`  <action>${escapeXml(action)}</action>`);
        }
        parts.push('</available_actions>');
    }
    if (options.details != null && Object.keys(options.details).length > 0) {
        parts.push('<details>');
        for (const [key, value] of Object.entries(options.details)) {
            parts.push(`  <detail key="${escapeXmlAttr(key)}">${escapeXml(value)}</detail>`);
        }
        parts.push('</details>');
    }
    if (options.retryAfter !== undefined && Number.isFinite(options.retryAfter) && options.retryAfter > 0) {
        parts.push(`<retry_after>${options.retryAfter} seconds</retry_after>`);
    }
    parts.push('</tool_error>');
    // Warnings are non-fatal — do not set isError so the response
    // flows through the success path while still carrying guidance.
    const isError = severity !== 'warning';
    const resp = { content: [{ type: "text", text: parts.join('\n') }], isError };
    Object.defineProperty(resp, TOOL_RESPONSE_BRAND, { value: true });
    return resp;
}
/**
 * Type-guard used by `ServerAttachment` to detect a `HandoffResponse`
 * without importing `@mcpfusion/swarm` (avoids circular dependency).
 */
export function isHandoffResponse(v) {
    if (typeof v !== 'object' || v === null)
        return false;
    const r = v;
    return r['_MCPFUSION_handoff'] === true && r['isHandoff'] === true;
}
/**
 * Initiate a federated agent handoff from a tool handler.
 *
 * The `ServerAttachment` detects this response and activates the
 * `SwarmGateway` tunnel to the specified upstream server.
 *
 * @param target  - Target server URI (e.g. `'mcp://finance-agent.internal:8080'`)
 * @param options - Carry-over state, system instruction, and tracing hints
 *
 * @example
 * ```typescript
 * import { f } from '@mcpfusion/core';
 *
 * export const triage = f.tool('system.triage')
 *     .withEnum('domain', ['finance', 'devops'])
 *     .withString('context', 'What the user needs.')
 *     .handle(async (input) => f.handoff(`mcp://${input.domain}-agent`, {
 *         carryOverState: { intent: input.context },
 *         reason: `Triage → ${input.domain}`,
 *     }));
 * ```
 */
export function handoff(target, options) {
    const resp = {
        _MCPFUSION_handoff: true,
        isHandoff: true,
        payload: { target, ...options },
    };
    Object.defineProperty(resp, TOOL_RESPONSE_BRAND, { value: true });
    return resp;
}
