// @mcpfusion/core/dist/core/StandardSchema.js
/**
 * StandardSchema — Universal Schema Abstraction Layer
 *
 * Decouples the MCP Fusion validation engine from Zod specifically,
 * enabling support for any validator that implements the Standard Schema
 * specification (`@standard-schema/spec`).
 *
 * This allows users to choose lighter alternatives:
 * - **Zod** (~14kb min): Full-featured, most popular
 * - **Valibot** (~1kb min via tree-shaking): Ultra-lightweight
 * - **ArkType** (~5kb min): Fastest runtime validation
 * - **TypeBox** (~4kb min): JSON Schema native
 *
 * ## Standard Schema Spec
 *
 * Any object with `~standard` property conforming to:
 * ```typescript
 * interface StandardSchema {
 *   '~standard': {
 *     version: 1;
 *     vendor: string;
 *     validate: (value: unknown) => { value: T } | { issues: Issue[] };
 *   }
 * }
 * ```
 *
 * @see https://github.com/standard-schema/standard-schema
 *
 * @example
 * ```typescript
 * import { toStandardValidator } from '@mcpfusion/core';
 * import * as v from 'valibot';
 *
 * // Valibot schemas work natively via Standard Schema
 * const schema = v.object({ name: v.string(), age: v.number() });
 * const validator = toStandardValidator(schema);
 *
 * const result = validator.validate({ name: 'Alice', age: 30 });
 * // { success: true, data: { name: 'Alice', age: 30 } }
 * ```
 *
 * @module
 */
// ── Adapters ─────────────────────────────────────────────
/**
 * Create a MCPFusionValidator from a Standard Schema v1 compatible schema.
 *
 * This is the primary entry point for non-Zod validators. Any schema
 * library implementing the Standard Schema spec can be used directly.
 *
 * @param schema - A Standard Schema v1 compatible schema
 * @returns A {@link MCPFusionValidator} wrapping the schema
 *
 * @example
 * ```typescript
 * import * as v from 'valibot'; // ~1kb tree-shaken
 *
 * const schema = v.object({ name: v.string() });
 * const validator = toStandardValidator(schema);
 *
 * const ok = validator.validate({ name: 'Alice' });
 * // { success: true, data: { name: 'Alice' } }
 *
 * const err = validator.validate({ name: 42 });
 * // { success: false, issues: [{ message: 'Expected string', path: ['name'] }] }
 * ```
 */
export function toStandardValidator(schema) {
    const spec = schema['~standard'];
    return {
        validate(value) {
            const result = spec.validate(value);
            // Guard: async validators return a Promise, which would silently
            // produce { success: false, issues: undefined } because
            // 'value' in Promise is false. Detect and throw early. ( fix)
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- defensive: Standard Schema implementations may violate sync contract
            if (result != null && typeof result.then === 'function') {
                throw new Error(`[mcpfusion] Schema validator "${spec.vendor}" returned a Promise. ` +
                    'Async validators are not supported — use a synchronous schema. ' +
                    'See: https://mcpfusion.vinkius.com/docs/standard-schema');
            }
            if ('value' in result) {
                return { success: true, data: result.value };
            }
            return { success: false, issues: result.issues };
        },
        vendor: spec.vendor,
        schema,
    };
}
/**
 * Create a MCPFusionValidator from a raw Zod schema.
 *
 * This adapter uses Zod's `.safeParse()` method and maps the result
 * to the standard MCPFusionValidator interface.
 *
 * @param schema - A Zod schema (z.object, z.string, etc.)
 * @returns A {@link MCPFusionValidator} wrapping the Zod schema
 *
 * @example
 * ```typescript
 * import { z } from 'zod';
 *
 * const schema = z.object({ name: z.string() });
 * const validator = fromZodSchema(schema);
 *
 * const ok = validator.validate({ name: 'Alice' });
 * // { success: true, data: { name: 'Alice' } }
 * ```
 */
export function fromZodSchema(schema) {
    return {
        validate(value) {
            const result = schema.safeParse(value);
            if (result.success) {
                return { success: true, data: result.data };
            }
            // Map Zod errors to StandardSchemaIssue
            const issues = (result.error.issues).map((issue) => {
                const mapped = { message: issue.message };
                if (issue.path) {
                    mapped.path = issue.path;
                }
                return mapped;
            });
            return { success: false, issues };
        },
        vendor: 'zod',
        schema,
    };
}
/**
 * Check if a value implements the Standard Schema v1 spec.
 *
 * @param value - Any value to check
 * @returns `true` if the value has a valid `~standard` property
 */
export function isStandardSchema(value) {
    if (typeof value !== 'object' || value === null)
        return false;
    if (!('~standard' in value))
        return false;
    const std = value['~standard'];
    return typeof std === 'object' && std !== null && std['version'] === 1;
}
/**
 * Auto-detect and create a MCPFusionValidator from any supported schema.
 *
 * Detection order:
 * 1. Standard Schema v1 (Valibot, ArkType, etc.)
 * 2. Zod-like (has `.safeParse()`)
 * 3. Throws if unrecognized
 *
 * @param schema - Any supported schema
 * @returns A {@link MCPFusionValidator}
 * @throws If the schema type is not recognized
 */
export function autoValidator(schema) {
    if (isStandardSchema(schema)) {
        return toStandardValidator(schema);
    }
    if (isZodLike(schema)) {
        return fromZodSchema(schema);
    }
    throw new Error('Unsupported schema type. Expected a Standard Schema v1 (' +
        'Valibot, ArkType) or Zod schema. See: https://mcpfusion.vinkius.com/docs/standard-schema');
}
/** Duck-type check for Zod-like schemas */
function isZodLike(value) {
    return (typeof value === 'object' &&
        value !== null &&
        typeof value.safeParse === 'function');
}
