// @mcpfusion/core/dist/core/elicitation/ask.js
/**
 * ask — Callable Namespace for MCP Elicitation
 *
 * The `ask` export is a **single symbol** that works as:
 * 1. A callable function: `await ask('message', { fields })`
 * 2. A DSL namespace: `ask.string()`, `ask.enum()`, `ask.boolean()`, `ask.number()`
 * 3. A URL redirect: `await ask.redirect('message', url)`
 *
 * The transport context (`sendRequest`) is bound per-request via
 * `AsyncLocalStorage` — the developer never sees it.
 *
 * @example
 * ```typescript
 * import { initMCPFusion, ask } from '@mcpfusion/core';
 *
 * const f = initMCPFusion<AppContext>();
 *
 * const deploy = f.mutation('infra.deploy')
 *     .withString('app_id', 'Application ID')
 *     .interactive()
 *     .handle(async (input) => {
 *         const prefs = await ask('Confirm deployment:', {
 *             region:  ask.enum(['us-east-1', 'eu-west-1'] as const, 'Region'),
 *             confirm: ask.boolean('I confirm this deployment'),
 *         });
 *
 *         if (prefs.declined) return f.error('CANCELLED', 'Aborted');
 *         return { region: prefs.data.region };
 *     });
 * ```
 *
 * @module
 */
import { AsyncLocalStorage } from 'node:async_hooks';
import { AskStringField, AskNumberField, AskBooleanField, AskEnumField, ElicitationUnsupportedError, createAskResponse, } from './types.js';
// ── AsyncLocalStorage (Per-Request Transport) ────────────
/**
 * Per-request storage for the elicitation transport function.
 *
 * Bound by `GroupedToolBuilder._executePipeline()` when the tool
 * declares `.interactive()` and the MCP SDK provides `sendRequest`.
 *
 * @internal
 */
export const _elicitStore = new AsyncLocalStorage();
// ── Field Compiler ───────────────────────────────────────
/**
 * Compile a record of `AskField<T>` descriptors into a JSON Schema
 * object suitable for `elicitation/create`.
 *
 * @param fields - Object of `ask.*` field descriptors
 * @returns JSON Schema `{ type: 'object', properties, required }`
 *
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function compileAskFields(fields) {
    const properties = {};
    const required = [];
    for (const [key, field] of Object.entries(fields)) {
        properties[key] = field._compile();
        // All fields are required by default (no optional fields in elicitation forms)
        required.push(key);
    }
    return { type: 'object', properties, required };
}
/**
 * Resolve the `ElicitSink` from `AsyncLocalStorage`.
 * Throws `ElicitationUnsupportedError` if not in an `.interactive()` context.
 * @internal
 */
function getSink() {
    const sink = _elicitStore.getStore();
    if (!sink)
        throw new ElicitationUnsupportedError();
    return sink;
}
/**
 * `ask` — The Callable Namespace for MCP Elicitation.
 *
 * Works as both a function (`await ask('msg', { fields })`) and a
 * namespace (`ask.string()`, `ask.enum()`, `ask.redirect()`).
 *
 * Transport context is bound via `AsyncLocalStorage` — zero `ctx` needed.
 *
 * @example
 * ```typescript
 * import { ask } from '@mcpfusion/core';
 *
 * // DSL — field descriptors
 * ask.string('Name')
 * ask.number('Age').min(18).max(120)
 * ask.boolean('Confirm').default(true)
 * ask.enum(['free', 'pro'] as const, 'Plan')
 *
 * // Form mode — request structured input
 * const result = await ask('Fill in:', { name: ask.string('Name') });
 *
 * // URL mode — redirect to external page
 * const auth = await ask.redirect('Authenticate:', oauthUrl);
 * ```
 */
export const ask = Object.assign(
// ── Callable: await ask('message', { fields }) ──────
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function ask(message, fields) {
    const sink = getSink();
    const schema = compileAskFields(fields);
    const raw = await sink({
        method: 'elicitation/create',
        params: {
            message,
            requestedSchema: schema,
        },
    });
    return createAskResponse(raw);
}, 
// ── Namespace: static field descriptor factories ────
{
    string(description) {
        return new AskStringField(description);
    },
    number(description) {
        return new AskNumberField(description);
    },
    boolean(description) {
        return new AskBooleanField(description);
    },
    enum(values, description) {
        return new AskEnumField(values, description);
    },
    async redirect(message, url) {
        const sink = getSink();
        const raw = await sink({
            method: 'elicitation/create',
            params: { message, url },
        });
        return createAskResponse(raw);
    },
});
