// @mcpfusion/core/dist/core/elicitation/types.js
/**
 * Elicitation — Human-in-the-Loop Interaction Types
 *
 * MCP Elicitation (Form + URL modes) type declarations.
 * Provides the `AskField<T>` phantom-typed field descriptors and the
 * `AskResponse<T>` result wrapper with boolean guards.
 *
 * @module
 */
/**
 * Base class for all `ask.*` field descriptors.
 *
 * Uses a phantom `__type` brand so TypeScript can infer the
 * concrete type from the DSL without manual generic annotations.
 *
 * @typeParam T - The inferred TypeScript type for this field
 *
 * @example
 * ```typescript
 * // The developer never interacts with this class directly.
 * // They use `ask.string()`, `ask.number()`, etc.
 * ```
 */
export class AskField {
    /** @internal */
    _description;
    /** @internal */
    _defaultValue;
    constructor(description) {
        this._description = description;
    }
    /**
     * Set a default value for this field.
     *
     * @param value - Default value shown pre-filled in the client form
     * @returns `this` for chaining
     */
    default(value) {
        this._defaultValue = value;
        return this;
    }
    /**
     * Set or override the field description.
     *
     * @param desc - Human-readable label/description
     * @returns `this` for chaining
     */
    describe(desc) {
        this._description = desc;
        return this;
    }
}
// ── Concrete Field Descriptors ───────────────────────────
/** String field descriptor — `ask.string('Name')` */
export class AskStringField extends AskField {
    /** @internal */
    _compile() {
        return {
            type: 'string',
            ...(this._description ? { description: this._description } : {}),
            ...(this._defaultValue !== undefined ? { default: this._defaultValue } : {}),
        };
    }
}
/** Number field descriptor — `ask.number('Age').min(18).max(120)` */
export class AskNumberField extends AskField {
    /** @internal */
    _min;
    /** @internal */
    _max;
    /**
     * Set the minimum value constraint.
     * @param n - Minimum (inclusive)
     * @returns `this` for chaining
     */
    min(n) {
        this._min = n;
        return this;
    }
    /**
     * Set the maximum value constraint.
     * @param n - Maximum (inclusive)
     * @returns `this` for chaining
     */
    max(n) {
        this._max = n;
        return this;
    }
    /** @internal */
    _compile() {
        return {
            type: 'number',
            ...(this._description ? { description: this._description } : {}),
            ...(this._defaultValue !== undefined ? { default: this._defaultValue } : {}),
            ...(this._min !== undefined ? { minimum: this._min } : {}),
            ...(this._max !== undefined ? { maximum: this._max } : {}),
        };
    }
}
/** Boolean field descriptor — `ask.boolean('Confirm').default(true)` */
export class AskBooleanField extends AskField {
    /** @internal */
    _compile() {
        return {
            type: 'boolean',
            ...(this._description ? { description: this._description } : {}),
            ...(this._defaultValue !== undefined ? { default: this._defaultValue } : {}),
        };
    }
}
/**
 * Enum field descriptor — `ask.enum(['a', 'b'] as const, 'Choose one')`
 *
 * @typeParam V - Literal union type inferred from the values array
 */
export class AskEnumField extends AskField {
    /** @internal */
    _values;
    constructor(values, description) {
        super(description);
        this._values = values;
    }
    /** @internal */
    _compile() {
        return {
            type: 'string',
            enum: [...this._values],
            ...(this._description ? { description: this._description } : {}),
            ...(this._defaultValue !== undefined ? { default: this._defaultValue } : {}),
        };
    }
}
// ── AskResponse Factory ─────────────────────────────────
/**
 * Create an `AskResponse<T>` from the raw MCP elicitation result.
 *
 * The `.data` getter uses a fail-fast pattern: accessing it when
 * the user declined/cancelled throws `ElicitationDeclinedError`.
 *
 * @internal
 */
export function createAskResponse(raw) {
    const action = (raw.action ?? 'cancel');
    const accepted = action === 'accept';
    const declined = action === 'decline';
    const cancelled = action === 'cancel';
    const content = raw.content;
    return {
        action,
        accepted,
        declined,
        cancelled,
        get data() {
            if (!accepted) {
                throw new ElicitationDeclinedError(action);
            }
            return content;
        },
    };
}
// ── Errors ───────────────────────────────────────────────
/**
 * Thrown when `ask()` or `ask.redirect()` is called outside an
 * `.interactive()` handler, or when the MCP client doesn't support
 * elicitation.
 */
export class ElicitationUnsupportedError extends Error {
    constructor() {
        super('[mcpfusion] Elicitation requested but no transport context is available. ' +
            'Ensure the tool uses `.interactive()` and the MCP client declares ' +
            '`{ capabilities: { elicitation: {} } }` during initialization.');
        this.name = 'ElicitationUnsupportedError';
    }
}
/**
 * Thrown when `.data` is accessed on an `AskResponse` that the user
 * declined or cancelled. Prevents silent `undefined` propagation.
 */
export class ElicitationDeclinedError extends Error {
    constructor(action) {
        super(`[mcpfusion] Cannot access .data — the user ${action === 'decline' ? 'declined' : 'cancelled'} the elicitation. ` +
            `Check .declined or .cancelled before accessing .data.`);
        this.name = 'ElicitationDeclinedError';
    }
}
