// @mcpfusion/core/dist/introspection/BehaviorDigest.js
import { sha256, canonicalize } from './canonicalize.js';
// ============================================================================
// Digest Computation
// ============================================================================
/**
 * Compute the behavioral digest for a single tool contract.
 *
 * The digest is deterministic: same contract → same digest.
 * It uses canonical JSON serialization (sorted keys) to ensure
 * platform-independent consistency.
 *
 * @param contract - The materialized tool contract
 * @returns Content-addressed digest with per-section components
 */
export async function computeDigest(contract) {
    const [surfaceHash, behaviorHash, tokenEconomicsHash, entitlementsHash] = await Promise.all([
        hashSection({
            name: contract.surface.name,
            description: contract.surface.description,
            tags: [...contract.surface.tags].sort(),
            inputSchemaDigest: contract.surface.inputSchemaDigest,
            actions: sortedActions(contract),
        }),
        hashSection({
            egressSchemaDigest: contract.behavior.egressSchemaDigest,
            systemRulesFingerprint: contract.behavior.systemRulesFingerprint,
            cognitiveGuardrails: contract.behavior.cognitiveGuardrails,
            middlewareChain: contract.behavior.middlewareChain,
            stateSyncFingerprint: contract.behavior.stateSyncFingerprint,
            concurrencyFingerprint: contract.behavior.concurrencyFingerprint,
            affordanceTopology: [...contract.behavior.affordanceTopology],
            embeddedPresenters: [...contract.behavior.embeddedPresenters],
        }),
        hashSection({
            schemaFieldCount: contract.tokenEconomics.schemaFieldCount,
            unboundedCollection: contract.tokenEconomics.unboundedCollection,
            baseOverheadTokens: contract.tokenEconomics.baseOverheadTokens,
            inflationRisk: contract.tokenEconomics.inflationRisk,
        }),
        hashSection({
            filesystem: contract.entitlements.filesystem,
            network: contract.entitlements.network,
            subprocess: contract.entitlements.subprocess,
            crypto: contract.entitlements.crypto,
            codeEvaluation: contract.entitlements.codeEvaluation,
            raw: [...contract.entitlements.raw].sort(),
        }),
    ]);
    // Composite digest over all components
    const compositeInput = [surfaceHash, behaviorHash, tokenEconomicsHash, entitlementsHash].join(':');
    const digest = await sha256(compositeInput);
    return {
        digest,
        components: {
            surface: surfaceHash,
            behavior: behaviorHash,
            tokenEconomics: tokenEconomicsHash,
            entitlements: entitlementsHash,
        },
        computedAt: new Date().toISOString(),
        toolName: contract.surface.name,
    };
}
/**
 * Compute a server-level digest covering all tools.
 *
 * The server digest is the SHA-256 over all per-tool digests,
 * sorted by tool name for determinism.
 *
 * @param contracts - Record of tool name → contract
 * @returns Server-level content-addressed digest
 */
export async function computeServerDigest(contracts) {
    const tools = {};
    const sortedNames = Object.keys(contracts).sort();
    const perToolDigests = [];
    for (const name of sortedNames) {
        const result = await computeDigest(contracts[name]);
        tools[name] = result;
        perToolDigests.push(`${name}:${result.digest}`);
    }
    const digest = await sha256(perToolDigests.join('\n'));
    return {
        digest,
        tools,
        computedAt: new Date().toISOString(),
    };
}
/**
 * Compare two server digests and return which tools changed.
 *
 * @param before - Previous server digest
 * @param after  - Current server digest
 * @returns Object with added, removed, and changed tool names
 */
export function compareServerDigests(before, after) {
    const beforeNames = new Set(Object.keys(before.tools));
    const afterNames = new Set(Object.keys(after.tools));
    const added = [];
    const removed = [];
    const changed = [];
    const unchanged = [];
    for (const name of afterNames) {
        if (!beforeNames.has(name)) {
            added.push(name);
        }
        else if (before.tools[name].digest !== after.tools[name].digest) {
            changed.push(name);
        }
        else {
            unchanged.push(name);
        }
    }
    for (const name of beforeNames) {
        if (!afterNames.has(name)) {
            removed.push(name);
        }
    }
    return {
        serverDigestChanged: before.digest !== after.digest,
        added,
        removed,
        changed,
        unchanged,
    };
}
// ============================================================================
// Internals
// ============================================================================
/**
 * Sort actions by key for deterministic hashing.
 * @internal
 */
function sortedActions(contract) {
    return Object.keys(contract.surface.actions)
        .sort()
        .map(key => ({
        key,
        ...contract.surface.actions[key],
    }));
}
/**
 * Hash a section object using canonical JSON serialization.
 * @internal
 */
function hashSection(obj) {
    return sha256(canonicalize(obj));
}
