// @mcpfusion/core/dist/introspection/CapabilityLockfile.js
/**
 * CapabilityLockfile — Behavioral Surface Snapshot
 *
 * Generates a deterministic, human-readable, git-diffable lockfile
 * that captures the complete behavioral surface of an MCP Fusion
 * server at a point in time.
 *
 * **Inspired by**: `yarn.lock`, `package-lock.json`, `Cargo.lock`.
 * Those files snapshot the dependency resolution graph so that every
 * `install` produces the same tree. `mcpfusion.lock` does the same
 * for the **behavioral surface** — ensuring that every deployment
 * exposes the same tool contracts, cognitive guardrails, entitlements,
 * and token economics that were reviewed and approved.
 *
 * **Why this matters for AI**:
 *
 * An LLM's calibration to a tool server depends on the *behavioral*
 * surface — not just the input schema. If a system rule changes, an
 * affordance link disappears, or a Presenter's egress schema mutates,
 * the LLM may hallucinate or violate requirements. The lockfile makes
 * these invisible changes **visible and auditable** in version control.
 *
 * **Workflow**:
 *
 * ```bash
 * # Generate or update the lockfile
 * mcpfusion lock
 *
 * # CI gate: fail if the lockfile is stale
 * mcpfusion lock --check
 * ```
 *
 * The lockfile is committed alongside the code. Pull request diffs
 * show exactly which behavioral surfaces changed, enabling reviewers
 * to assess AI-facing impact before merge.
 *
 * Pure-function module for generation and verification.
 * Side-effectful I/O is clearly separated.
 *
 * @module
 */
import { readFile, writeFile } from 'node:fs/promises';
import { resolve } from 'node:path';
import { computeServerDigest } from './BehaviorDigest.js';
import { sha256, canonicalize } from './canonicalize.js';
// ============================================================================
// Lockfile Schema
// ============================================================================
/** Current lockfile format version. */
const LOCKFILE_VERSION = 1;
/** Default lockfile name. */
export const LOCKFILE_NAME = 'mcpfusion.lock';
/**
 * Generate a `CapabilityLockfile` from compiled tool contracts
 * and (optionally) prompt builders.
 *
 * This is a **pure function**: given the same contracts and metadata,
 * it always produces the same lockfile (modulo `generatedAt` timestamp).
 *
 * @param serverName    - MCP server name
 * @param contracts     - Record of tool name → materialized contract
 * @param mcpfusionVersion - fusion version string
 * @param options       - Optional: prompt builders to include
 * @returns A fully materialized lockfile
 */
export async function generateLockfile(serverName, contracts, mcpfusionVersion, options) {
    const serverDigest = await computeServerDigest(contracts);
    const sortedNames = Object.keys(contracts).sort();
    const tools = {};
    for (const name of sortedNames) {
        tools[name] = snapshotTool(contracts[name], serverDigest.tools[name]);
    }
    // ── Prompt Snapshots ─────────────────────────────────
    const promptBuilders = options?.prompts ?? [];
    let prompts;
    const promptDigestParts = [];
    if (promptBuilders.length > 0) {
        prompts = {};
        // Deduplicate by name — last-registered builder wins
        const buildersByName = new Map();
        for (const b of promptBuilders) {
            buildersByName.set(b.getName(), b);
        }
        const sortedPromptNames = [...buildersByName.keys()].sort();
        for (const name of sortedPromptNames) {
            const builder = buildersByName.get(name);
            const snapshot = await snapshotPrompt(builder);
            prompts[name] = snapshot;
            promptDigestParts.push(`${name}:${snapshot.integrityDigest}`);
        }
    }
    // ── Integrity Digest (tools + prompts) ───────────────
    const digestInput = promptDigestParts.length > 0
        ? `${serverDigest.digest}\n---prompts---\n${promptDigestParts.join('\n')}`
        : serverDigest.digest;
    const integrityDigest = promptDigestParts.length > 0
        ? await sha256(digestInput)
        : serverDigest.digest;
    const capabilities = prompts
        ? { tools, prompts }
        : { tools };
    return {
        lockfileVersion: LOCKFILE_VERSION,
        serverName,
        mcpfusionVersion,
        generatedAt: new Date().toISOString(),
        integrityDigest: `sha256:${integrityDigest}`,
        capabilities,
    };
}
/**
 * Serialize a lockfile to a deterministic JSON string.
 *
 * Uses sorted keys and 2-space indentation for git-friendly diffs.
 * The output is canonical — identical lockfiles produce identical strings.
 *
 * @param lockfile - The lockfile to serialize
 * @returns Deterministic JSON string with trailing newline
 */
export function serializeLockfile(lockfile) {
    return JSON.stringify(lockfile, (_key, value) => {
        if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
            return Object.keys(value)
                .sort()
                .reduce((sorted, k) => {
                sorted[k] = value[k];
                return sorted;
            }, {});
        }
        return value;
    }, 2) + '\n';
}
/**
 * Verify that a lockfile matches the current server contracts and prompts.
 *
 * This is the **CI gate**: `mcpfusion lock --check` calls this function
 * and exits non-zero if the lockfile is stale.
 *
 * @param lockfile  - Previously generated lockfile (from disk)
 * @param contracts - Current compiled tool contracts (from code)
 * @param options   - Optional: prompt builders for prompt verification
 * @returns Check result with per-tool and per-prompt diff details
 */
export async function checkLockfile(lockfile, contracts, options) {
    // Regenerate a fresh lockfile to compare digests
    const fresh = await generateLockfile(lockfile.serverName, contracts, lockfile.mcpfusionVersion, options);
    // Fast path: integrity digest matches → everything is identical
    if (lockfile.integrityDigest === fresh.integrityDigest) {
        return {
            ok: true,
            message: 'Lockfile is up to date.',
            added: [],
            removed: [],
            changed: [],
            unchanged: Object.keys(contracts).sort(),
            addedPrompts: [],
            removedPrompts: [],
            changedPrompts: [],
            unchangedPrompts: Object.keys(fresh.capabilities.prompts ?? {}).sort(),
        };
    }
    // Slow path: per-tool comparison
    const lockedToolNames = new Set(Object.keys(lockfile.capabilities.tools));
    const currentToolNames = new Set(Object.keys(contracts));
    const added = [];
    const removed = [];
    const changed = [];
    const unchanged = [];
    for (const name of currentToolNames) {
        if (!lockedToolNames.has(name)) {
            added.push(name);
        }
        else {
            const lockedDigest = lockfile.capabilities.tools[name].integrityDigest;
            const freshDigest = fresh.capabilities.tools[name].integrityDigest;
            if (lockedDigest === freshDigest) {
                unchanged.push(name);
            }
            else {
                changed.push(name);
            }
        }
    }
    for (const name of lockedToolNames) {
        if (!currentToolNames.has(name)) {
            removed.push(name);
        }
    }
    // ── Per-prompt comparison ────────────────────────────
    const lockedPrompts = lockfile.capabilities.prompts ?? {};
    const freshPrompts = fresh.capabilities.prompts ?? {};
    const lockedPromptNames = new Set(Object.keys(lockedPrompts));
    const currentPromptNames = new Set(Object.keys(freshPrompts));
    const addedPrompts = [];
    const removedPrompts = [];
    const changedPrompts = [];
    const unchangedPrompts = [];
    for (const name of currentPromptNames) {
        if (!lockedPromptNames.has(name)) {
            addedPrompts.push(name);
        }
        else {
            const lockedDigest = lockedPrompts[name].integrityDigest;
            const freshDigest = freshPrompts[name].integrityDigest;
            if (lockedDigest === freshDigest) {
                unchangedPrompts.push(name);
            }
            else {
                changedPrompts.push(name);
            }
        }
    }
    for (const name of lockedPromptNames) {
        if (!currentPromptNames.has(name)) {
            removedPrompts.push(name);
        }
    }
    // ── Build message ────────────────────────────────────
    const driftParts = [];
    if (added.length > 0)
        driftParts.push(`tools added: [${added.join(', ')}]`);
    if (removed.length > 0)
        driftParts.push(`tools removed: [${removed.join(', ')}]`);
    if (changed.length > 0)
        driftParts.push(`tools changed: [${changed.join(', ')}]`);
    if (addedPrompts.length > 0)
        driftParts.push(`prompts added: [${addedPrompts.join(', ')}]`);
    if (removedPrompts.length > 0)
        driftParts.push(`prompts removed: [${removedPrompts.join(', ')}]`);
    if (changedPrompts.length > 0)
        driftParts.push(`prompts changed: [${changedPrompts.join(', ')}]`);
    return {
        ok: false,
        message: `Lockfile is stale. ${driftParts.join('; ')}. Run \`mcpfusion lock\` to update.`,
        added,
        removed,
        changed,
        unchanged,
        addedPrompts,
        removedPrompts,
        changedPrompts,
        unchangedPrompts,
    };
}
/**
 * Parse and validate a lockfile JSON string.
 *
 * @param content - Raw JSON string from `mcpfusion.lock`
 * @returns Parsed lockfile, or null if invalid
 */
export function parseLockfile(content) {
    try {
        const parsed = JSON.parse(content);
        if (parsed['lockfileVersion'] !== LOCKFILE_VERSION)
            return null;
        if (typeof parsed['serverName'] !== 'string')
            return null;
        if (typeof parsed['integrityDigest'] !== 'string')
            return null;
        if (typeof parsed['generatedAt'] !== 'string')
            return null;
        if (typeof parsed['mcpfusionVersion'] !== 'string')
            return null;
        if (parsed['capabilities'] == null || typeof parsed['capabilities'] !== 'object')
            return null;
        const caps = parsed['capabilities'];
        if (caps['tools'] == null || typeof caps['tools'] !== 'object')
            return null;
        return parsed;
    }
    catch {
        /* malformed JSON or unexpected structure — treat as missing */
        return null;
    }
}
// ============================================================================
// Persistence (Side-effectful — clearly separated)
// ============================================================================
/**
 * Write a lockfile to disk.
 *
 * @param lockfile    - The lockfile to persist
 * @param projectRoot - Project root directory
 */
export async function writeLockfile(lockfile, projectRoot) {
    const filePath = resolve(projectRoot, LOCKFILE_NAME);
    await writeFile(filePath, serializeLockfile(lockfile), 'utf-8');
}
/**
 * Read a lockfile from disk.
 *
 * @param projectRoot - Project root directory
 * @returns Parsed lockfile, or null if not found / invalid
 */
export async function readLockfile(projectRoot) {
    const filePath = resolve(projectRoot, LOCKFILE_NAME);
    try {
        const content = await readFile(filePath, 'utf-8');
        return parseLockfile(content);
    }
    catch {
        /* file not found or unreadable — treat as no lockfile */
        return null;
    }
}
// ============================================================================
// Internals
// ============================================================================
/**
 * Snapshot a single tool's contract into lockfile format.
 * @internal
 */
function snapshotTool(contract, digest) {
    const actionKeys = Object.keys(contract.surface.actions).sort();
    const destructiveActions = actionKeys.filter(key => contract.surface.actions[key].destructive);
    const readOnlyActions = actionKeys.filter(key => contract.surface.actions[key].readOnly);
    return {
        integrityDigest: `sha256:${digest.digest}`,
        surface: {
            description: contract.surface.description ?? null,
            actions: actionKeys,
            inputSchemaDigest: `sha256:${contract.surface.inputSchemaDigest}`,
            tags: [...contract.surface.tags].sort(),
        },
        behavior: {
            egressSchemaDigest: contract.behavior.egressSchemaDigest
                ? `sha256:${contract.behavior.egressSchemaDigest}`
                : null,
            systemRulesFingerprint: contract.behavior.systemRulesFingerprint,
            destructiveActions,
            readOnlyActions,
            middlewareChain: [...contract.behavior.middlewareChain],
            affordanceTopology: [...contract.behavior.affordanceTopology],
            cognitiveGuardrails: {
                agentLimitMax: contract.behavior.cognitiveGuardrails.agentLimitMax,
                egressMaxBytes: contract.behavior.cognitiveGuardrails.egressMaxBytes,
            },
        },
        tokenEconomics: {
            inflationRisk: contract.tokenEconomics.inflationRisk,
            schemaFieldCount: contract.tokenEconomics.schemaFieldCount,
            unboundedCollection: contract.tokenEconomics.unboundedCollection,
        },
        entitlements: {
            filesystem: contract.entitlements.filesystem,
            network: contract.entitlements.network,
            subprocess: contract.entitlements.subprocess,
            crypto: contract.entitlements.crypto,
            codeEvaluation: contract.entitlements.codeEvaluation,
        },
    };
}
/**
 * Snapshot a single prompt builder into lockfile format.
 *
 * Extracts all declarative metadata and computes a SHA-256 digest
 * over the canonical representation.
 *
 * @internal
 */
async function snapshotPrompt(builder) {
    const def = builder.buildPromptDefinition();
    const tags = [...builder.getTags()].sort();
    const hasMiddleware = builder.hasMiddleware();
    const hydrationTimeout = builder.getHydrationTimeout() ?? null;
    // Normalize arguments for deterministic serialization
    const args = (def.arguments ?? [])
        .map(a => ({
        name: a.name,
        description: a.description ?? null,
        required: a.required ?? false,
    }))
        .sort((a, b) => a.name.localeCompare(b.name));
    const argumentsDigest = `sha256:${await sha256(canonicalize(args))}`;
    // Compute integrity digest over all declarative fields
    const surface = {
        name: def.name,
        description: def.description ?? null,
        title: def.title ?? null,
        tags,
        arguments: args,
        hasMiddleware,
        hydrationTimeout,
    };
    const integrityDigest = `sha256:${await sha256(canonicalize(surface))}`;
    return {
        integrityDigest,
        description: def.description ?? null,
        title: def.title ?? null,
        tags,
        arguments: args,
        argumentsDigest,
        hasMiddleware,
        hydrationTimeout,
    };
}
