// @mcpfusion/core/dist/introspection/CryptoAttestation.js
// Lazy resolution instead of top-level await (breaks CJS consumers)
let _subtle;
async function getSubtle() {
    if (_subtle)
        return _subtle;
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- defensive: globalThis.crypto.subtle may not exist in all runtimes (Workers, Deno, older Node.js)
    _subtle = globalThis.crypto?.subtle
        ?? (await import('node:crypto')).webcrypto.subtle;
    return _subtle;
}
// ============================================================================
// HMAC Signer (Built-in)
// ============================================================================
/**
 * Create an HMAC-SHA256 signer from a shared secret.
 *
 * @param secret - The shared secret (should be ≥32 bytes)
 * @returns An `AttestationSigner` using HMAC-SHA256
 */
export function createHmacSigner(secret) {
    // Security #5: reject weak secrets
    if (secret.length < 32) {
        const msg = `[mcpfusion] HMAC secret must be at least 32 characters (got ${secret.length}). ` +
            `Use a cryptographically random value (e.g. openssl rand -hex 32).`;
        if (process.env['NODE_ENV'] === 'production') {
            throw new Error(msg);
        }
        console.warn(`⚠️  ${msg}`);
    }
    return {
        name: 'hmac-sha256',
        async sign(digest) {
            return hmacSign(digest, secret);
        },
        async verify(digest, signature) {
            const expected = await hmacSign(digest, secret);
            return timingSafeCompare(expected, signature);
        },
    };
}
// ============================================================================
// Attestation Operations
// ============================================================================
/**
 * Sign a server digest and produce an attestation result.
 *
 * @param serverDigest - The server's behavioral digest
 * @param config - Zero-trust configuration
 * @returns Attestation result with signature
 */
export async function attestServerDigest(serverDigest, config) {
    const signer = resolveSigner(config);
    const signature = await signer.sign(serverDigest.digest);
    const attestedAt = new Date().toISOString();
    // Verify against expected digest if configured
    if (config.expectedDigest) {
        const matches = serverDigest.digest === config.expectedDigest;
        return {
            valid: matches,
            computedDigest: serverDigest.digest,
            expectedDigest: config.expectedDigest,
            signature,
            signerName: signer.name,
            attestedAt,
            error: matches ? null : `Attestation failed: computed digest ${serverDigest.digest} does not match expected ${config.expectedDigest}`,
        };
    }
    return {
        valid: true,
        computedDigest: serverDigest.digest,
        expectedDigest: null,
        signature,
        signerName: signer.name,
        attestedAt,
        error: null,
    };
}
/**
 * Verify a previously signed attestation.
 *
 * @param serverDigest - The current server digest
 * @param signature - The signature to verify
 * @param config - Zero-trust configuration
 * @returns Attestation result with verification status
 */
export async function verifyAttestation(serverDigest, signature, config) {
    const signer = resolveSigner(config);
    const verified = await signer.verify(serverDigest.digest, signature);
    const attestedAt = new Date().toISOString();
    return {
        valid: verified,
        computedDigest: serverDigest.digest,
        expectedDigest: config.expectedDigest ?? null,
        signature,
        signerName: signer.name,
        attestedAt,
        error: verified ? null : `Signature verification failed for digest ${serverDigest.digest}`,
    };
}
/**
 * Runtime capability pinning — check that the current digest
 * matches what was attested at build time.
 *
 * Designed to be called once at server startup in the `attach()` flow.
 *
 * @param currentDigest - Re-computed server digest
 * @param config - Zero-trust configuration
 * @returns Attestation result
 * @throws If `failOnMismatch` is true and the digest doesn't match
 */
export async function verifyCapabilityPin(currentDigest, config) {
    const result = await attestServerDigest(currentDigest, config);
    if (!result.valid && (config.failOnMismatch ?? true)) {
        throw new AttestationError(`[mcpfusion] Zero-Trust attestation failed: ${result.error}`, result);
    }
    return result;
}
/**
 * Build the `mcpfusionTrust` capability object for MCP exposure.
 *
 * @param attestation - A completed attestation result
 * @param toolCount - Number of tools in the registry
 * @returns Capability structure for `server.capabilities`
 */
export function buildTrustCapability(attestation, toolCount) {
    return {
        serverDigest: attestation.computedDigest,
        signature: attestation.signature,
        signerName: attestation.signerName,
        attestedAt: attestation.attestedAt,
        toolCount,
        verified: attestation.valid,
    };
}
// ============================================================================
// Attestation Error
// ============================================================================
/**
 * Error thrown when zero-trust attestation fails with failOnMismatch.
 *
 * Carries the full `AttestationResult` for programmatic inspection.
 */
export class AttestationError extends Error {
    attestation;
    constructor(message, attestation) {
        super(message);
        this.name = 'AttestationError';
        this.attestation = attestation;
    }
}
// ============================================================================
// Internals
// ============================================================================
/**
 * Resolve the signer from config.
 * @internal
 */
function resolveSigner(config) {
    if (config.signer === 'hmac') {
        if (!config.secret) {
            throw new Error('[mcpfusion] HMAC signer requires a secret. Set zeroTrust.secret or use a custom signer.');
        }
        return createHmacSigner(config.secret);
    }
    return config.signer;
}
/**
 * HMAC-SHA256 sign using Web Crypto API.
 * @internal
 */
async function hmacSign(data, secret) {
    const encoder = new TextEncoder();
    const keyData = encoder.encode(secret);
    const msgData = encoder.encode(data);
    const subtle = await getSubtle();
    const key = await subtle.importKey('raw', keyData, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']);
    const signature = await subtle.sign('HMAC', key, msgData);
    return Array.from(new Uint8Array(signature))
        .map(b => b.toString(16).padStart(2, '0'))
        .join('');
}
/**
 * Constant-time string comparison to prevent timing attacks.
 *
 * Compares byte-by-byte using XOR accumulation — the comparison
 * always processes all bytes regardless of where differences occur.
 * @internal
 */
function timingSafeCompare(a, b) {
    const encoder = new TextEncoder();
    const bufA = encoder.encode(a);
    const bufB = encoder.encode(b);
    // Constant-time: always process max(lenA, lenB) bytes.
    // Length mismatch contributes to diff but does NOT cause early return.
    const maxLen = Math.max(bufA.length, bufB.length);
    let diff = bufA.length ^ bufB.length; // length mismatch is a difference
    for (let i = 0; i < maxLen; i++) {
        diff |= (bufA[i] ?? 0) ^ (bufB[i] ?? 0);
    }
    return diff === 0;
}
