// @mcpfusion/core/dist/introspection/GovernanceObserver.js
import { SpanStatusCode } from '../observability/Tracing.js';
import { toErrorMessage } from '../core/ErrorUtils.js';
/**
 * Create a governance observer that emits debug events and/or tracing
 * spans for governance operations.
 *
 * @param config - Observer configuration (debug handler and/or tracer)
 * @returns A `GovernanceObserver` instance
 *
 * @example
 * ```typescript
 * import { createGovernanceObserver } from '@mcpfusion/core/introspection';
 * import { createDebugObserver } from '@mcpfusion/core';
 *
 * const observer = createGovernanceObserver({
 *     debug: createDebugObserver(),
 * });
 *
 * const contracts = observer.observe(
 *     'contract.compile',
 *     'Compiling 5 tool contracts',
 *     () => compileContracts(builders),
 * );
 * ```
 */
export function createGovernanceObserver(config) {
    const { debug, tracer } = config;
    /**
     * Shared span + event logic for both sync and async paths.
     * Eliminates duplication between observe() and observeAsync().
     * @internal
     */
    function finalize(operation, label, start, span, resultOrError) {
        const durationMs = Date.now() - start;
        if (resultOrError.ok) {
            span?.setAttribute('mcp.governance.outcome', 'success');
            span?.setAttribute('mcp.durationMs', durationMs);
            span?.setStatus({ code: SpanStatusCode.OK });
            debug?.({
                type: 'governance',
                operation,
                label,
                outcome: 'success',
                durationMs,
                timestamp: Date.now(),
            });
            span?.end();
            return resultOrError.value;
        }
        const message = toErrorMessage(resultOrError.error);
        span?.setAttribute('mcp.governance.outcome', 'failure');
        span?.setAttribute('mcp.durationMs', durationMs);
        span?.setStatus({ code: SpanStatusCode.ERROR, message });
        span?.recordException(resultOrError.error instanceof Error ? resultOrError.error : new Error(message));
        debug?.({
            type: 'governance',
            operation,
            label,
            outcome: 'failure',
            detail: message,
            durationMs,
            timestamp: Date.now(),
        });
        span?.end();
        throw resultOrError.error;
    }
    function startSpan(operation, label) {
        return tracer?.startSpan(`mcp.governance.${operation}`, {
            attributes: {
                'mcp.governance.operation': operation,
                'mcp.governance.label': label,
            },
        });
    }
    function observe(operation, label, fn) {
        const start = Date.now();
        const span = startSpan(operation, label);
        try {
            const result = fn();
            // Runtime guard — reject async callbacks passed to sync observe()
            if (result != null && typeof result['then'] === 'function') {
                throw new Error('[mcpfusion] observe() received an async callback. Use observeAsync() for async operations.');
            }
            return finalize(operation, label, start, span, { ok: true, value: result });
        }
        catch (err) {
            return finalize(operation, label, start, span, { ok: false, error: err });
        }
    }
    async function observeAsync(operation, label, fn) {
        const start = Date.now();
        const span = startSpan(operation, label);
        try {
            const result = await fn();
            return finalize(operation, label, start, span, { ok: true, value: result });
        }
        catch (err) {
            return finalize(operation, label, start, span, { ok: false, error: err });
        }
    }
    return { observe, observeAsync };
}
/**
 * Create a no-op governance observer.
 *
 * Used when observability is not configured. Zero overhead.
 */
export function createNoopObserver() {
    return {
        observe: (_op, _label, fn) => fn(),
        observeAsync: (_op, _label, fn) => fn(),
    };
}
