// @mcpfusion/core/dist/introspection/index.js
/** Introspection Module — Barrel Export */
export { compileManifest, cloneManifest } from './ManifestCompiler.js';
export { registerIntrospectionResource } from './IntrospectionResource.js';
// ── Contract Materialization ─────────────────────────────
export { materializeContract, compileContracts, sha256, canonicalize, } from './ToolContract.js';
// ── Contract Diffing ─────────────────────────────────────
export { diffContracts, formatDiffReport, formatDeltasAsXml, } from './ContractDiff.js';
// ── Behavioral Fingerprinting ────────────────────────────
export { computeDigest, computeServerDigest, compareServerDigests, } from './BehaviorDigest.js';
// ── Capability Lockfile ──────────────────────────────────
export { generateLockfile, serializeLockfile, checkLockfile, parseLockfile, writeLockfile, readLockfile, LOCKFILE_NAME, } from './CapabilityLockfile.js';
// ── Zero-Trust Runtime ───────────────────────────────────
export { createHmacSigner, attestServerDigest, verifyAttestation, verifyCapabilityPin, buildTrustCapability, AttestationError, } from './CryptoAttestation.js';
// ── Token Economics ──────────────────────────────────────
export { estimateTokens, profileBlock, profileResponse, computeStaticProfile, aggregateProfiles, } from './TokenEconomics.js';
// ── Entitlement Scanner (Blast Radius) ───────────────────
export { scanSource, buildEntitlements, validateClaims, scanAndValidate, scanEvasionIndicators, } from './EntitlementScanner.js';
// ── Semantic Probing (LLM-as-Judge) ──────────────────────
export { createProbe, buildJudgePrompt, parseJudgeResponse, evaluateProbe, evaluateProbes, aggregateResults, } from './SemanticProbe.js';
// ── Self-Healing Context ─────────────────────────────────
export { enrichValidationError, createToolEnhancer, } from './ContractAwareSelfHealing.js';
// ── Governance Observer (Observability Bridge) ───────────
export { createGovernanceObserver, createNoopObserver, } from './GovernanceObserver.js';
// ── Server Card (SEP-1649 Auto-Discovery) ────────────────
export { compileServerCard, SERVER_CARD_PATH } from './ServerCard.js';
