// @mcpfusion/core/dist/introspection/IntrospectionResource.js
/**
 * IntrospectionResource — MCP Resource Registration for Dynamic Manifest
 *
 * When introspection is enabled, this module silently registers two
 * MCP request handlers on the low-level Server:
 *
 *   1. `resources/list` — Advertises the manifest resource
 *   2. `resources/read` — Returns the RBAC-filtered manifest JSON
 *
 * Uses the native MCP Resource protocol — no custom HTTP endpoints.
 * Zero overhead when introspection is disabled (this module is never imported).
 *
 * @module
 */
import { ListResourcesRequestSchema, ReadResourceRequestSchema, } from '@modelcontextprotocol/sdk/types.js';
import {} from '../core/types.js';
import {} from './types.js';
import { compileManifest, cloneManifest } from './ManifestCompiler.js';
// ── Types ────────────────────────────────────────────────
/** Default URI for the introspection resource */
const DEFAULT_URI = 'mcpfusion://manifest.json';
/** Default resource name */
const RESOURCE_NAME = 'mcpfusion:manifest';
/** Default resource description */
const RESOURCE_DESCRIPTION = 'Dynamic introspection manifest exposing all registered tools, actions, ' +
    'and presenters. RBAC-filtered per session context.';
/** Default MIME type for the manifest resource */
const RESOURCE_MIME_TYPE = 'application/json';
// ── Public API ───────────────────────────────────────────
/**
 * Register MCP Resource handlers for introspection on the low-level Server.
 *
 * Called by `attachToServer()` when `options.introspection.enabled` is `true`.
 *
 * @param server - The resolved low-level MCP Server instance
 * @param config - Introspection configuration (from AttachOptions)
 * @param serverName - Server name for the manifest payload
 * @param builders - Registry delegate providing builder iteration
 * @param contextFactory - Factory to create per-request context (from AttachOptions)
 */
export function registerIntrospectionResource(server, config, serverName, builders, contextFactory) {
    const resolved = server;
    const uri = config.uri ?? DEFAULT_URI;
    // ── resources/list handler ────────────────────────────
    // Advertises the manifest resource to clients/orchestrators
    const existingListHandler = safeGetExistingListHandler(resolved);
    resolved.setRequestHandler(ListResourcesRequestSchema, (() => {
        // Merge with any previously registered resources
        const existingResources = existingListHandler ? existingListHandler() : [];
        return {
            resources: [
                ...existingResources,
                {
                    uri,
                    name: RESOURCE_NAME,
                    description: RESOURCE_DESCRIPTION,
                    mimeType: RESOURCE_MIME_TYPE,
                },
            ],
        };
    }));
    // ── resources/read handler ───────────────────────────
    // Returns the RBAC-filtered manifest when requested
    resolved.setRequestHandler(ReadResourceRequestSchema, (async (request, extra) => {
        const requestedUri = request.params.uri;
        // Only handle our manifest URI
        if (requestedUri !== uri) {
            return {
                contents: [],
            };
        }
        // Compile the full manifest from current registry state
        const fullManifest = compileManifest(serverName, builders.values());
        // Apply RBAC filter if configured
        let manifest;
        if (config.filter && contextFactory) {
            const ctx = await contextFactory(extra);
            // Clone before mutation to protect the compiled tree
            manifest = config.filter(cloneManifest(fullManifest), ctx);
        }
        else {
            manifest = fullManifest;
        }
        return {
            contents: [
                {
                    uri,
                    mimeType: RESOURCE_MIME_TYPE,
                    text: JSON.stringify(manifest, null, 2),
                },
            ],
        };
    }));
}
// ── Private Helpers ──────────────────────────────────────
/**
 * Attempt to preserve an existing resources/list handler.
 *
 * Since the low-level Server replaces handlers (not appends), we
 * capture any existing handler before overwriting it. This allows
 * introspection to coexist with user-registered resources.
 *
 * @returns A function returning existing resources, or undefined
 */
function safeGetExistingListHandler(_server) {
    // The low-level MCP SDK doesn't expose a getter for existing handlers.
    // For now, introspection assumes it's the first to register resources/list.
    // If the user also registers resources, they should do so AFTER
    // attachToServer() or use the high-level McpServer.resource() API.
    return undefined;
}
