// @mcpfusion/core/dist/introspection/ServerCard.js
/**
 * ServerCard — MCP Server Card Compiler (SEP-1649)
 *
 * Compiles a standard `/.well-known/mcp/server-card.json` payload from
 * the MCP Fusion registry metadata for auto-discovery by MCP clients
 * (Claude, Cursor, ChatGPT, etc.).
 *
 * The Server Card enables AI agents to automatically discover and
 * inspect a server's capabilities, authentication requirements,
 * and available primitives before establishing a connection.
 *
 * Pure-function module: no state, no side effects.
 *
 * @module
 */
import {} from './types.js';
// ── Constants ────────────────────────────────────────────
/** Default schema URL for the MCP Server Card format */
const DEFAULT_SCHEMA = 'https://modelcontextprotocol.io/schemas/server-card/v1';
/** Default server card schema version */
const DEFAULT_CARD_VERSION = '1.0';
/** Default MCP protocol version (2025-06-18 spec) */
const DEFAULT_PROTOCOL_VERSION = '2025-06-18';
/** Well-known path for the server card endpoint */
export const SERVER_CARD_PATH = '/.well-known/mcp/server-card.json';
// ── Public API ───────────────────────────────────────────
/**
 * Compile a Server Card payload from MCP Fusion registry metadata.
 *
 * Extracts tool names, descriptions, prompt definitions, resource
 * URIs, and auth requirements into the standard Server Card format.
 *
 * @param config - Server card configuration
 * @param builders - Iterable of all registered tool builders
 * @param prompts - Optional list of prompt definitions
 * @param resources - Optional list of resource definitions
 * @returns A JSON-serializable Server Card payload
 *
 * @example
 * ```typescript
 * const card = compileServerCard(
 *     { name: 'billing-api', version: '1.0.0', transport: 'http' },
 *     registry.getBuilders(),
 * );
 * // Serve at /.well-known/mcp/server-card.json
 * ```
 */
export function compileServerCard(config, builders, prompts, resources) {
    // ── Tools ─────────────────────────────────────────────
    const tools = [];
    for (const builder of builders) {
        const def = builder.buildToolDefinition();
        tools.push({
            name: builder.getName(),
            description: def?.description,
            tags: builder.getTags ? builder.getTags() : [],
        });
    }
    // ── Prompts ──────────────────────────────────────────
    const promptEntries = [];
    if (prompts) {
        for (const p of prompts) {
            promptEntries.push({ name: p.name, ...(p.description ? { description: p.description } : {}) });
        }
    }
    // ── Resources ────────────────────────────────────────
    const resourceEntries = [];
    if (resources) {
        for (const r of resources) {
            resourceEntries.push({
                uri: r.uri,
                name: r.name,
                ...(r.description ? { description: r.description } : {}),
                ...(r.mimeType ? { mimeType: r.mimeType } : {}),
            });
        }
    }
    // ── Transport ────────────────────────────────────────
    const transport = config.transport === 'stdio'
        ? { type: 'stdio' }
        : {
            type: 'streamable-http',
            endpoint: config.endpoint ?? '/mcp',
        };
    // ── Capabilities ─────────────────────────────────────
    const capabilities = {};
    if (tools.length > 0)
        capabilities.tools = {};
    if (promptEntries.length > 0)
        capabilities.prompts = {};
    if (resourceEntries.length > 0)
        capabilities.resources = {};
    // ── Payload ──────────────────────────────────────────
    const payload = {
        $schema: DEFAULT_SCHEMA,
        version: DEFAULT_CARD_VERSION,
        protocolVersion: config.protocolVersion ?? DEFAULT_PROTOCOL_VERSION,
        serverInfo: {
            name: config.name,
            version: config.version ?? '1.0.0',
            ...(config.title ? { title: config.title } : {}),
        },
        ...(config.description ? { description: config.description } : {}),
        ...(config.iconUrl ? { iconUrl: config.iconUrl } : {}),
        ...(config.documentationUrl ? { documentationUrl: config.documentationUrl } : {}),
        transport,
        capabilities,
        ...(tools.length > 0 ? { tools } : {}),
        ...(promptEntries.length > 0 ? { prompts: promptEntries } : {}),
        ...(resourceEntries.length > 0 ? { resources: resourceEntries } : {}),
    };
    return payload;
}
