// @mcpfusion/core/dist/introspection/ToolContract.js
/**
 * ToolContract — Behavioral Contract Materialization
 *
 * Unifies the declarative surface (tool names, schemas, tags) with
 * the behavioral contract (Presenter egress schema, system rules,
 * cognitive guardrails, middleware chain, state-sync policies) into
 * a single, serializable, diffable, hashable artifact.
 *
 * **Key insight**: MCP Fusion's Presenter is not just a response
 * formatter — it's a declarative behavioral specification. The Zod
 * schema, system rules, agent limits, and affordances are all
 * explicit, serializable contracts. By materializing them into a
 * `ToolContract`, we create a fingerprint that changes when behavior
 * changes — even if the MCP tool declaration stays identical.
 *
 * **Zero developer effort**: The contract materializes from what the
 * developer has already declared. No annotations, no config, no
 * ceremony — just `materializeContract(builder)`.
 *
 * Pure-function module: no state, no side effects.
 *
 * @module
 */
import {} from '../core/types.js';
import { sha256, canonicalize } from './canonicalize.js';
import { scanSource, buildEntitlements } from './EntitlementScanner.js';
// ============================================================================
// Materialization
// ============================================================================
/**
 * Materialize a `ToolContract` from a builder's public API.
 *
 * Extracts all behavioral metadata from the builder's introspection
 * methods without accessing any private internals. This guarantees
 * compatibility with any `ToolBuilder` implementation.
 *
 * @param builder - A registered tool builder
 * @returns A fully materialized `ToolContract`
 */
export async function materializeContract(builder) {
    const toolDef = builder.buildToolDefinition();
    const metadata = builder.getActionMetadata();
    // Surface
    const actions = {};
    for (const action of metadata) {
        actions[action.key] = {
            description: action.description,
            destructive: action.destructive,
            idempotent: action.idempotent,
            readOnly: action.readOnly,
            requiredFields: action.requiredFields,
            presenterName: action.presenterName,
            inputSchemaDigest: await sha256(JSON.stringify(action.requiredFields)),
            hasMiddleware: action.hasMiddleware,
        };
    }
    const surface = {
        name: builder.getName(),
        description: toolDef.description,
        tags: builder.getTags(),
        actions,
        inputSchemaDigest: await sha256(canonicalize(toolDef.inputSchema)),
    };
    // Behavior — extract from action metadata (Presenter introspection accessors)
    const behavior = await materializeBehavior(metadata, builder);
    // Token Economics
    const tokenEconomics = computeTokenEconomics(metadata, behavior);
    // Entitlements — real static analysis via EntitlementScanner
    let entitlements;
    try {
        const builderAny = builder;
        const actions = typeof builderAny['getActions'] === 'function'
            ? builderAny.getActions()
            : [];
        const combinedSource = actions
            .filter(a => typeof a.handler === 'function')
            .map(a => a.handler.toString())
            .join('\n');
        if (combinedSource.length > 0) {
            const matches = scanSource(combinedSource);
            entitlements = buildEntitlements(matches);
        }
        else {
            entitlements = { filesystem: false, network: false, subprocess: false, crypto: false, codeEvaluation: false, raw: [] };
        }
    }
    catch {
        /* static analysis unavailable — default to no entitlements */
        entitlements = { filesystem: false, network: false, subprocess: false, crypto: false, codeEvaluation: false, raw: [] };
    }
    return { surface, behavior, tokenEconomics, entitlements };
}
/**
 * Materialize the behavioral contract from action metadata.
 * @internal
 */
async function materializeBehavior(metadata, _builder) {
    // Collect Presenter data from the first action that declares one
    let egressSchemaDigest = null;
    let systemRulesFingerprint = 'none';
    const affordanceTopology = [];
    const embeddedPresenters = [];
    // Collect unique Presenter metadata across all actions
    const presenterSchemaKeys = new Set();
    const staticRuleStrings = [];
    for (const action of metadata) {
        if (action.presenterSchemaKeys) {
            for (const key of action.presenterSchemaKeys) {
                presenterSchemaKeys.add(key);
            }
        }
        if (action.presenterStaticRules) {
            for (const rule of action.presenterStaticRules) {
                staticRuleStrings.push(rule);
            }
        }
        if (action.presenterHasContextualRules) {
            systemRulesFingerprint = 'dynamic';
        }
    }
    if (presenterSchemaKeys.size > 0) {
        const sortedKeys = [...presenterSchemaKeys].sort();
        egressSchemaDigest = await sha256(sortedKeys.join(','));
    }
    // condition should only check staticRuleStrings, not presenterSchemaKeys
    // always hash static rules even when dynamic rules coexist
    if (staticRuleStrings.length > 0) {
        const sortedRules = [...new Set(staticRuleStrings)].sort();
        const hash = await sha256(sortedRules.join(','));
        // Composite fingerprint preserves static rule coverage alongside dynamic indicator
        const prefix = systemRulesFingerprint === 'dynamic' ? 'dynamic' : 'static';
        systemRulesFingerprint = prefix + ':' + hash;
    }
    // Middleware chain — extract names from builder
    const middlewareChain = [];
    // Cannot access private _middlewares, but we can infer from action metadata
    for (const action of metadata) {
        if (action.hasMiddleware) {
            middlewareChain.push(`${action.key}:mw`);
        }
    }
    return {
        egressSchemaDigest,
        systemRulesFingerprint,
        cognitiveGuardrails: {
            agentLimitMax: null, // Extracted per-Presenter in extended materialization
            egressMaxBytes: null,
        },
        middlewareChain,
        stateSyncFingerprint: null, // Injected by StateSyncLayer integration
        concurrencyFingerprint: null,
        affordanceTopology,
        embeddedPresenters,
    };
}
/**
 * Compute token economics profile from contract metadata.
 * @internal
 */
function computeTokenEconomics(metadata, behavior) {
    // Count schema fields across all Presenters
    let schemaFieldCount = 0;
    const seenPresenters = new Set();
    for (const action of metadata) {
        if (action.presenterName && !seenPresenters.has(action.presenterName)) {
            seenPresenters.add(action.presenterName);
            schemaFieldCount += action.presenterSchemaKeys?.length ?? 0;
        }
    }
    // Unbounded collection risk
    const unboundedCollection = behavior.cognitiveGuardrails.agentLimitMax === null
        && schemaFieldCount > 0;
    // Base overhead estimation (~4 chars per token, GPT heuristic)
    // Rules: ~50 tokens each, UI blocks: ~100 tokens each
    const rulesOverhead = behavior.systemRulesFingerprint !== 'none' ? 50 : 0;
    const affordanceOverhead = behavior.affordanceTopology.length * 20;
    const baseOverheadTokens = rulesOverhead + affordanceOverhead;
    // Inflation risk — derived from guardrails and schema density
    const inflationRisk = classifyInflationRisk(unboundedCollection, schemaFieldCount);
    return {
        schemaFieldCount,
        unboundedCollection,
        baseOverheadTokens,
        inflationRisk,
    };
}
// ============================================================================
// Server-Level Contract Compilation
// ============================================================================
/**
 * Compile contracts for all tools in a registry.
 *
 * @param builders - Iterable of all registered tool builders
 * @returns Record mapping tool names to their materialized contracts
 */
export async function compileContracts(builders) {
    const contracts = {};
    for (const builder of builders) {
        const contract = await materializeContract(builder);
        contracts[contract.surface.name] = contract;
    }
    return contracts;
}
// ============================================================================
// Internals
// ============================================================================
/** @internal */
function classifyInflationRisk(unbounded, fieldCount) {
    if (unbounded && fieldCount > 10)
        return 'critical';
    if (unbounded)
        return 'high';
    if (fieldCount > 20)
        return 'medium';
    return 'low';
}
// ── Re-export shared utilities for public API stability ──
export { sha256, canonicalize } from './canonicalize.js';
