// @mcpfusion/core/dist/presenter/definePresenter.js
/**
 * definePresenter() — Declarative Presenter Definition
 *
 * Zero-friction object-config API for creating Presenters.
 * Replaces the fluent builder pattern with a single object literal
 * that enables instant Ctrl+Space autocomplete and zero generic noise.
 *
 * The `schema` field drives type inference: the `ui`, `rules`, and
 * `suggestActions` callbacks automatically receive the inferred type,
 * so the developer never writes a generic parameter manually.
 *
 * @example
 * ```typescript
 * import { definePresenter, ui } from '@mcpfusion/core';
 * import { z } from 'zod';
 *
 * export const InvoicePresenter = definePresenter({
 *   name: 'Invoice',
 *   schema: z.object({
 *     id: z.string(),
 *     amount_cents: z.number().describe('CRITICAL: in CENTS. Divide by 100.'),
 *     status: z.enum(['paid', 'pending']),
 *   }),
 *   rules: ['CRITICAL: Divide amount_cents by 100 before displaying.'],
 *   ui: (inv) => [ui.echarts({ series: [{ type: 'gauge', data: [{ value: inv.amount_cents / 100 }] }] })],
 *   suggestActions: (inv) =>
 *     inv.status === 'pending'
 *       ? [{ tool: 'billing.pay', reason: 'Offer immediate payment' }]
 *       : [],
 * });
 * ```
 *
 * @module
 */
import {} from 'zod';
import { Presenter } from './Presenter.js';
import {} from './ui.js';
import { extractZodDescriptions } from './ZodDescriptionExtractor.js';
/**
 * Implementation
 * @internal
 */
export function definePresenter(config) {
    const presenter = new Presenter(config.name);
    // Resolve the Zod schema (supports both raw ZodType and Model objects)
    let resolvedSchema;
    if (config.schema) {
        // Support passing a Model object directly (duck-type detection)
        // eslint-disable-next-line @typescript-eslint/no-explicit-any -- duck-type detection for Model vs raw ZodType
        const schemaVal = config.schema;
        if (schemaVal != null && typeof schemaVal === 'object' && 'fields' in schemaVal && 'schema' in schemaVal) {
            resolvedSchema = schemaVal.schema;
        }
        else {
            resolvedSchema = config.schema;
        }
        presenter.schema(resolvedSchema);
    }
    // ── Zod-Driven Prompts: auto-extract .describe() annotations ──
    const autoRules = config.autoRules !== false; // default: true
    const zodDescriptions = (autoRules && resolvedSchema)
        ? extractZodDescriptions(resolvedSchema)
        : [];
    if (config.rules) {
        if (typeof config.rules === 'function') {
            // Dynamic rules — wrap to prepend Zod descriptions
            if (zodDescriptions.length > 0) {
                const dynamicFn = config.rules;
                presenter.systemRules((data, ctx) => [
                    ...zodDescriptions,
                    ...dynamicFn(data, ctx),
                ]);
            }
            else {
                presenter.systemRules(config.rules);
            }
        }
        else {
            // Static rules — merge with Zod descriptions
            const merged = [...zodDescriptions, ...config.rules];
            presenter.systemRules(merged);
        }
    }
    else if (zodDescriptions.length > 0) {
        // No explicit rules — use Zod descriptions alone
        presenter.systemRules(zodDescriptions);
    }
    if (config.ui) {
        presenter.uiBlocks(config.ui);
    }
    if (config.collectionUi) {
        presenter.collectionUiBlocks(config.collectionUi);
    }
    if (config.agentLimit) {
        presenter.agentLimit(config.agentLimit.max, config.agentLimit.onTruncate);
    }
    if (config.suggestActions) {
        presenter.suggestActions(config.suggestActions);
    }
    if (config.embeds) {
        for (const embed of config.embeds) {
            presenter.embed(embed.key, embed.presenter);
        }
    }
    if (config.redactPII) {
        presenter.redactPII(config.redactPII.paths, config.redactPII.censor);
    }
    if (config.collectionSuggestions) {
        presenter.collectionSuggestActions(config.collectionSuggestions);
    }
    if (config.collectionRules) {
        presenter.collectionRules(config.collectionRules);
    }
    // Async callbacks
    if (config.asyncUi) {
        presenter.asyncUiBlocks(config.asyncUi);
    }
    if (config.asyncCollectionUi) {
        presenter.asyncCollectionUiBlocks(config.asyncCollectionUi);
    }
    if (config.asyncRules) {
        presenter.asyncRules(config.asyncRules);
    }
    if (config.asyncSuggestActions) {
        presenter.asyncSuggestActions(config.asyncSuggestActions);
    }
    return presenter;
}
/**
 * Implementation
 * @internal
 */
export function extendPresenter(base, overrides) {
    const merged = {
        // Name and schema: override always wins
        name: overrides.name,
        // Auto-rules: override wins when explicitly set
        autoRules: overrides.autoRules ?? base.autoRules,
        // Rules: merge strategy
        rules: _mergeRules(base.rules, overrides.rules),
        // UI: override wins if defined
        ui: overrides.ui ?? base.ui,
        collectionUi: overrides.collectionUi ?? base.collectionUi,
        // Agent limit: override wins if defined
        agentLimit: overrides.agentLimit ?? base.agentLimit,
        // Suggestions: override wins if defined
        suggestActions: overrides.suggestActions ?? base.suggestActions,
        collectionSuggestions: overrides.collectionSuggestions ?? base.collectionSuggestions,
        // Collection rules: override wins if defined
        collectionRules: overrides.collectionRules ?? base.collectionRules,
        // Embeds: merge (additive)
        embeds: [
            ...(base.embeds ?? []),
            ...(overrides.embeds ?? []),
        ],
        // Redaction: merge paths
        redactPII: _mergeRedactPII(base.redactPII, overrides.redactPII),
        // Async callbacks: override wins if defined
        asyncUi: overrides.asyncUi ?? base.asyncUi,
        asyncCollectionUi: overrides.asyncCollectionUi ?? base.asyncCollectionUi,
        asyncRules: overrides.asyncRules ?? base.asyncRules,
        asyncSuggestActions: overrides.asyncSuggestActions ?? base.asyncSuggestActions,
    };
    // Schema: override wins, set only if defined (exactOptionalPropertyTypes)
    const schema = overrides.schema ?? base.schema;
    if (schema != null)
        merged['schema'] = schema;
    // Remove undefined values to satisfy exactOptionalPropertyTypes
    for (const key of Object.keys(merged)) {
        if (merged[key] === undefined)
            delete merged[key];
    }
    return definePresenter(merged);
}
/**
 * Merge two rules configs: static arrays are concatenated, dynamic functions are chained.
 * @internal
 */
function _mergeRules(base, override) {
    if (!base)
        return override;
    if (!override)
        return base;
    const baseIsStatic = typeof base !== 'function';
    const overrideIsStatic = typeof override !== 'function';
    // Both static: concatenate
    if (baseIsStatic && overrideIsStatic) {
        return [...base, ...override];
    }
    // Mix or both dynamic: chain
    const baseFn = baseIsStatic
        ? () => [...base]
        : base;
    const overrideFn = overrideIsStatic
        ? () => [...override]
        : override;
    return (data, ctx) => [
        ...baseFn(data, ctx),
        ...overrideFn(data, ctx),
    ];
}
/**
 * Merge two redactPII configurations by concatenating paths.
 * The override's censor function takes priority.
 * @internal
 */
function _mergeRedactPII(base, override) {
    if (!base)
        return override;
    if (!override)
        return base;
    const result = {
        paths: [...base.paths, ...override.paths],
    };
    const censor = override.censor ?? base.censor;
    if (censor !== undefined)
        result.censor = censor;
    return result;
}
