// @mcpfusion/core/dist/presenter/PresenterPipeline.js
/**
 * PresenterPipeline — Decomposed Make Pipeline
 *
 * Extracted from `Presenter.make()` to uphold SRP and enable isolated testing.
 * Each step is a pure function that receives a `PresenterSnapshot` (read-only
 * config) and produces side-effects on a `ResponseBuilder`.
 *
 * The `executePipeline()` orchestrator replaces the inline logic previously
 * in `Presenter.make()`, maintaining the exact same behavior and output.
 *
 * @internal
 * @module
 */
import { ZodError } from 'zod';
import { ResponseBuilder } from './ResponseBuilder.js';
import {} from './ui.js';
import {} from './Presenter.js';
import { PresenterValidationError } from './PresenterValidationError.js';
import { applySelectFilter } from './SelectUtils.js';
import { compileRedactor } from './RedactEngine.js';
import {} from '../core/serialization/JsonSerializer.js';
import {} from './Presenter.js';
// ── Step Functions ──────────────────────────────────────
/**
 * Step 0: Cognitive Guardrails — truncate array by agentLimit.
 *
 * If the data is an array larger than the configured limit, it is sliced
 * and a truncation UiBlock is returned for insertion.
 *
 * @returns Potentially truncated data + optional truncation UiBlock
 * @internal
 */
export function stepTruncate(data, isArray, snapshot) {
    if (isArray && snapshot.agentLimit && data.length > snapshot.agentLimit.max) {
        const fullLength = data.length;
        const omitted = fullLength - snapshot.agentLimit.max;
        return {
            data: data.slice(0, snapshot.agentLimit.max),
            truncationBlock: snapshot.agentLimit.onTruncate(omitted),
        };
    }
    return { data };
}
/**
 * Step 1: Validate data through the Zod schema (if configured).
 * For arrays, each item is validated independently.
 *
 * @throws {PresenterValidationError} on schema mismatch
 * @internal
 */
export function stepValidate(data, isArray, snapshot) {
    if (!snapshot.schema)
        return data;
    try {
        if (isArray) {
            return data.map(item => snapshot.schema.parse(item));
        }
        return snapshot.schema.parse(data);
    }
    catch (err) {
        if (err instanceof ZodError) {
            throw new PresenterValidationError(snapshot.name, err);
        }
        throw err;
    }
}
/**
 * Step 2: Apply PII redaction to wire-facing data.
 * Creates a deep clone to preserve original data for UI/rules.
 *
 * @internal
 */
export function stepRedact(data, isArray, snapshot) {
    // Lazy recompilation: if redactPII was called before fast-redact loaded
    if (!snapshot.compiledRedactor && snapshot.redactConfig) {
        snapshot.compiledRedactor = compileRedactor(snapshot.redactConfig);
        // Propagate back to Presenter so subsequent make() calls reuse it
        if (snapshot.compiledRedactor && snapshot.writeBackRedactor) {
            snapshot.writeBackRedactor(snapshot.compiledRedactor);
        }
        if (!snapshot.compiledRedactor) {
            console.warn(`[mcpfusion] Presenter "${snapshot.name}": PII redaction configured but fast-redact is not available. ` +
                `Data will pass through WITHOUT redaction. Ensure initMCPFusion() completes before .make() is called, ` +
                `or install fast-redact as a dependency.`);
        }
    }
    if (!snapshot.compiledRedactor)
        return data;
    if (isArray) {
        return data.map(item => snapshot.compiledRedactor(item));
    }
    return snapshot.compiledRedactor(data);
}
/**
 * Step 3: Generate and attach UI blocks to the response builder.
 * Auto-detects single vs collection. Filters `null` blocks.
 *
 * @internal
 */
export function stepUiBlocks(builder, data, isArray, snapshot, ctx) {
    if (isArray && snapshot.collectionUiBlocks) {
        const blocks = snapshot.collectionUiBlocks(data, ctx).filter(Boolean);
        if (blocks.length > 0)
            builder.uiBlocks(blocks);
    }
    else if (!isArray && snapshot.itemUiBlocks) {
        const blocks = snapshot.itemUiBlocks(data, ctx).filter(Boolean);
        if (blocks.length > 0)
            builder.uiBlocks(blocks);
    }
}
/**
 * Step 4: Resolve and attach domain rules to the response builder.
 * Supports both static arrays and dynamic context-aware functions.
 *
 * For collections: also evaluates `collectionRules` (if defined) and
 * merges them after per-item rules.
 *
 * @internal
 */
export function stepRules(builder, data, isArray, snapshot, ctx) {
    // Per-item rules
    if (typeof snapshot.rules === 'function') {
        const singleData = isArray ? data[0] : data;
        if (singleData !== undefined) {
            const resolved = snapshot.rules(singleData, ctx)
                .filter((r) => r !== null);
            if (resolved.length > 0)
                builder.systemRules(resolved);
        }
    }
    else if (snapshot.rules.length > 0) {
        builder.systemRules(snapshot.rules);
    }
    // Collection-level rules (additive — appended after per-item rules)
    const hasCollectionRules = typeof snapshot.collectionRules === 'function'
        || snapshot.collectionRules.length > 0;
    if (isArray && hasCollectionRules) {
        const items = data;
        if (typeof snapshot.collectionRules === 'function') {
            const resolved = snapshot.collectionRules(items, ctx)
                .filter((r) => r !== null);
            if (resolved.length > 0)
                builder.systemRules(resolved);
        }
        else if (snapshot.collectionRules.length > 0) {
            builder.systemRules(snapshot.collectionRules);
        }
    }
}
/**
 * Step 5: Evaluate and attach action suggestions to the response.
 *
 * For collections: uses `collectionSuggestActions` if defined,
 * otherwise falls back to per-item evaluation on the first item.
 *
 * @internal
 */
export function stepSuggestions(builder, data, isArray, snapshot, ctx) {
    // Collection-level suggestions (preferred for arrays)
    if (isArray && snapshot.collectionSuggestActions) {
        const items = data;
        if (items.length === 0)
            return;
        const suggestions = snapshot.collectionSuggestActions(items, ctx)
            .filter((s) => s !== null);
        if (suggestions.length > 0) {
            builder.systemHint(suggestions);
        }
        return;
    }
    // Per-item fallback (single item or first item of array)
    if (!snapshot.suggestActions)
        return;
    const singleData = isArray ? data[0] : data;
    if (singleData === undefined)
        return;
    const suggestions = snapshot.suggestActions(singleData, ctx);
    if (suggestions.length > 0) {
        builder.systemHint(suggestions);
    }
}
/**
 * Step 6: Process embedded child Presenters for nested relational data.
 * Merges child UI blocks and rules into the parent builder.
 *
 * For collections: iterates all array items. Rules are deduplicated
 * via a Set to avoid repetition. UI blocks are emitted only for the
 * first item to prevent context window explosion.
 *
 * @internal
 */
export function stepEmbeds(builder, data, isArray, snapshot, ctx) {
    if (snapshot.embeds.length === 0)
        return;
    const items = isArray
        ? data.filter((item) => item !== undefined && typeof item === 'object')
        : [data].filter((item) => item !== undefined && typeof item === 'object');
    if (items.length === 0)
        return;
    // Track which rule blocks we've already emitted (dedup for arrays)
    const emittedRules = new Set();
    for (let itemIdx = 0; itemIdx < items.length; itemIdx++) {
        const item = items[itemIdx];
        for (const embed of snapshot.embeds) {
            const nestedData = item[embed.key];
            if (nestedData === undefined || nestedData === null)
                continue;
            const childBuilder = embed.presenter.make(nestedData, ctx);
            const childResponse = childBuilder.build();
            // Skip the first block (data) — parent already has it
            for (let i = 1; i < childResponse.content.length; i++) {
                const blockText = childResponse.content[i].text;
                // For arrays: deduplicate rule/hint blocks, emit UI only for first item
                if (isArray && itemIdx > 0) {
                    if (emittedRules.has(blockText))
                        continue;
                    if (blockText.includes('<ui_passthrough'))
                        continue;
                }
                emittedRules.add(blockText);
                builder.rawBlock(blockText);
            }
        }
    }
}
/**
 * Execute the full Presenter pipeline.
 *
 * Orchestrates: truncate → validate → embed → render UI → attach rules
 * → suggest actions → **Late Guillotine** (`_select` filter).
 *
 * @param data - Raw data from handler
 * @param snapshot - Read-only configuration snapshot
 * @param ctx - Optional request context
 * @param selectFields - Optional field names for context window optimization
 * @returns A fully-composed ResponseBuilder
 * @internal
 */
export function executePipeline(data, snapshot, ctx, selectFields) {
    return executePipelineWithData(data, snapshot, ctx, selectFields).builder;
}
/**
 * Execute the full Presenter pipeline and return the validated data.
 *
 * Extended version of `executePipeline()` that also returns the
 * truncated+validated data for reuse by `makeAsync()`, ensuring
 * each Zod schema runs exactly once per call — critical for
 * `.transform()` schemas with non-idempotent logic.
 *
 * @param data - Raw data from handler
 * @param snapshot - Read-only configuration snapshot
 * @param ctx - Optional request context
 * @param selectFields - Optional field names for context window optimization
 * @returns A {@link PipelineResult} with both builder and validated data
 * @internal
 */
export function executePipelineWithData(data, snapshot, ctx, selectFields) {
    const isArray = Array.isArray(data);
    // Step 0: Cognitive Guardrails — truncate if needed
    const truncated = stepTruncate(data, isArray, snapshot);
    data = truncated.data;
    // Step 1: Process embedded child Presenters (on RAW data, before validation)
    const rawForEmbeds = data;
    // Step 2: Validate — produces the FULL validated data
    const validated = stepValidate(data, isArray, snapshot);
    // ── Late Guillotine ──────────────────────────────────
    // Steps 3-6 use the FULL validated data so that UI blocks,
    // system rules, and action suggestions never see undefined
    // for pruned fields.
    // Step 3: Determine wire-facing data (filtered or full)
    const wireData = (selectFields && selectFields.length > 0)
        ? applySelectFilter(validated, selectFields, isArray)
        : validated;
    // Step 3.1: DLP Redaction — mask PII on the wire-facing data
    const safeWireData = stepRedact(wireData, isArray, snapshot);
    const builder = new ResponseBuilder(safeWireData, snapshot.compiledStringify);
    // Step 3.5: Truncation warning (first UI block, before all others)
    if (truncated.truncationBlock) {
        builder.uiBlock(truncated.truncationBlock);
    }
    // Step 4: Merge embedded child Presenter blocks (using FULL data)
    stepEmbeds(builder, rawForEmbeds, isArray, snapshot, ctx);
    // Step 5: Attach UI blocks (using FULL validated data)
    stepUiBlocks(builder, validated, isArray, snapshot, ctx);
    // Step 6: Attach rules (using FULL validated data)
    stepRules(builder, validated, isArray, snapshot, ctx);
    // Step 7: Attach action suggestions (using FULL validated data)
    stepSuggestions(builder, validated, isArray, snapshot, ctx);
    return { builder, validated, isArray };
}
