// @mcpfusion/core/dist/presenter/Presenter.js
/**
 * Presenter — MVA View Layer for AI Agents
 *
 * Domain-level "lens" that defines how data is presented to an LLM.
 * A Presenter validates data through a Zod schema, attaches JIT system
 * rules, and renders deterministic UI blocks — all at Node.js speed.
 *
 * The Presenter is **domain-level**, not tool-level. The same Presenter
 * (e.g. `InvoicePresenter`) can be reused across any tool that returns
 * invoices, guaranteeing consistent AI perception of the domain.
 *
 * ## Advanced Features
 *
 * - **Context-Aware**: Callbacks receive `TContext` for RBAC, DLP, locale
 * - **Cognitive Guardrails**: `.agentLimit()` truncates large arrays
 * - **Agentic Affordances**: `.suggestActions()` for HATEOAS-style hints
 * - **Composition**: `.embed()` nests child Presenters (DRY relational data)
 *
 * @example
 * ```typescript
 * import { createPresenter, ui } from '@mcpfusion/core';
 * import { z } from 'zod';
 *
 * export const InvoicePresenter = createPresenter('Invoice')
 *     .schema(z.object({
 *         id: z.string(),
 *         amount_cents: z.number(),
 *         status: z.enum(['paid', 'pending']),
 *     }))
 *     .systemRules((invoice, ctx) => [
 *         'CRITICAL: amount_cents is in CENTS. Divide by 100.',
 *         ctx?.user?.role !== 'admin' ? 'Mask sensitive financial data.' : null,
 *     ])
 *     .uiBlocks((invoice) => [
 *         ui.echarts({
 *             series: [{ type: 'gauge', data: [{ value: invoice.amount_cents / 100 }] }],
 *         }),
 *     ])
 *     .agentLimit(50, (omitted) =>
 *         ui.summary(`⚠️ Dataset truncated. 50 shown, ${omitted} hidden.`)
 *     )
 *     .suggestActions((invoice) => {
 *         if (invoice.status === 'pending') {
 *             return [{ tool: 'billing.pay', reason: 'Offer immediate payment' }];
 *         }
 *         return [];
 *     });
 * ```
 *
 * @see {@link createPresenter} for the factory function
 * @see {@link ResponseBuilder} for manual response composition
 *
 * @module
 */
import { z, ZodType } from 'zod';
import {} from './ui.js';
import { extractZodKeys } from './SelectUtils.js';
import { defaultSerializer } from '../core/serialization/JsonSerializer.js';
import { compileRedactor } from './RedactEngine.js';
import { executePipelineWithData, } from './PresenterPipeline.js';
import { evaluateRules } from './PromptFirewall.js';
// ── Brand ────────────────────────────────────────────────
const PRESENTER_BRAND = 'MCPFusionPresenter';
/**
 * Check if a value is a {@link Presenter} instance.
 *
 * Used by the execution pipeline to detect Presenters declared
 * via the `returns` field in action configuration.
 *
 * @param value - Any value
 * @returns `true` if the value is a Presenter
 */
export function isPresenter(value) {
    return (typeof value === 'object' &&
        value !== null &&
        '__brand' in value &&
        value.__brand === PRESENTER_BRAND);
}
// ── Presenter ────────────────────────────────────────────
/**
 * Domain-level Presenter — the "View" in MVA (Model-View-Agent).
 *
 * Encapsulates:
 * - **Schema** (Zod): Validates and filters data before it reaches the LLM
 * - **System Rules**: JIT context directives that travel with the data
 * - **UI Blocks**: SSR-rendered visual blocks (charts, diagrams, tables)
 * - **Agent Limit**: Smart truncation for large collections
 * - **Action Suggestions**: HATEOAS-style next-action hints
 * - **Embeds**: Relational Presenter composition (DRY)
 *
 * @typeParam T - The validated output type (inferred from the Zod schema)
 *
 * @see {@link createPresenter} for the factory function
 */
export class Presenter {
    /** @internal Brand for type detection in the pipeline */
    __brand = PRESENTER_BRAND;
    /** @internal Human-readable domain name (for debugging) */
    name;
    _schema;
    _rules = [];
    _itemUiBlocks;
    _collectionUiBlocks;
    _suggestActions;
    _collectionSuggestActions;
    _agentLimit;
    _embeds = [];
    _sealed = false;
    _collectionRules = [];
    _compiledStringify;
    _compiledRedactor;
    _redactConfig;
    _redactPaths = [];
    _asyncItemUiBlocks;
    _asyncCollectionUiBlocks;
    _asyncRules;
    _asyncSuggestActions;
    _promptFirewall;
    /** @internal Use {@link createPresenter} factory instead */
    constructor(name) {
        this.name = name;
    }
    // ── Configuration Guards ─────────────────────────────
    /**
     * Ensure the Presenter is not sealed.
     * Throws a clear error if `.make()` has already been called.
     * @internal
     */
    _assertNotSealed() {
        if (this._sealed) {
            throw new Error(`Presenter "${this.name}" is sealed after first .make() call. ` +
                `Configuration must be done before .make() is called.`);
        }
    }
    // Implementation — accepts ZodType, plain object shapes, or Model objects
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
    schema(schemaOrShape) {
        this._assertNotSealed();
        const narrowed = this;
        // Detect Model object (duck typing via `fields` + `schema` properties)
        if (schemaOrShape != null && typeof schemaOrShape === 'object' && 'fields' in schemaOrShape && 'schema' in schemaOrShape && schemaOrShape.schema instanceof ZodType) {
            narrowed._schema = schemaOrShape.schema;
        }
        else if (schemaOrShape instanceof ZodType) {
            // Detect if it's an already-constructed ZodType (has _def property)
            narrowed._schema = schemaOrShape;
        }
        else {
            // Plain object shape → wrap in z.object()
            narrowed._schema = z.object(schemaOrShape);
        }
        // AOT: Lazy-compile a fast stringify function for this schema.
        // Falls back to JSON.stringify if fast-json-stringify is unavailable.
        narrowed._compiledStringify = defaultSerializer.compile(narrowed._schema);
        return narrowed;
    }
    /**
     * Set JIT system rules that travel with the data.
     *
     * Rules are **Context Tree-Shaking**: they only appear in the LLM's
     * context when this specific domain's data is returned.
     *
     * Accepts either a **static array** or a **dynamic function** that
     * receives the data and request context for RBAC/DLP/locale rules.
     * Return `null` from dynamic rules to conditionally exclude them.
     *
     * @param rules - Array of rule strings, or a function `(data, ctx?) => (string | null)[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * // Static rules
     * .systemRules(['CRITICAL: amounts in CENTS.'])
     *
     * // Dynamic context-aware rules (RBAC)
     * .systemRules((user, ctx) => [
     *     ctx?.user?.role !== 'admin' ? 'Mask email and phone.' : null,
     *     `Format dates using ${ctx?.tenant?.locale ?? 'en-US'}`
     * ])
     * ```
     */
    systemRules(rules) {
        this._assertNotSealed();
        this._rules = rules;
        return this;
    }
    /**
     * Define UI blocks generated for a **single data item**.
     *
     * The callback receives the validated data item and optionally the
     * request context. Return `null` for conditional blocks (filtered
     * automatically).
     *
     * @param fn - `(item, ctx?) => (UiBlock | null)[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * .uiBlocks((invoice, ctx) => [
     *     ui.echarts({ series: [...] }),
     *     ctx?.user?.department === 'finance' ? ui.echarts(advancedChart) : null,
     * ])
     * ```
     */
    uiBlocks(fn) {
        this._assertNotSealed();
        this._itemUiBlocks = fn;
        return this;
    }
    /**
     * Define aggregated UI blocks for a **collection** (array) of items.
     *
     * When the handler returns an array, this callback is called **once**
     * with the entire validated array. Prevents N individual charts
     * from flooding the LLM's context.
     *
     * @param fn - `(items[], ctx?) => (UiBlock | null)[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * .collectionUiBlocks((invoices) => [
     *     ui.echarts({ xAxis: { data: invoices.map(i => i.id) } }),
     *     ui.summary(`${invoices.length} invoices found.`),
     * ])
     * ```
     */
    collectionUiBlocks(fn) {
        this._assertNotSealed();
        this._collectionUiBlocks = fn;
        return this;
    }
    /**
     * Set a cognitive guardrail that truncates large collections.
     *
     * Protects against context DDoS: if a tool returns thousands of
     * records, the Presenter automatically truncates the data array
     * and injects a summary block teaching the AI to use filters.
     *
     * @param max - Maximum items to keep in the data array
     * @param onTruncate - Callback that generates a warning UI block.
     *                     Receives the count of omitted items.
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * .agentLimit(50, (omitted) =>
     *     ui.summary(`⚠️ Truncated. 50 shown, ${omitted} hidden. Use filters.`)
     * )
     * ```
     */
    agentLimit(max, onTruncate) {
        this._assertNotSealed();
        this._agentLimit = { max, onTruncate };
        return this;
    }
    /**
     * Define HATEOAS-style action suggestions based on data state.
     *
     * The callback inspects the data and returns suggested next tools,
     * guiding the AI through the business state machine. Eliminates
     * routing hallucinations by providing explicit next-step hints.
     *
     * @param fn - `(data, ctx?) => ActionSuggestion[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * .suggestActions((invoice, ctx) => {
     *     if (invoice.status === 'pending') {
     *         return [
     *             { tool: 'billing.pay', reason: 'Offer immediate payment' },
     *             { tool: 'billing.send_reminder', reason: 'Send reminder email' },
     *         ];
     *     }
     *     return [];
     * })
     * ```
     */
    suggestActions(fn) {
        this._assertNotSealed();
        this._suggestActions = fn;
        return this;
    }
    /**
     * Define HATEOAS-style action suggestions for **collections**.
     *
     * Unlike `.suggestActions()`, this callback receives the **entire array**
     * of validated items, enabling aggregate-level suggestions like batch
     * operations, bulk approvals, or summary insights.
     *
     * When both `.suggestActions()` and `.collectionSuggestActions()` are set,
     * only collectionSuggestActions is used for arrays.
     *
     * @param fn - `(items[], ctx?) => (ActionSuggestion | null)[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * .collectionSuggestActions((invoices) => [
     *     invoices.some(i => i.status === 'overdue')
     *         ? { tool: 'billing.batch_remind', reason: 'Send batch reminders' }
     *         : null,
     *     invoices.length > 100
     *         ? { tool: 'billing.export', reason: 'Export results for offline review' }
     *         : null,
     * ])
     * ```
     */
    collectionSuggestActions(fn) {
        this._assertNotSealed();
        this._collectionSuggestActions = fn;
        return this;
    }
    // ── Fluent Aliases ───────────────────────────────────
    /**
     * Alias for `.suggestActions()` — fluent shorthand.
     *
     * Define HATEOAS-style action suggestions based on data state.
     * Use with the `suggest()` helper for maximum fluency.
     *
     * @param fn - `(data, ctx?) => (ActionSuggestion | null)[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * import { suggest } from '@mcpfusion/core';
     *
     * .suggest((inv) => [
     *     suggest('invoices.get', 'View details'),
     *     inv.status === 'overdue'
     *         ? suggest('billing.remind', 'Send reminder')
     *         : null,
     * ])
     * ```
     */
    suggest(fn) {
        return this.suggestActions(fn);
    }
    /**
     * Alias for `.collectionSuggestActions()` — fluent shorthand.
     *
     * @param fn - `(items[], ctx?) => (ActionSuggestion | null)[]`
     * @returns `this` for chaining
     */
    collectionSuggest(fn) {
        return this.collectionSuggestActions(fn);
    }
    /**
     * Alias for `.systemRules()` — fluent shorthand.
     *
     * @param rules - Static rules array or dynamic `(data, ctx?) => (string | null)[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * .rules(['CRITICAL: amounts in CENTS.'])
     * .rules((inv) => [
     *     inv.status === 'overdue' ? '⚠️ OVERDUE' : null,
     * ])
     * ```
     */
    rules(rules) {
        return this.systemRules(rules);
    }
    /**
     * Set collection-level system rules.
     *
     * Unlike `.systemRules()`, these rules are evaluated with the **entire
     * array** of validated items. Use for aggregate context (totals, counts,
     * mixed-status warnings) that cannot be derived from a single item.
     *
     * Both per-item and collection rules are merged in the response.
     *
     * @param rules - Static rules array, or dynamic `(items[], ctx?) => (string | null)[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * .collectionRules((invoices, ctx) => [
     *     `Total: ${invoices.length} invoices found.`,
     *     invoices.some(i => i.status === 'overdue')
     *         ? '⚠️ Some invoices are OVERDUE. Highlight them.'
     *         : null,
     * ])
     * ```
     */
    collectionRules(rules) {
        this._assertNotSealed();
        this._collectionRules = rules;
        return this;
    }
    /**
     * Alias for `.uiBlocks()` — fluent shorthand.
     *
     * @param fn - `(item, ctx?) => (UiBlock | null)[]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * .ui((inv) => [
     *     ui.echarts({ series: [{ type: 'gauge', data: [{ value: inv.amount / 100 }] }] }),
     * ])
     * ```
     */
    ui(fn) {
        return this.uiBlocks(fn);
    }
    /**
     * Cognitive guardrail shorthand with auto-generated message.
     *
     * Truncates large collections and injects a smart summary block.
     * For custom truncation messages, use `.agentLimit(max, onTruncate)` instead.
     *
     * @param max - Maximum items to keep in the data array
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * // Auto-generated message:
     * .limit(50)
     * // → "⚠️ Dataset truncated. 50 shown, {omitted} hidden. Use filters to narrow results."
     *
     * // For custom message, use agentLimit():
     * .agentLimit(50, (omitted) => ui.summary(`Custom: ${omitted} hidden`))
     * ```
     */
    limit(max) {
        return this.agentLimit(max, (omitted) => ({
            type: 'summary',
            content: `📊 **Summary**: ⚠️ Dataset truncated. ${max} shown, ${omitted} hidden. Use filters to narrow results.`,
        }));
    }
    /**
     * Compose a child Presenter for a nested relation.
     *
     * When `data[key]` exists, the child Presenter's rules and UI blocks
     * are merged into the parent response. This is the DRY solution for
     * relational data: define `ClientPresenter` once, embed it everywhere.
     *
     * @param key - The property key containing the nested data
     * @param childPresenter - The Presenter to apply to `data[key]`
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * import { ClientPresenter } from './ClientPresenter';
     *
     * export const InvoicePresenter = createPresenter('Invoice')
     *     .schema(invoiceSchema)
     *     .embed('client', ClientPresenter);
     * ```
     */
    embed(key, childPresenter) {
        this._assertNotSealed();
        this._embeds.push({ key, presenter: childPresenter });
        return this;
    }
    // ── DLP Compliance (PII Redaction) ────────────────────
    /**
     * Declare PII fields to redact before data leaves the framework.
     *
     * Uses `fast-redact` (Pino's V8-optimized serialization engine) to
     * compile object paths into hyper-fast masking functions at config
     * time — zero overhead on the hot path.
     *
     * The redaction is applied **structurally** on the wire-facing data
     * (after `_select` filter, before `ResponseBuilder`). UI blocks and
     * system rules still see the **full unmasked data** (Late Guillotine
     * pattern preserved).
     *
     * Requires `fast-redact` as an optional peer dependency.
     * If not installed, passes data through unmodified (defensive fallback).
     *
     * @param paths - Object paths to redact. Supports dot notation,
     *   bracket notation, wildcards (`'*'`), and array indices.
     * @param censor - Replacement value. Default: `'[REDACTED]'`.
     *   Can be a function `(originalValue) => maskedValue`.
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * // Basic PII masking
     * createPresenter('Patient')
     *     .schema({ name: t.string, ssn: t.string, diagnosis: t.string })
     *     .redactPII(['ssn', 'diagnosis'])
     *
     * // Wildcard — redact all nested SSN fields
     * createPresenter('Users')
     *     .redactPII(['*.ssn', '*.password', 'credit_card.number'])
     *
     * // Array wildcard — redact diagnosis in all patients
     * createPresenter('Hospital')
     *     .redactPII(['patients[*].diagnosis', 'patients[*].ssn'])
     *
     * // Custom censor — last 4 digits visible
     * createPresenter('Payment')
     *     .redactPII(['credit_card.number'], (v) => '****-****-****-' + String(v).slice(-4))
     * ```
     *
     * @see {@link https://github.com/davidmarkclements/fast-redact | fast-redact}
     */
    redactPII(paths, censor) {
        this._assertNotSealed();
        const config = {
            paths,
            ...(censor !== undefined ? { censor } : {}),
        };
        this._redactConfig = config;
        this._compiledRedactor = compileRedactor(config);
        this._redactPaths = paths;
        return this;
    }
    /**
     * Alias for `.redactPII()` — fluent shorthand.
     *
     * @param paths - Object paths to redact
     * @param censor - Replacement value or function
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * createPresenter('User')
     *     .schema({ name: t.string, ssn: t.string })
     *     .redact(['ssn'])
     * ```
     */
    redact(paths, censor) {
        return this.redactPII(paths, censor);
    }
    // ── Introspection (read-only metadata accessors) ─────
    /**
     * Get the Zod schema's top-level keys.
     *
     * Returns the field names declared in the Presenter's schema.
     * Safe to call at any time — does NOT seal the Presenter.
     *
     * @returns Array of key names, or empty array if no schema is set
     */
    getSchemaKeys() {
        if (!this._schema)
            return [];
        return extractZodKeys(this._schema);
    }
    /**
     * Get the DLP redaction paths configured on this Presenter.
     *
     * Returns the paths passed to `.redactPII()` / `.redact()`.
     * Empty array if no redaction is configured.
     *
     * @internal Used by PostProcessor for `dlp.redact` telemetry
     */
    getRedactPaths() {
        return this._redactPaths;
    }
    /**
     * Get the agent limit config (if set via `.agentLimit()` or `.limit()`).
     *
     * @internal Used by PostProcessor for guardrail telemetry
     * @returns The max limit, or `undefined` if no limit is configured
     */
    getAgentLimitMax() {
        return this._agentLimit?.max;
    }
    /**
     * Get which UI block factory methods were configured.
     *
     * Inspects the configuration callbacks to determine supported
     * UI block types. Does NOT execute any callbacks.
     *
     * @returns Array of UI block type labels
     */
    getUiBlockTypes() {
        const types = [];
        if (this._itemUiBlocks)
            types.push('item');
        if (this._collectionUiBlocks)
            types.push('collection');
        // Note: specific types (echarts, mermaid, etc.) are only known at
        // runtime when callbacks execute. We report callback presence here.
        return types;
    }
    /**
     * Whether the Presenter uses dynamic (context-aware) system rules.
     *
     * Static rules (string arrays) are NOT contextual.
     * Functions `(data, ctx?) => ...` ARE contextual.
     *
     * @returns `true` if rules are a function
     */
    hasContextualRules() {
        return typeof this._rules === 'function';
    }
    /**
     * Return static rule strings for introspection hashing.
     *
     * If rules are dynamic (function), returns an empty array because
     * the actual rule content depends on runtime data/context.
     *
     * @returns Static rule strings, or empty array if rules are contextual
     */
    getStaticRuleStrings() {
        if (typeof this._rules === 'function')
            return [];
        return this._rules;
    }
    // ── Async Configuration ──────────────────────────────
    /**
     * Register an **async** UI block callback for single items.
     *
     * Use when UI generation requires I/O (database, API, file system).
     * Must be consumed via `makeAsync()` — `make()` ignores async callbacks.
     *
     * @param fn - Async function receiving validated data + optional context
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * createPresenter('Invoice')
     *     .asyncUiBlocks(async (inv, ctx) => {
     *         const history = await ctx.db.payments.history(inv.id);
     *         return [ui.echarts(buildTimeline(history))];
     *     });
     * ```
     */
    asyncUiBlocks(fn) {
        this._asyncItemUiBlocks = fn;
        return this;
    }
    /**
     * Register an **async** UI block callback for collections.
     *
     * @param fn - Async function receiving the full validated array + optional context
     * @returns `this` for chaining
     */
    asyncCollectionUiBlocks(fn) {
        this._asyncCollectionUiBlocks = fn;
        return this;
    }
    /**
     * Register **async** system rules generation.
     *
     * @param fn - Async function receiving validated data + optional context
     * @returns `this` for chaining
     */
    asyncRules(fn) {
        this._asyncRules = fn;
        return this;
    }
    /**
     * Register **async** action suggestions generation.
     *
     * @param fn - Async function receiving validated data + optional context
     * @returns `this` for chaining
     */
    asyncSuggestActions(fn) {
        this._asyncSuggestActions = fn;
        return this;
    }
    /**
     * Enable LLM-as-Judge evaluation of dynamic system rules.
     *
     * The PromptFirewall evaluates all resolved rules (sync + async)
     * through a judge chain before they reach the LLM. This prevents
     * prompt injection via tainted data interpolated in `.systemRules()`.
     *
     * **Important:** Presenters with a firewall MUST use `makeAsync()`.
     * Calling `make()` will throw an error.
     *
     * @param config - Firewall configuration (adapter or chain)
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * createPresenter('Invoice')
     *     .systemRules((inv) => [`Status: ${inv.description}`])
     *     .promptFirewall({
     *         adapter: { name: 'gpt-4o-mini', evaluate: (p) => openai.chat(p) },
     *     });
     * ```
     */
    promptFirewall(config) {
        this._assertNotSealed();
        this._promptFirewall = config;
        return this;
    }
    // ── Execution ────────────────────────────────────────
    /**
     * Compose a {@link ResponseBuilder} from raw data.
     *
     * Orchestrates: truncate → validate → embed → render UI → attach rules
     * → suggest actions → **Late Guillotine** (`_select` filter).
     *
     * **Late Guillotine pattern**: UI blocks, system rules, and action
     * suggestions are computed using the **full** validated data, ensuring
     * charts and rules never see `undefined` for pruned fields. Only the
     * wire-facing data block in the ResponseBuilder is filtered by `_select`
     * — the UI consumes full data in RAM, the AI consumes pruned data on
     * the wire.
     *
     * After the first call, the Presenter is sealed (immutable).
     *
     * **Auto-detection**: If `data` is an array, items are validated
     * individually and `collectionUiBlocks` is called (if defined).
     * Otherwise, `uiBlocks` is called for the single item.
     *
     * @param data - Raw data from the handler (object or array)
     * @param ctx - Optional request context (for RBAC, locale, etc.)
     * @param selectFields - Optional top-level field names to keep in the
     *   data block. When provided, only these keys survive in the JSON
     *   payload sent to the AI. Nested objects are kept whole (shallow).
     * @returns A {@link ResponseBuilder} ready for chaining or `.build()`
     * @throws If Zod validation fails
     *
     * @example
     * ```typescript
     * // Full data (default)
     * return InvoicePresenter.make(rawInvoice).build();
     *
     * // With _select filtering — only 'status' reaches the AI
     * return InvoicePresenter.make(rawInvoice, ctx, ['status']).build();
     * ```
     */
    make(data, ctx, selectFields) {
        // Guard: PromptFirewall requires makeAsync()
        if (this._promptFirewall) {
            throw new Error(`Presenter "${this.name}" has a PromptFirewall configured. ` +
                `Use makeAsync() instead of make(). The firewall requires async ` +
                `evaluation to filter system rules through the LLM judge.`);
        }
        // Seal on first use — configuration is frozen from here
        this._sealed = true;
        // Delegate entirely to the decomposed pipeline
        return executePipelineWithData(data, this._toSnapshot(), ctx, selectFields).builder;
    }
    // ── Async Make ───────────────────────────────────────
    /**
     * Check if this Presenter has any async callbacks configured.
     *
     * Used by the pipeline to decide between sync `make()` and async
     * `makeAsync()`. When no async callbacks are set, `makeAsync()` is
     * equivalent to `Promise.resolve(make())`, so the sync path is preferred.
     *
     * @returns `true` if any async callback is configured
     */
    hasAsyncCallbacks() {
        return !!(this._asyncItemUiBlocks || this._asyncCollectionUiBlocks
            || this._asyncRules || this._asyncSuggestActions
            || this._promptFirewall);
    }
    /**
     * Async version of `make()` — enriches the response with async callbacks.
     *
     * Runs all sync steps first (via `make()`), then awaits async callbacks
     * and appends their results to the builder. The sync `make()` method
     * remains unchanged (zero breaking changes).
     *
     * @param data - Raw data from the handler (object or array)
     * @param ctx - Optional request context
     * @param selectFields - Optional top-level field names for context window optimization
     * @returns A Promise resolving to a {@link ResponseBuilder}
     *
     * @example
     * ```typescript
     * const presenter = createPresenter('Invoice')
     *     .schema(invoiceSchema)
     *     .asyncUiBlocks(async (inv, ctx) => {
     *         const history = await ctx.db.payments.history(inv.id);
     *         return [ui.echarts(buildTimeline(history))];
     *     });
     *
     * // In handler:
     * const builder = await presenter.makeAsync(data, ctx);
     * return builder.build();
     * ```
     */
    async makeAsync(data, ctx, selectFields) {
        // Step 1: Run all sync steps (via pipeline)
        // Bypass the make() guard for firewall-enabled presenters
        this._sealed = true;
        const snapshot = this._toSnapshot();
        // executePipelineWithData returns both the builder AND the validated data,
        // so async callbacks receive the same truncated+validated dataset
        // without re-running stepTruncate()+stepValidate().
        const { builder, validated, isArray } = executePipelineWithData(data, snapshot, ctx, selectFields);
        // Step 2: Async enrichment — append after sync blocks.
        // Reuses the pipeline's validated data for all async callbacks.
        // Async UI blocks
        if (isArray && this._asyncCollectionUiBlocks) {
            const blocks = await this._asyncCollectionUiBlocks(validated, ctx);
            const filtered = blocks.filter(Boolean);
            if (filtered.length > 0)
                builder.uiBlocks(filtered);
        }
        else if (!isArray && this._asyncItemUiBlocks) {
            const blocks = await this._asyncItemUiBlocks(validated, ctx);
            const filtered = blocks.filter(Boolean);
            if (filtered.length > 0)
                builder.uiBlocks(filtered);
        }
        // Async rules
        if (this._asyncRules) {
            const singleData = isArray ? validated[0] : validated;
            if (singleData !== undefined) {
                const rules = await this._asyncRules(singleData, ctx);
                const filtered = rules.filter((r) => r !== null);
                if (filtered.length > 0)
                    builder.systemRules(filtered);
            }
        }
        // Async suggestions
        if (this._asyncSuggestActions) {
            const singleData = isArray ? validated[0] : validated;
            if (singleData !== undefined) {
                const suggestions = await this._asyncSuggestActions(singleData, ctx);
                const filtered = suggestions.filter((s) => s !== null);
                if (filtered.length > 0)
                    builder.systemHint(filtered);
            }
        }
        // Step 3: PromptFirewall — filter ALL accumulated rules via LLM judge
        if (this._promptFirewall) {
            const currentRules = builder.getRules();
            if (currentRules.length > 0) {
                const verdict = await evaluateRules(currentRules, this._promptFirewall);
                builder.replaceRules(verdict.allowed);
            }
        }
        return builder;
    }
    // ── Pipeline Snapshot ─────────────────────────────────
    /**
     * Create a read-only snapshot of the Presenter's configuration
     * for use by the decomposed pipeline steps.
     *
     * @returns A {@link PresenterSnapshot} capturing the current config
     * @internal
     */
    _toSnapshot() {
        return {
            name: this.name,
            schema: this._schema,
            rules: this._rules,
            collectionRules: this._collectionRules,
            itemUiBlocks: this._itemUiBlocks,
            collectionUiBlocks: this._collectionUiBlocks,
            suggestActions: this._suggestActions,
            collectionSuggestActions: this._collectionSuggestActions,
            agentLimit: this._agentLimit,
            embeds: this._embeds,
            redactConfig: this._redactConfig,
            compiledRedactor: this._compiledRedactor,
            compiledStringify: this._compiledStringify,
            writeBackRedactor: (fn) => { this._compiledRedactor = fn; },
        };
    }
}
// ── Factory ──────────────────────────────────────────────
/**
 * Create a new domain-level Presenter.
 *
 * The Presenter defines how a specific domain model (Invoice, Task,
 * Project) is presented to AI agents. It is **reusable** across any
 * tool that returns that model.
 *
 * @param name - Human-readable domain name (for debugging/logging)
 * @returns A new {@link Presenter} ready for configuration
 *
 * @example
 * ```typescript
 * import { createPresenter, ui } from '@mcpfusion/core';
 *
 * export const TaskPresenter = createPresenter('Task')
 *     .schema(taskSchema)
 *     .systemRules(['Use emojis: 🔄 In Progress, ✅ Done'])
 *     .uiBlocks((task) => [ui.markdown(`**${task.title}**: ${task.status}`)]);
 * ```
 *
 * @see {@link Presenter} for the full API
 */
export function createPresenter(name) {
    return new Presenter(name);
}
