// @mcpfusion/core/dist/presenter/RedactEngine.js
/**
 * RedactEngine — DLP Compliance Engine (Zero-Leak PII Protection)
 *
 * Compiles object paths into V8-optimized redaction functions using
 * `fast-redact` (the Pino logger serialization engine). Masks sensitive
 * fields structurally before data leaves the framework.
 *
 * Architecture:
 *   ┌──────────────────────────────────────────────────┐
 *   │  Boot / .redactPII() call                       │
 *   │                                                  │
 *   │  ['*.ssn', 'credit_card.number']                │
 *   │       │                                          │
 *   │       ▼                                          │
 *   │  fast-redact({ paths, censor, serialize: false })│
 *   │       │                                          │
 *   │       ▼                                          │
 *   │  Compiled RedactFn (V8-optimized)               │
 *   │  Applied in Presenter.make() on wireData        │
 *   └──────────────────────────────────────────────────┘
 *
 * Properties:
 * - Compiles paths into V8-optimized functions at config time
 * - Zero overhead on hot path (pre-compiled redaction)
 * - Wildcards supported: `'*.ssn'`, `'patients[*].diagnosis'`
 * - Custom censor: string or function `(value) => maskedValue`
 * - Zero-risk fallback: if `fast-redact` is not installed, logs warning
 *   and passes data through unmodified
 *
 * @module
 * @internal
 */
import { toErrorMessage } from '../core/ErrorUtils.js';
let _fastRedact = null;
// promise gate to serialize concurrent lazy-import calls
let _loadPromise = null;
async function loadFastRedact() {
    if (_fastRedact === false)
        return null; // already tried, not available
    if (_fastRedact !== null)
        return _fastRedact;
    // if another call is already importing, await the same promise
    if (_loadPromise)
        return _loadPromise;
    _loadPromise = (async () => {
        try {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const mod = await import('fast-redact');
            _fastRedact = (typeof mod === 'function' ? mod : mod.default);
            return _fastRedact;
        }
        catch {
            _fastRedact = false; // mark as unavailable
            return null;
        }
    })();
    return _loadPromise;
}
// Synchronous version — uses the cached function if already loaded.
function getFastRedact() {
    if (_fastRedact === false || _fastRedact === null)
        return null;
    return _fastRedact;
}
// ── Compile Logic ────────────────────────────────────────
/**
 * Compile a redaction config into a V8-optimized redaction function.
 *
 * The compiled function uses `structuredClone` to avoid mutating the
 * original data (required for Late Guillotine: UI blocks and rules
 * need the full unmasked data).
 *
 * @param config - Redaction configuration (paths + optional censor)
 * @returns A compiled `RedactFn`, or `undefined` if `fast-redact` is unavailable
 *
 * @example
 * ```typescript
 * const redact = compileRedactor({
 *     paths: ['*.ssn', 'credit_card.number'],
 *     censor: '[REDACTED]',
 * });
 *
 * if (redact) {
 *     const safe = redact(sensitiveData);
 *     // safe.ssn === '[REDACTED]'
 * }
 * ```
 */
export function compileRedactor(config) {
    if (config.paths.length === 0)
        return undefined;
    const factory = getFastRedact();
    if (!factory)
        return undefined;
    try {
        const redactor = factory({
            paths: [...config.paths],
            censor: config.censor ?? '[REDACTED]',
            serialize: false, // mutate-in-place mode (we clone first)
        });
        // Return a safe function that clones → redacts → returns clone
        return (data) => {
            if (data === null || data === undefined || typeof data !== 'object') {
                return data; // primitives pass through
            }
            try {
                const clone = structuredClone(data);
                redactor(clone);
                return clone;
            }
            catch (err) {
                // never return unredacted data — throw explicitly.
                // The caller must handle the error rather than silently leaking PII.
                throw new Error(`[mcpfusion] PII redaction failed: ${toErrorMessage(err)}. ` +
                    'Data withheld to prevent PII leak.', { cause: err });
            }
        };
    }
    catch {
        // Compilation failed (invalid paths, etc.)
        return undefined;
    }
}
// ── Async Initialization ─────────────────────────────────
/**
 * Pre-load the `fast-redact` module.
 *
 * Called during `initMCPFusion()` boot sequence alongside other
 * optional dependencies. Not required — the engine degrades
 * gracefully if `fast-redact` is not installed.
 *
 * @returns `true` if `fast-redact` is available
 */
export async function initRedactEngine() {
    const factory = await loadFastRedact();
    return factory !== null;
}
