// @mcpfusion/core/dist/presenter/TelemetryCollector.js
/**
 * TelemetryCollector — Presenter Telemetry Emission
 *
 * Extracted from PostProcessor to uphold SRP.
 * Emits `presenter.slice`, `presenter.rules`, and `dlp.redact` events
 * with zero coupling to the post-processing priority logic.
 *
 * @internal
 * @module
 */
import {} from '../core/response.js';
import {} from './Presenter.js';
import {} from '../observability/TelemetryEvent.js';
// ── Shared Encoder ──────────────────────────────────────
const _encoder = new TextEncoder();
// ── Main Emitter ────────────────────────────────────────
/**
 * Emit all Presenter telemetry events for a single post-processing cycle.
 *
 * Emits up to 3 events:
 * 1. `presenter.slice` — raw vs wire bytes, rows, selectFields, guardrail info
 * 2. `presenter.rules` — extracted rule strings from `<domain_rules>` XML
 * 3. `dlp.redact` — PII redaction path count and list
 *
 * @param opts - Telemetry emission options
 * @internal
 */
export function emitPresenterTelemetry(opts) {
    const { sink, tool, action, response, presenter, rawJson, rawRows, selectFields } = opts;
    // ── 1. presenter.slice ──────────────────────────────
    const rawBytes = _encoder.encode(rawJson).byteLength;
    let wireBytes = 0;
    for (const c of response.content) {
        if ('text' in c && typeof c.text === 'string') {
            wireBytes += _encoder.encode(c.text).byteLength;
        }
    }
    const agentLimitMax = presenter.getAgentLimitMax();
    const wireRows = (agentLimitMax !== undefined && rawRows > agentLimitMax)
        ? agentLimitMax
        : rawRows;
    sink({
        type: 'presenter.slice',
        tool,
        action,
        rawBytes,
        wireBytes,
        rowsRaw: rawRows,
        rowsWire: wireRows,
        ...(selectFields && selectFields.length > 0 ? {
            selectFields,
            totalFields: presenter.getSchemaKeys().length || undefined,
        } : {}),
        ...(agentLimitMax !== undefined && rawRows > agentLimitMax ? {
            guardrailFrom: rawRows,
            guardrailTo: agentLimitMax,
            guardrailHint: 'Results truncated by agentLimit. Use pagination or filters.',
        } : {}),
        timestamp: Date.now(),
        // eslint-disable-next-line @typescript-eslint/no-explicit-any -- TelemetrySink accepts extensible event shapes
    });
    // ── 2. presenter.rules ──────────────────────────────
    const rulesFromResponse = [];
    for (const c of response.content) {
        if ('text' in c && typeof c.text === 'string') {
            const match = c.text.match(/<domain_rules>\n([\s\S]*?)\n<\/domain_rules>/);
            if (match) {
                rulesFromResponse.push(...match[1].split('\n').filter(Boolean).map(r => r.replace(/^- /, '')));
            }
        }
    }
    if (rulesFromResponse.length > 0) {
        sink({
            type: 'presenter.rules',
            tool,
            action,
            rules: rulesFromResponse,
            timestamp: Date.now(),
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- TelemetrySink accepts extensible event shapes
        });
    }
    // ── 3. dlp.redact ───────────────────────────────────
    const redactPaths = presenter.getRedactPaths();
    if (redactPaths.length > 0) {
        sink({
            type: 'dlp.redact',
            tool,
            action,
            fieldsRedacted: redactPaths.length,
            paths: [...redactPaths],
            timestamp: Date.now(),
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- TelemetrySink accepts extensible event shapes
        });
    }
}
