// @mcpfusion/core/dist/core/serialization/JsonSerializer.js
/**
 * JsonSerializer — AOT JSON Serialization Engine
 *
 * Compiles Zod schemas into hyper-fast stringify functions using
 * `fast-json-stringify` (the Fastify serialization engine).
 *
 * Architecture:
 *   ┌──────────────────────────────────────────────────┐
 *   │  Boot / First Call (AOT / Lazy)                  │
 *   │                                                  │
 *   │  ZodType ──► zod-to-json-schema ──► JSON Schema  │
 *   │                    │                             │
 *   │                    ▼                             │
 *   │            fast-json-stringify                   │
 *   │                    │                             │
 *   │                    ▼                             │
 *   │         Compiled StringifyFn (C-like)            │
 *   │         Cached via WeakMap<ZodType>              │
 *   └──────────────────────────────────────────────────┘
 *
 * Properties:
 * - 2-5x faster than `JSON.stringify` for known schemas
 * - Lazy-by-default: compiles on first use, cached via WeakMap
 * - Zero-risk fallback: if `fast-json-stringify` is missing or
 *   compilation fails, falls back to native `JSON.stringify`
 * - Produces compact/minified JSON (no indentation) for max throughput
 * - Safe optional/nullable handling via schema transformation
 *
 * @module
 * @internal
 */
import { zodToJsonWithOptions } from '../schema/ZodCompat.js';
let _fjsBuild = null;
async function loadFastJsonStringify() {
    if (_fjsBuild === false)
        return null; // already tried, not available
    if (_fjsBuild !== null)
        return _fjsBuild;
    try {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const mod = await import('fast-json-stringify');
        _fjsBuild = (typeof mod === 'function' ? mod : mod.default);
        return _fjsBuild;
    }
    catch {
        _fjsBuild = false; // mark as unavailable
        return null;
    }
}
// Synchronous version — uses the cached function if already loaded.
function getFjsBuild() {
    if (_fjsBuild === false || _fjsBuild === null)
        return null;
    return _fjsBuild;
}
// ── Schema Cache ─────────────────────────────────────────
// WeakMap ensures schemas are compiled exactly once and can be GC'd
// when the schema reference is released.
const _cache = new WeakMap();
// ── JSON Schema Transformation ───────────────────────────
/**
 * Transform a JSON Schema for safe use with fast-json-stringify.
 *
 * - Strips `$schema` (not supported by fjs)
 * - Adds `additionalProperties: true` to objects so unknown fields
 *   are serialized instead of silently dropped
 * - Wraps types with nullable variants where appropriate
 *
 * @internal
 */
function transformForFjs(jsonSchema) {
    const schema = { ...jsonSchema };
    // Remove $schema — fast-json-stringify doesn't support it
    delete schema['$schema'];
    // For object-type schemas, allow additional properties
    // so dynamic/runtime fields are not silently dropped
    if (schema['type'] === 'object') {
        if (schema['additionalProperties'] === undefined) {
            schema['additionalProperties'] = true;
        }
    }
    // Recursively transform nested properties
    if (schema['properties'] != null && typeof schema['properties'] === 'object') {
        const props = { ...schema['properties'] };
        for (const [key, value] of Object.entries(props)) {
            if (value != null && typeof value === 'object') {
                props[key] = transformForFjs(value);
            }
        }
        schema['properties'] = props;
    }
    // Transform array items
    if (schema['items'] != null && typeof schema['items'] === 'object') {
        schema['items'] = transformForFjs(schema['items']);
    }
    // Handle anyOf / oneOf for nullable unions
    for (const combiner of ['anyOf', 'oneOf', 'allOf']) {
        if (Array.isArray(schema[combiner])) {
            schema[combiner] = schema[combiner]
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- runtime: anyOf entries may be null/primitive in real-world JSON Schemas
                .map(s => (s != null && typeof s === 'object' ? transformForFjs(s) : s));
        }
    }
    return schema;
}
// ── Compile Logic ────────────────────────────────────────
/**
 * Core compilation: Zod schema → JSON Schema → fast-json-stringify function.
 *
 * @internal
 */
function compileSchema(schema) {
    const build = getFjsBuild();
    if (!build)
        return undefined;
    // Check cache first (schema as WeakMap key)
    if (typeof schema === 'object' && schema !== null && _cache.has(schema)) {
        return _cache.get(schema);
    }
    try {
        // Step 1: Zod → JSON Schema
        const rawJsonSchema = zodToJsonWithOptions(schema, {
            $refStrategy: 'none',
        });
        // Step 2: Transform for fjs compatibility
        const fjsSchema = transformForFjs(rawJsonSchema);
        // Step 3: Compile the stringify function
        const stringify = build(fjsSchema, {
            largeArrayMechanism: 'json-stringify',
        });
        // Cache the compiled function
        if (typeof schema === 'object' && schema !== null) {
            _cache.set(schema, stringify);
        }
        return stringify;
    }
    catch {
        // Compilation failed — fall back to native JSON.stringify
        // This is expected for very complex schemas that fjs can't handle
        return undefined;
    }
}
// ── Public Factory ───────────────────────────────────────
/**
 * Create an AOT JSON serializer.
 *
 * The serializer lazy-loads `fast-json-stringify` and provides
 * compile/stringify methods with automatic fallback.
 *
 * @example
 * ```typescript
 * const serializer = createSerializer();
 * await serializer.init(); // optional: pre-load the fjs module
 *
 * const stringify = serializer.compile(myZodSchema);
 * if (stringify) {
 *     const json = stringify(data); // 2-5x faster
 * }
 * ```
 */
export function createSerializer() {
    return {
        async init() {
            await loadFastJsonStringify();
        },
        compile(schema) {
            return compileSchema(schema);
        },
        stringify(data, compiled) {
            if (compiled) {
                try {
                    return compiled(data);
                }
                catch {
                    // Defensive: if compiled function fails at runtime
                    // (e.g., data shape mismatch), fall back to native
                }
            }
            return JSON.stringify(data);
        },
    };
}
// ── Singleton ────────────────────────────────────────────
/**
 * Global singleton serializer.
 *
 * Shared across the engine so all Presenters and response helpers
 * use the same cache and fjs module reference.
 */
export const defaultSerializer = createSerializer();
