// @mcpfusion/core/dist/core/builder/ActionGroupBuilder.js
/**
 * ActionGroupBuilder — Sub-builder for Hierarchical Action Groups
 *
 * Used within {@link GroupedToolBuilder.group} callbacks to register
 * actions under a named group. Supports group-scoped middleware and
 * generates compound keys (e.g., `"users.create"`).
 *
 * @typeParam TContext - Application context type
 * @typeParam TCommon - Common schema shape (inferred from parent builder)
 *
 * @example
 * ```typescript
 * createTool<AppContext>('platform')
 *     .group('users', 'User management', g => g
 *         .use(requireAdmin)
 *         .query('list', async (ctx) => success(await ctx.db.users.findMany()))
 *         .mutation('ban', async (ctx, args) => {
 *             await ctx.db.users.ban(args.user_id);
 *             return success('User banned');
 *         })
 *     );
 * ```
 *
 * @see {@link GroupedToolBuilder.group} for creating groups
 * @see {@link MiddlewareFn} for middleware signature
 *
 * @module
 */
import {} from 'zod';
import {} from '../types.js';
import { resolveMiddleware } from '../middleware/ContextDerivation.js';
// ── Shared Config → InternalAction Mapper ────────────────
/**
 * Map `ActionConfig` properties to `InternalAction` base fields.
 *
 * Both `GroupedToolBuilder.action()` and `ActionGroupBuilder.action()`
 * perform this same mapping. Extracted here to eliminate duplication
 * and ensure a single source of truth.
 *
 * @param config - The action configuration from the public API
 * @param omitCommonFields - Resolved omitCommon fields (already merged/deduped)
 * @returns Base fields for building an `InternalAction`
 *
 * @internal
 */
export function mapConfigToActionFields(config, omitCommonFields) {
    return {
        actionName: config.name,
        description: config.description ?? undefined,
        compactDescription: config.compactDescription ?? undefined,
        schema: config.schema ?? undefined,
        destructive: config.destructive ?? undefined,
        idempotent: config.idempotent ?? undefined,
        readOnly: config.readOnly ?? undefined,
        handler: config.handler,
        omitCommonFields: (omitCommonFields?.length ?? 0) > 0 ? omitCommonFields : undefined,
        returns: config.returns ?? undefined,
    };
}
// ── ActionGroupBuilder ───────────────────────────────────
export class ActionGroupBuilder {
    /** @internal */
    _actions = [];
    _groupName;
    _groupDescription;
    _groupMiddlewares = [];
    _groupOmitCommon = [];
    constructor(groupName, description) {
        this._groupName = groupName;
        this._groupDescription = description || '';
    }
    /**
     * Add middleware scoped to this group only.
     *
     * Unlike {@link GroupedToolBuilder.use}, this middleware runs
     * only for actions within this group — not globally.
     *
     * Accepts both `MiddlewareDefinition` from `f.middleware()` and
     * raw `MiddlewareFn` functions.
     *
     * @param mw - Middleware function or MiddlewareDefinition
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * builder.group('admin', 'Admin operations', (g) => g
     *     .use(requireAdmin)
     *     .query('list', listHandler)
     *     .mutation('reset', resetHandler)
     * );
     * ```
     *
     * @see {@link MiddlewareFn} for the middleware signature
     */
    use(mw) {
        this._groupMiddlewares.push(resolveMiddleware(mw));
        return this;
    }
    /**
     * Omit common schema fields for all actions in this group.
     *
     * Use when an entire group derives common fields from context
     * (e.g. a "profile" group that resolves `workspace_id` from the JWT).
     *
     * Per-action `omitCommon` merges with group-level omissions.
     *
     * @param fields - Common field names to omit
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * builder.group('profile', 'User profile', (g) => g
     *     .omitCommon('workspace_id')
     *     .query('me', meHandler)
     * );
     * ```
     */
    omitCommon(...fields) {
        this._groupOmitCommon.push(...fields);
        return this;
    }
    // ── Semantic Verb Shortcuts ──────────────────────────
    /**
     * Register a **read-only** action (readOnly: true).
     *
     * Semantic shortcut — eliminates the need for config objects.
     * The action name is automatically prefixed with the group name
     * (e.g., `"list"` in group `"users"` → `"users.list"`).
     *
     * @param name - Action name (must not contain dots)
     * @param handler - Handler function
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * builder.group('users', 'User management', (g) => g
     *     .query('list', async (ctx) => success(await ctx.db.users.findMany()))
     *     .query('get', async (ctx, args) => success(await ctx.db.users.find(args.id)))
     * );
     * ```
     */
    query(name, handler) {
        return this.action({ name, readOnly: true, handler });
    }
    /**
     * Register a **destructive** action (destructive: true).
     *
     * Semantic shortcut — eliminates the need for config objects.
     * Signals to the LLM that this action has irreversible side effects.
     *
     * @param name - Action name (must not contain dots)
     * @param handler - Handler function
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * builder.group('users', 'User management', (g) => g
     *     .mutation('ban', async (ctx, args) => {
     *         await ctx.db.users.ban(args.user_id);
     *         return success('User banned');
     *     })
     * );
     * ```
     */
    mutation(name, handler) {
        return this.action({ name, destructive: true, handler });
    }
    action(configOrName, maybeHandler) {
        // 2-arg shorthand: action('invite', inviteHandler)
        if (typeof configOrName === 'string' && typeof maybeHandler !== 'function') {
            throw new Error(`action("${configOrName}") requires a handler function as the second argument.`);
        }
        const config = typeof configOrName === 'string'
            ? { name: configOrName, handler: maybeHandler }
            : configOrName;
        if (config.name.includes('.')) {
            throw new Error(`Action name "${config.name}" must not contain dots. ` +
                `The framework uses dots internally for group.action compound keys.`);
        }
        // Merge group-level + per-action omissions (deduped)
        const perAction = config.omitCommon ?? [];
        const mergedOmit = [...new Set([...this._groupOmitCommon, ...perAction])];
        this._actions.push({
            key: `${this._groupName}.${config.name}`,
            groupName: this._groupName,
            groupDescription: this._groupDescription,
            ...mapConfigToActionFields(config, mergedOmit),
            middlewares: this._groupMiddlewares.length > 0
                ? [...this._groupMiddlewares] : undefined,
        });
        return this;
    }
}
