// @mcpfusion/core/dist/core/builder/defineTool.js
/**
 * defineTool() — High-Level Tool Definition (Vercel/Stripe Style)
 *
 * The recommended entry point for building MCP tools. Write a plain
 * JSON-like config object — zero Zod imports, zero builder patterns.
 * The framework converts everything to Zod schemas internally.
 *
 * @example
 * ```typescript
 * import { defineTool, success } from '@mcpfusion/core';
 *
 * export const projects = defineTool('projects', {
 *     description: 'Manage workspace projects',
 *     shared: { workspace_id: 'string' },
 *     actions: {
 *         list:   { readOnly: true, handler: async (ctx, args) => success([]) },
 *         create: { params: { name: 'string' }, handler: async (ctx, args) => success(args.name) },
 *     },
 * });
 * ```
 *
 * @see {@link createTool} for the power-user builder API
 * @see {@link ToolRegistry} for registering tools
 *
 * @module
 */
import {} from 'zod';
import { GroupedToolBuilder } from './GroupedToolBuilder.js';
import {} from '../types.js';
import { resolveMiddleware } from '../middleware/ContextDerivation.js';
import {} from '../../presenter/Presenter.js';
import { convertParamsToZod, } from './ParamDescriptors.js';
import { isZodSchema } from '../schema/SchemaUtils.js';
// ============================================================================
// defineTool()
// ============================================================================
/**
 * Resolve params: if ParamsMap → convertParamsToZod, if ZodObject → passthrough.
 * @internal
 */
function resolveSchema(params) {
    if (!params)
        return undefined;
    if (isZodSchema(params))
        return params;
    return convertParamsToZod(params);
}
/**
 * Define a tool using a high-level JSON-like config.
 *
 * This is the recommended entry point for most developers.
 * The framework handles all Zod schema creation, validation,
 * and MCP protocol details internally.
 *
 * @example
 * ```typescript
 * const echo = defineTool('echo', {
 *     actions: {
 *         say: {
 *             params: { message: 'string' },
 *             handler: async (ctx, args) => success(args.message),
 *         },
 *     },
 * });
 *
 * // Tool with shared params + groups + middleware
 * const platform = defineTool<AppContext>('platform', {
 *     description: 'Platform management',
 *     tags: ['admin'],
 *     shared: { workspace_id: { type: 'string', description: 'Workspace ID' } },
 *     middleware: [requireAuth],
 *     groups: {
 *         users: {
 *             description: 'User management',
 *             actions: {
 *                 list: { readOnly: true, handler: listUsers },
 *                 ban:  { destructive: true, params: { user_id: 'string' }, handler: banUser },
 *             },
 *         },
 *     },
 * });
 *
 * // Register normally
 * const registry = new ToolRegistry<AppContext>();
 * registry.register(platform);
 * ```
 *
 * @see {@link createTool} for the power-user builder API
 * @see {@link ToolRegistry.register} for registration
 */
export function defineTool(name, config) {
    const { description, tags, discriminator, toonDescription, annotations, shared, middleware, actions, groups } = config;
    // ── Guard: actions and groups are mutually exclusive ──
    if (actions && groups) {
        throw new Error(`defineTool("${name}"): "actions" and "groups" are mutually exclusive. ` +
            `Use "actions" for flat tools OR "groups" for hierarchical tools, not both.`);
    }
    const builder = new GroupedToolBuilder(name);
    // ── Builder configuration (declarative) ──
    if (description)
        builder.description(description);
    if (tags != null && tags.length > 0)
        builder.tags(...tags);
    if (discriminator)
        builder.discriminator(discriminator);
    if (toonDescription)
        builder.toonDescription();
    if (annotations)
        builder.annotations(annotations);
    const sharedSchema = resolveSchema(shared);
    if (sharedSchema)
        builder.commonSchema(sharedSchema);
    middleware?.forEach(mw => builder.use(resolveMiddleware(mw)));
    // ── Register actions/groups ──
    if (actions) {
        for (const [actionName, actionDef] of Object.entries(actions)) {
            builder.action(buildActionConfig(actionName, actionDef));
        }
    }
    if (groups) {
        for (const [groupName, groupDef] of Object.entries(groups)) {
            registerGroup(builder, groupName, groupDef);
        }
    }
    return builder;
}
// ============================================================================
// Internal Helpers
// ============================================================================
/**
 * Build an ActionConfig from an ActionDef — single source of truth
 * for translating high-level descriptors to builder config.
 * @internal
 */
function buildActionConfig(actionName, def) {
    const schema = resolveSchema(def.params);
    return {
        name: actionName,
        handler: def.handler,
        ...(def.description && { description: def.description }),
        ...(schema && { schema }),
        ...(def.readOnly !== undefined && { readOnly: def.readOnly }),
        ...(def.destructive !== undefined && { destructive: def.destructive }),
        ...(def.idempotent !== undefined && { idempotent: def.idempotent }),
        ...((def.omitCommon?.length ?? 0) > 0 && { omitCommon: def.omitCommon }),
        ...(def.returns && { returns: def.returns }),
    };
}
/**
 * Register a group with its actions on a builder.
 * @internal
 */
function registerGroup(builder, groupName, def) {
    builder.group(groupName, def.description ?? '', g => {
        if (def.omitCommon != null && def.omitCommon.length > 0) {
            g.omitCommon(...def.omitCommon);
        }
        def.middleware?.forEach(mw => g.use(resolveMiddleware(mw)));
        for (const [actionName, actionDef] of Object.entries(def.actions)) {
            g.action(buildActionConfig(actionName, actionDef));
        }
    });
}
