// @mcpfusion/core/dist/core/builder/FluentToolBuilder.js
/**
 * FluentToolBuilder — Type-Chaining Builder for Semantic Verb Tools
 *
 * The core builder behind `f.query()`, `f.mutation()`, and `f.action()`.
 * Uses TypeScript generic accumulation so that each fluent step narrows
 * the types — the IDE "magically" knows the exact shape of `input` and
 * `ctx` inside `.handle()` without any manual Interface declaration.
 *
 * @example
 * ```typescript
 * const f = initMCPFusion<AppContext>();
 *
 * const listUsers = f.query('users.list')
 *     .describe('List users from the database')
 *     .withNumber('limit', 'Max results to return')
 *     .withOptionalEnum('status', ['active', 'inactive'], 'Filter by status')
 *     .returns(UserPresenter)
 *     .handle(async (input, ctx) => {
 *         return ctx.db.user.findMany({ take: input.limit });
 *     });
 * ```
 *
 * @see {@link FluentRouter} for prefix grouping
 * @see {@link initMCPFusion} for the factory that creates these builders
 *
 * @module
 */
import { z } from 'zod';
import { compileFieldForInput } from '../../model/defineModel.js';
import {} from '../types.js';
import {} from '../../presenter/Presenter.js';
import {} from '../execution/ConcurrencyGuard.js';
import {} from '../../sandbox/SandboxEngine.js';
import {} from '../middleware/ContextDerivation.js';
import {} from './SemanticDefaults.js';
import { createProxyHandler } from './ProxyHandler.js';
import { buildToolFromFluent } from './BuildPipeline.js';
// Re-export SemanticDefaults for backward compatibility
export { QUERY_DEFAULTS, MUTATION_DEFAULTS, ACTION_DEFAULTS } from './SemanticDefaults.js';
// ── Array Item Type Resolution ───────────────────────────
/** Resolve Zod type from array item type string */
function resolveArrayItemType(itemType) {
    switch (itemType) {
        case 'string': return z.string();
        case 'number': return z.number();
        case 'boolean': return z.boolean();
    }
}
// ── Schema Description Helper ─────────────────────────────
/**
 * Apply a `.describe()` annotation to a Zod schema only when a description
 * is provided. Eliminates the repeated ternary in every `with*()` method:
 *   `description ? z.X().describe(description) : z.X()`
 * → `withDesc(z.X(), description)`
 *
 * @internal
 */
function withDesc(schema, description) {
    return description ? schema.describe(description) : schema;
}
// ── FluentToolBuilder ────────────────────────────────────
/**
 * Fluent builder that accumulates types at each step.
 *
 * @typeParam TContext - Base application context (from `initMCPFusion<TContext>()`)
 * @typeParam TInput - Accumulated input type (built by `with*()` methods)
 * @typeParam TCtx - Accumulated context type (enriched by `.use()`)
 */
export class FluentToolBuilder {
    /** @internal */ _name;
    /** @internal */ _description;
    /** @internal */ _instructions;
    /** @internal */ _inputSchema;
    /** @internal */ _withParams = {};
    /** @internal */ _tags = [];
    /** @internal */ _modelRef;
    /**
     * @internal reject duplicate parameter names.
     * All `withXxx()` methods delegate to this instead of assigning directly.
     */
    _addParam(name, schema) {
        if (!name || !name.trim()) {
            throw new Error(`Empty parameter name on tool "${this._name}". ` +
                `Each parameter must have a non-empty name.`);
        }
        if (Object.prototype.hasOwnProperty.call(this._withParams, name)) {
            throw new Error(`Duplicate parameter name "${name}" on tool "${this._name}". ` +
                `Each parameter must have a unique name.`);
        }
        this._withParams[name] = schema;
    }
    /** @internal */ _middlewares = [];
    /** @internal */ _returns;
    /** @internal */ _semanticDefaults;
    /** @internal */ _readOnly;
    /** @internal */ _destructive;
    /** @internal */ _idempotent;
    /** @internal */ _toonMode = false;
    /** @internal */ _annotations;
    /** @internal */ _invalidatesPatterns = [];
    /** @internal */ _cacheControl;
    /** @internal */ _concurrency;
    /** @internal */ _egressMaxBytes;
    /** @internal */ _sandboxConfig;
    /** @internal */ _handlerSet = false;
    /** @internal */ _fsmStates;
    /** @internal */ _fsmTransition;
    /** @internal */ _interactive = false;
    /** @internal */ _compactDescription;
    /**
     * @param name - Tool name in `domain.action` format (e.g. `'users.list'`)
     * @param defaults - Semantic defaults from the verb (`query`, `mutation`, `action`)
     */
    constructor(name, defaults = {}) {
        if (!name || !name.trim()) {
            throw new Error('Tool name must be a non-empty string. ' +
                'Use the "domain.action" format (e.g. "users.list").');
        }
        this._name = name;
        this._semanticDefaults = defaults;
    }
    // ── Configuration (fluent, each returns narrowed type) ──
    /**
     * Set the tool description shown to the LLM.
     *
     * @param text - Human-readable description
     * @returns `this` for chaining
     */
    describe(text) {
        this._description = text;
        return this;
    }
    /**
     * Set a compact description for FSM progressive disclosure.
     *
     * This short description (~200 chars) is shown during `tools/list` discovery
     * when the FSM is in 'compact' state. After the first successful call, the
     * FSM transitions to 'expert' state and the full `.describe()` content is
     * exposed via `listChanged` re-fetch.
     *
     * @param text - Short, functional description of what the tool does
     * @returns `this` for chaining
     */
    compactDescription(text) {
        this._compactDescription = text;
        return this;
    }
    /**
     * Set AI-First instructions — injected as system-level guidance in the tool description.
     *
     * This is **Prompt Engineering embedded in the framework**. The instructions
     * tell the LLM WHEN and HOW to use this tool, reducing hallucination.
     *
     * @param text - System prompt for the tool
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * f.query('docs.search')
     *     .describe('Search internal documentation')
     *     .instructions('Use ONLY when the user asks about internal policies.')
     *     .withString('query', 'Search term')
     *     .handle(async (input) => { ... });
     * ```
     */
    instructions(text) {
        this._instructions = text;
        return this;
    }
    // ── Parameter Declaration (with* methods) ────────────
    /**
     * Add a required string parameter.
     *
     * @param name - Parameter name
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('projects.get')
     *     .withString('project_id', 'The project ID to retrieve')
     *     .handle(async (input) => { ... });
     * // input.project_id: string ✅
     * ```
     */
    withString(name, description) {
        this._addParam(name, withDesc(z.string(), description));
        return this;
    }
    /**
     * Add an optional string parameter.
     *
     * @param name - Parameter name
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     */
    withOptionalString(name, description) {
        this._addParam(name, withDesc(z.string(), description).optional());
        return this;
    }
    /**
     * Add a required number parameter.
     *
     * @param name - Parameter name
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     */
    withNumber(name, description) {
        this._addParam(name, withDesc(z.number(), description));
        return this;
    }
    /**
     * Add an optional number parameter.
     *
     * @param name - Parameter name
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     */
    withOptionalNumber(name, description) {
        this._addParam(name, withDesc(z.number(), description).optional());
        return this;
    }
    /**
     * Add a required boolean parameter.
     *
     * @param name - Parameter name
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     */
    withBoolean(name, description) {
        this._addParam(name, withDesc(z.boolean(), description));
        return this;
    }
    /**
     * Add an optional boolean parameter.
     *
     * @param name - Parameter name
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     */
    withOptionalBoolean(name, description) {
        this._addParam(name, withDesc(z.boolean(), description).optional());
        return this;
    }
    // ── Bulk Parameter Declaration ────────────────────────
    //
    // Batch variants accept a Record<string, string> where
    // keys are parameter names and values are descriptions.
    //
    // Before:
    //   .withOptionalString('title', 'Filter by title')
    //   .withOptionalString('status', 'Status filter')
    //
    // After:
    //   .withOptionalStrings({ title: 'Filter by title', status: 'Status filter' })
    /**
     * Add multiple required string parameters in bulk.
     *
     * @param fields - Record of `{ paramName: 'description' }`
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('tasks.get')
     *     .withStrings({
     *         company_slug: 'Workspace identifier',
     *         project_slug: 'Project identifier',
     *     })
     *     .handle(async (input) => { ... });
     * // input.company_slug: string ✅
     * // input.project_slug: string ✅
     * ```
     */
    withStrings(fields) {
        for (const [name, description] of Object.entries(fields)) {
            this._addParam(name, withDesc(z.string(), description));
        }
        return this;
    }
    /**
     * Add multiple optional string parameters in bulk.
     *
     * @param fields - Record of `{ paramName: 'description' }`
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('tasks.filter')
     *     .withOptionalStrings({
     *         title:    'Filter by title (partial match)',
     *         workflow: 'Column name (e.g. "In Progress")',
     *         labels:   'Comma-separated label names',
     *     })
     *     .handle(async (input) => { ... });
     * // input.title?: string | undefined ✅
     * ```
     */
    withOptionalStrings(fields) {
        for (const [name, description] of Object.entries(fields)) {
            this._addParam(name, withDesc(z.string(), description).optional());
        }
        return this;
    }
    /**
     * Add multiple required number parameters in bulk.
     *
     * @param fields - Record of `{ paramName: 'description' }`
     * @returns Builder with narrowed `TInput` type
     */
    withNumbers(fields) {
        for (const [name, description] of Object.entries(fields)) {
            this._addParam(name, withDesc(z.number(), description));
        }
        return this;
    }
    /**
     * Add multiple optional number parameters in bulk.
     *
     * @param fields - Record of `{ paramName: 'description' }`
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('tasks.filter')
     *     .withOptionalNumbers({
     *         per_page:  'Results per page (default: 50)',
     *         offset:    'Pagination offset',
     *     })
     *     .handle(async (input) => { ... });
     * // input.per_page?: number | undefined ✅
     * ```
     */
    withOptionalNumbers(fields) {
        for (const [name, description] of Object.entries(fields)) {
            this._addParam(name, withDesc(z.number(), description).optional());
        }
        return this;
    }
    /**
     * Add multiple required boolean parameters in bulk.
     *
     * @param fields - Record of `{ paramName: 'description' }`
     * @returns Builder with narrowed `TInput` type
     */
    withBooleans(fields) {
        for (const [name, description] of Object.entries(fields)) {
            this._addParam(name, withDesc(z.boolean(), description));
        }
        return this;
    }
    /**
     * Add multiple optional boolean parameters in bulk.
     *
     * @param fields - Record of `{ paramName: 'description' }`
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('tasks.filter')
     *     .withOptionalBooleans({
     *         is_blocker:  'Only blockers',
     *         is_bug:      'Only bugs',
     *         unassigned:  'Only unassigned tasks',
     *         is_archived: 'Include archived tasks',
     *     })
     *     .handle(async (input) => { ... });
     * // input.is_blocker?: boolean | undefined ✅
     * ```
     */
    withOptionalBooleans(fields) {
        for (const [name, description] of Object.entries(fields)) {
            this._addParam(name, withDesc(z.boolean(), description).optional());
        }
        return this;
    }
    /**
     * Add a required enum parameter.
     *
     * @param name - Parameter name
     * @param values - Allowed enum values
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('invoices.list')
     *     .withEnum('status', ['draft', 'sent', 'paid'], 'Filter by status')
     *     .handle(async (input) => { ... });
     * // input.status: 'draft' | 'sent' | 'paid' ✅
     * ```
     */
    withEnum(name, values, description) {
        this._addParam(name, withDesc(z.enum(values), description));
        return this;
    }
    /**
     * Add an optional enum parameter.
     *
     * @param name - Parameter name
     * @param values - Allowed enum values
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     */
    withOptionalEnum(name, values, description) {
        this._addParam(name, withDesc(z.enum(values), description).optional());
        return this;
    }
    /**
     * Add a required array parameter.
     *
     * @param name - Parameter name
     * @param itemType - Type of array items (`'string'`, `'number'`, `'boolean'`)
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.mutation('tasks.tag')
     *     .withString('task_id', 'The task to tag')
     *     .withArray('tags', 'string', 'Tags to apply')
     *     .handle(async (input) => { ... });
     * // input.tags: string[] ✅
     * ```
     */
    withArray(name, itemType, description) {
        this._addParam(name, withDesc(z.array(resolveArrayItemType(itemType)), description));
        return this;
    }
    /**
     * Add an optional array parameter.
     *
     * @param name - Parameter name
     * @param itemType - Type of array items (`'string'`, `'number'`, `'boolean'`)
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     */
    withOptionalArray(name, itemType, description) {
        this._addParam(name, withDesc(z.array(resolveArrayItemType(itemType)), description).optional());
        return this;
    }
    // ── JSON Object Parameters ─────────────────────────────
    /**
     * Add a required JSON object parameter.
     *
     * Accepts an arbitrary key–value object. Useful for tools that
     * receive structured payloads, configuration maps, or filter sets.
     *
     * @param name - Parameter name
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.action('reports.generate')
     *     .withJson('payload', 'Report configuration (dates, metrics, grouping)')
     *     .handle(async (input) => { ... });
     * // input.payload: Record<string, unknown> ✅
     * ```
     */
    withJson(name, description) {
        this._addParam(name, withDesc(z.record(z.string(), z.any()), description));
        return this;
    }
    /**
     * Add an optional JSON object parameter.
     *
     * @param name - Parameter name
     * @param description - Human-readable description for the LLM
     * @returns Builder with narrowed `TInput` type
     */
    withOptionalJson(name, description) {
        this._addParam(name, withDesc(z.record(z.string(), z.any()), description).optional());
        return this;
    }
    // ── Bulk Enum & Array Declaration ─────────────────────
    /**
     * Add multiple required enum parameters in bulk.
     *
     * Each entry is `paramName: [allowedValues, description?]`.
     *
     * @param fields - Record of `{ paramName: [values, description?] }`
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('tasks.filter')
     *     .withEnums({
     *         status:   [['open', 'closed', 'archived'], 'Task status'],
     *         priority: [['low', 'medium', 'high'], 'Priority level'],
     *     })
     *     .handle(async (input) => { ... });
     * // input.status: 'open' | 'closed' | 'archived' ✅
     * ```
     */
    withEnums(fields) {
        for (const [name, [values, description]] of Object.entries(fields)) {
            this._addParam(name, withDesc(z.enum(values), description));
        }
        return this;
    }
    /**
     * Add multiple optional enum parameters in bulk.
     *
     * Each entry is `paramName: [allowedValues, description?]`.
     *
     * @param fields - Record of `{ paramName: [values, description?] }`
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('tasks.filter')
     *     .withOptionalEnums({
     *         status:   [['open', 'closed'], 'Filter by status'],
     *         priority: [['low', 'medium', 'high'], 'Filter by priority'],
     *     })
     *     .handle(async (input) => { ... });
     * // input.status?: 'open' | 'closed' | undefined ✅
     * ```
     */
    withOptionalEnums(fields) {
        for (const [name, [values, description]] of Object.entries(fields)) {
            this._addParam(name, withDesc(z.enum(values), description).optional());
        }
        return this;
    }
    /**
     * Add multiple required array parameters in bulk, sharing the same item type.
     *
     * @param itemType - Type of array items (`'string'`, `'number'`, `'boolean'`)
     * @param fields - Record of `{ paramName: 'description' }`
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.mutation('tasks.update')
     *     .withArrays('string', {
     *         tags:   'Tags to apply',
     *         labels: 'Category labels',
     *     })
     *     .handle(async (input) => { ... });
     * // input.tags: string[] ✅
     * // input.labels: string[] ✅
     * ```
     */
    withArrays(itemType, fields) {
        const base = z.array(resolveArrayItemType(itemType));
        for (const [name, description] of Object.entries(fields)) {
            this._addParam(name, withDesc(base, description));
        }
        return this;
    }
    /**
     * Add multiple optional array parameters in bulk, sharing the same item type.
     *
     * @param itemType - Type of array items (`'string'`, `'number'`, `'boolean'`)
     * @param fields - Record of `{ paramName: 'description' }`
     * @returns Builder with narrowed `TInput` type
     *
     * @example
     * ```typescript
     * f.query('tasks.filter')
     *     .withOptionalArrays('string', {
     *         tags:   'Filter by tags',
     *         labels: 'Filter by labels',
     *     })
     *     .handle(async (input) => { ... });
     * // input.tags?: string[] | undefined ✅
     * ```
     */
    withOptionalArrays(itemType, fields) {
        const base = z.array(resolveArrayItemType(itemType));
        for (const [name, description] of Object.entries(fields)) {
            this._addParam(name, withDesc(base, description).optional());
        }
        return this;
    }
    // ── Model Integration ─────────────────────────────────
    /**
     * Derive tool input parameters from a Model's fillable profile.
     *
     * Reads the specified operation from `model.input` (fillable profiles),
     * then adds each field as a parameter using the type and description from
     * the Model's casts. For `create`, fields respect their schema optionality.
     * For `update` and `filter`, all fields become optional.
     *
     * @param model - A `Model` from `defineModel()`
     * @param operation - The fillable profile name (e.g. `'create'`, `'update'`, `'filter'`)
     * @returns Builder with additional parameters from the Model
     *
     * @example
     * ```typescript
     * f.mutation('task.create')
     *   .fromModel(TaskModel, 'create')
     *   .withStrings({ company_slug: '...', project_slug: '...' })
     *   .handle(async (input, ctx) => { ... });
     * ```
     */
    fromModel(model, operation) {
        const fieldNames = model.input[operation];
        if (!fieldNames) {
            throw new Error(`Model "${model.name}" has no fillable profile "${operation}". ` +
                `Available: ${Object.keys(model.input).join(', ') || 'none'}`);
        }
        // Determine if all fields should be forced optional (update/filter semantics)
        const forceOptional = operation !== 'create';
        for (const fieldName of fieldNames) {
            const fieldDef = model.fields[fieldName];
            if (!fieldDef) {
                throw new Error(`Model "${model.name}" fillable profile "${operation}" references ` +
                    `field "${fieldName}" which is not defined in casts.`);
            }
            // Compile FieldDef → Zod schema for input context
            const schema = compileFieldForInput(fieldDef, forceOptional);
            this._addParam(fieldName, schema);
        }
        // Store Model reference for .proxy() alias resolution
        this._modelRef = model;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return this;
    }
    use(mw) {
        // Handle MiddlewareDefinition from f.middleware()
        if (typeof mw !== 'function' && '__brand' in mw && mw.__brand === 'MiddlewareDefinition') {
            const def = mw;
            this._middlewares.push(def.toMiddlewareFn());
            return this;
        }
        // Handle inline function
        const inlineMw = mw;
        // Convert the fluent middleware signature to the standard MiddlewareFn
        const standardMw = async (ctx, args, next) => {
            const wrappedNext = async (enrichedCtx) => {
                // sanitize enriched context before merging
                // to prevent prototype pollution via __proto__/constructor/prototype keys.
                const safe = enrichedCtx;
                const target = ctx;
                for (const key of Object.keys(safe)) {
                    if (key === '__proto__' || key === 'constructor' || key === 'prototype')
                        continue;
                    target[key] = safe[key];
                }
                return next();
            };
            return inlineMw({ ctx: ctx, next: wrappedNext });
        };
        this._middlewares.push(standardMw);
        return this;
    }
    /**
     * Set the MVA Presenter for automatic response formatting.
     *
     * When a Presenter is attached, the handler can return raw data
     * and the framework pipes it through schema validation, system rules,
     * and UI block generation.
     *
     * @param presenter - A Presenter instance
     * @returns `this` for chaining
     */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- covariant: accept any Presenter subtype
    returns(presenter) {
        this._returns = presenter;
        return this;
    }
    /**
     * Add capability tags for selective tool exposure.
     *
     * Tags are accumulated — calling `.tags()` multiple times
     * (or inheriting from a router) appends rather than replaces.
     *
     * @param tags - Tag strings for filtering
     * @returns `this` for chaining
     */
    tags(...tags) {
        this._tags.push(...tags);
        return this;
    }
    // ── Semantic Overrides ───────────────────────────────
    /** Override: mark this tool as read-only (no side effects) */
    readOnly() {
        this._readOnly = true;
        return this;
    }
    /** Override: mark this tool as destructive (irreversible) */
    destructive() {
        this._destructive = true;
        return this;
    }
    /** Override: mark this tool as idempotent (safe to retry) */
    idempotent() {
        this._idempotent = true;
        return this;
    }
    /**
     * Enable TOON-formatted descriptions for token optimization.
     *
     * @returns `this` for chaining
     */
    toonDescription() {
        this._toonMode = true;
        return this;
    }
    /**
     * Set MCP tool annotations.
     *
     * @param a - Annotation key-value pairs
     * @returns `this` for chaining
     */
    annotations(a) {
        this._annotations = { ...this._annotations, ...a };
        return this;
    }
    // ── State Sync (Fluent) ──────────────────────────────
    /**
     * Declare glob patterns invalidated when this tool succeeds.
     *
     * @param patterns - Glob patterns (e.g. `'sprints.*'`, `'tasks.*'`)
     * @returns `this` for chaining
     */
    invalidates(...patterns) {
        this._invalidatesPatterns.push(...patterns);
        return this;
    }
    /**
     * Mark this tool's data as immutable (safe to cache forever).
     *
     * @returns `this` for chaining
     */
    cached() {
        this._cacheControl = 'immutable';
        return this;
    }
    /**
     * Mark this tool's data as volatile (never cache).
     *
     * @returns `this` for chaining
     */
    stale() {
        this._cacheControl = 'no-store';
        return this;
    }
    // ── Runtime Guards (Fluent) ──────────────────────────
    /**
     * Set concurrency limits for this tool (Semaphore + Queue pattern).
     *
     * @param config - Concurrency configuration
     * @returns `this` for chaining
     */
    concurrency(config) {
        this._concurrency = config;
        return this;
    }
    /**
     * Set maximum payload size for tool responses (Egress Guard).
     *
     * @param bytes - Maximum payload size in bytes
     * @returns `this` for chaining
     */
    egress(bytes) {
        this._egressMaxBytes = bytes;
        return this;
    }
    // ── Sandbox (Zero-Trust Compute) ─────────────────
    /**
     * Enable zero-trust sandboxed execution for this tool.
     *
     * When enabled:
     * 1. A `SandboxEngine` is lazily created on the `GroupedToolBuilder`
     * 2. A system instruction is auto-injected into the tool description
     *    (HATEOAS auto-prompting) so the LLM knows to send JS functions
     * 3. The handler can use `SandboxEngine.execute()` to run LLM code
     *    in a sealed V8 isolate (no process, require, fs, network)
     *
     * @param config - Optional sandbox configuration (timeout, memory, output size)
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * f.query('data.compute')
     *     .describe('Run safe computation on server data')
     *     .sandboxed({ timeout: 3000, memoryLimit: 64 })
     *     .withString('expression', 'JS arrow function: (data) => result')
     *     .handle(async (input, ctx) => {
     *         const data = await ctx.db.records.findMany();
     *         const engine = new SandboxEngine({ timeout: 3000 });
     *         const result = await engine.execute(input.expression, data);
     *         return result.ok ? result.value : { error: result.error };
     *     });
     * ```
     */
    sandboxed(config) {
        this._sandboxConfig = config ?? {};
        return this;
    }
    // ── Elicitation (Human-in-the-Loop) ──────────────────
    /**
     * Enable human-in-the-loop interaction for this tool.
     *
     * When enabled, the handler can use the standalone `ask()` function
     * to pause execution and request user input via a client-rendered form,
     * or `ask.redirect()` to send the user to an external URL.
     *
     * The transport context is bound via `AsyncLocalStorage` — the
     * developer never needs to pass or receive any transport objects.
     *
     * @returns `this` for chaining (no type augmentation needed — `ask()` is standalone)
     *
     * @example
     * ```typescript
     * import { ask } from '@mcpfusion/core';
     *
     * const deploy = f.mutation('infra.deploy')
     *     .withString('app_id', 'Application ID')
     *     .interactive()
     *     .handle(async (input) => {
     *         const prefs = await ask('Confirm deployment:', {
     *             region: ask.enum(['us-east-1', 'eu-west-1'] as const, 'Region'),
     *         });
     *         if (prefs.declined) return f.error('CANCELLED', 'Aborted');
     *         return { region: prefs.data.region };
     *     });
     * ```
     */
    interactive() {
        this._interactive = true;
        return this;
    }
    // ── FSM State Gate (Temporal Anti-Hallucination) ─────
    /**
     * Bind this tool to specific FSM states.
     *
     * When a `StateMachineGate` is configured on the server, this tool
     * will only appear in `tools/list` when the FSM is in one of the
     * specified states. The LLM physically cannot call tools that
     * don't exist in its reality.
     *
     * @param states - FSM state(s) where this tool is visible
     * @param transition - Event to send to the FSM on successful execution
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * // Visible only in 'has_items' state, sends CHECKOUT on success
     * const checkout = f.mutation('cart.checkout')
     *     .bindState('has_items', 'CHECKOUT')
     *     .handle(async (input, ctx) => { ... });
     *
     * // Visible in multiple states
     * const addItem = f.mutation('cart.add_item')
     *     .bindState(['empty', 'has_items'], 'ADD_ITEM')
     *     .handle(async (input, ctx) => { ... });
     * ```
     */
    bindState(states, transition) {
        this._fsmStates = Array.isArray(states) ? states : [states];
        if (transition !== undefined)
            this._fsmTransition = transition;
        return this;
    }
    // ── Terminal: handle() ───────────────────────────────
    /**
     * Set the handler and build the tool — the terminal step.
     *
     * The handler receives `(input, ctx)` with fully typed `TInput` and `TCtx`.
     * **Implicit `success()` wrapping**: if the handler returns raw data
     * (not a `ToolResponse`), the framework wraps it with `success()`.
     *
     * @param handler - Async function receiving typed `(input, ctx)`
     * @returns A `GroupedToolBuilder` ready for registration
     *
     * @example
     * ```typescript
     * const getProject = f.query('projects.get')
     *     .describe('Get a project by ID')
     *     .withString('project_id', 'The exact project ID')
     *     .handle(async (input, ctx) => {
     *         return await ctx.db.projects.findUnique({ where: { id: input.project_id } });
     *     });
     * ```
     */
    handle(handler) {
        return this._build(handler);
    }
    // ── Terminal: proxy() ────────────────────────────────
    /**
     * Auto-generate a handler that proxies to `ctx.client` HTTP methods.
     *
     * Eliminates boilerplate by deriving the HTTP method from the semantic
     * verb (`query` → GET, `mutation` → POST) and passing model input
     * directly as query params or request body.
     *
     * For tools with non-trivial logic (data transformation, multi-step
     * calls, conditional behavior), use `.handle()` instead.
     *
     * @param endpoint - API path (e.g. `'companies/standup/summary'`)
     * @param options - Optional overrides
     * @param options.method - Force HTTP method (`'GET'`, `'POST'`, `'PUT'`, `'DELETE'`)
     * @param options.unwrap - Auto-unwrap `response.data` (default: `true`)
     * @returns A `GroupedToolBuilder` ready for registration
     *
     * @example
     * ```typescript
     * const pulse = analytics.query('pulse')
     *     .fromModel(AnalyticsModel, 'query')
     *     .proxy('companies/manager-dashboard/pulse');
     *
     * const create = note.mutation('create')
     *     .fromModel(NoteModel, 'create')
     *     .proxy('notes');
     *
     * const update = note.mutation('update')
     *     .fromModel(NoteModel, 'update')
     *     .proxy('notes/:uuid', { method: 'PUT' });
     * ```
     */
    proxy(endpoint, options) {
        const httpMethod = options?.method
            ?? (this._semanticDefaults.readOnly ? 'GET' : 'POST');
        const shouldUnwrap = options?.unwrap ?? true;
        const handler = createProxyHandler(endpoint, httpMethod, shouldUnwrap, this._modelRef);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return this._build(handler);
    }
    /**
     * Alias for `.handle()` — for backward compatibility.
     * @internal
     */
    resolve(handler) {
        // Adapt { input, ctx } signature to (input, ctx)
        return this._build((input, ctx) => handler({ input, ctx }));
    }
    // ── Internal Build ───────────────────────────────────
    /** @internal */
    _build(handler) {
        // guard against double-invocation of handle()/resolve()
        if (this._handlerSet) {
            throw new Error(`handle() already called on tool "${this._name}". ` +
                `Each FluentToolBuilder can only have one handler.`);
        }
        this._handlerSet = true;
        return buildToolFromFluent({
            name: this._name,
            description: this._description,
            compactDescription: this._compactDescription,
            instructions: this._instructions,
            withParams: this._withParams,
            tags: this._tags,
            middlewares: this._middlewares,
            returns: this._returns,
            semanticDefaults: this._semanticDefaults,
            readOnly: this._readOnly,
            destructive: this._destructive,
            idempotent: this._idempotent,
            toonMode: this._toonMode,
            annotations: this._annotations,
            invalidatesPatterns: this._invalidatesPatterns,
            cacheControl: this._cacheControl,
            concurrency: this._concurrency,
            egressMaxBytes: this._egressMaxBytes,
            sandboxConfig: this._sandboxConfig,
            fsmStates: this._fsmStates,
            fsmTransition: this._fsmTransition,
            interactive: this._interactive,
            handler: handler,
        });
    }
}
