// @mcpfusion/core/dist/core/builder/FluentSchemaHelpers.js
// ── Sentinel for fluent descriptor detection ──────────────
/**
 * Symbol used to detect fluent schema helpers at runtime.
 * @internal
 */
export const FLUENT_DESCRIPTOR = Symbol.for('mcpfusion.fluent-descriptor');
/**
 * Type guard: is this value a FluentDescriptor?
 * @internal
 */
export function isFluentDescriptor(value) {
    return (typeof value === 'object' &&
        value !== null &&
        FLUENT_DESCRIPTOR in value &&
        value[FLUENT_DESCRIPTOR] === true);
}
// ── String Helper ────────────────────────────────────────
export class FluentString {
    [FLUENT_DESCRIPTOR] = true;
    _desc;
    /** @internal */ _optional = false;
    _min;
    _max;
    _regex;
    _default;
    _examples;
    /** Set minimum string length */
    min(n) { this._min = n; return this; }
    /** Set maximum string length */
    max(n) { this._max = n; return this; }
    /** Set regex validation pattern */
    regex(pattern) { this._regex = pattern; return this; }
    /** Mark as optional */
    optional() {
        this._optional = true;
        return this;
    }
    /** Set human-readable description */
    describe(text) { this._desc = text; return this; }
    /** Set default value (appended to description for LLM, auto-marks as optional) */
    default(value) { this._default = value; this._optional = true; return this; }
    /** Add a single LLM few-shot example (AI-First DX) */
    example(value) {
        if (!this._examples)
            this._examples = [];
        this._examples.push(value);
        return this;
    }
    /** Set multiple LLM few-shot examples */
    examples(...values) { this._examples = values; return this; }
    toDescriptor() {
        const desc = this._default !== undefined
            ? (this._desc ? `${this._desc} (default: '${this._default}')` : `(default: '${this._default}')`)
            : this._desc;
        return {
            type: 'string',
            ...(desc && { description: desc }),
            ...(this._optional && { optional: true }),
            ...(this._min !== undefined && { min: this._min }),
            ...(this._max !== undefined && { max: this._max }),
            ...(this._regex !== undefined && { regex: this._regex }),
            ...(this._examples && { examples: this._examples }),
        };
    }
}
// ── Number Helper ────────────────────────────────────────
export class FluentNumber {
    [FLUENT_DESCRIPTOR] = true;
    _desc;
    /** @internal */ _optional = false;
    _min;
    _max;
    _int = false;
    _default;
    _examples;
    /** Set minimum value */
    min(n) { this._min = n; return this; }
    /** Set maximum value */
    max(n) { this._max = n; return this; }
    /** Restrict to integers */
    int() { this._int = true; return this; }
    /** Mark as optional */
    optional() {
        this._optional = true;
        return this;
    }
    /** Set human-readable description */
    describe(text) { this._desc = text; return this; }
    /** Set default value (appended to description for LLM, auto-marks as optional) */
    default(value) { this._default = value; this._optional = true; return this; }
    /** Add a single LLM few-shot example (AI-First DX) */
    example(value) {
        if (!this._examples)
            this._examples = [];
        this._examples.push(value);
        return this;
    }
    /** Set multiple LLM few-shot examples */
    examples(...values) { this._examples = values; return this; }
    toDescriptor() {
        const desc = this._default !== undefined
            ? (this._desc ? `${this._desc} (default: ${this._default})` : `(default: ${this._default})`)
            : this._desc;
        return {
            type: 'number',
            ...(desc && { description: desc }),
            ...(this._optional && { optional: true }),
            ...(this._min !== undefined && { min: this._min }),
            ...(this._max !== undefined && { max: this._max }),
            ...(this._int && { int: true }),
            ...(this._examples && { examples: this._examples }),
        };
    }
}
// ── Boolean Helper ───────────────────────────────────────
export class FluentBoolean {
    [FLUENT_DESCRIPTOR] = true;
    _desc;
    /** @internal */ _optional = false;
    _default;
    /** Mark as optional */
    optional() {
        this._optional = true;
        return this;
    }
    /** Set human-readable description */
    describe(text) { this._desc = text; return this; }
    /** Set default value (appended to description for LLM, auto-marks as optional) */
    default(value) { this._default = value; this._optional = true; return this; }
    toDescriptor() {
        const desc = this._default !== undefined
            ? (this._desc ? `${this._desc} (default: ${this._default})` : `(default: ${this._default})`)
            : this._desc;
        return {
            type: 'boolean',
            ...(desc && { description: desc }),
            ...(this._optional && { optional: true }),
        };
    }
}
// ── Enum Helper ──────────────────────────────────────────
export class FluentEnum {
    [FLUENT_DESCRIPTOR] = true;
    _desc;
    /** @internal */ _optional = false;
    _examples;
    _default;
    _values;
    constructor(...values) {
        this._values = values;
    }
    /** Mark as optional */
    optional() {
        this._optional = true;
        return this;
    }
    /** Set human-readable description */
    describe(text) { this._desc = text; return this; }
    /** Set default value (appended to description for LLM) */
    default(value) { this._default = value; this._optional = true; return this; }
    /** Add a single LLM few-shot example (AI-First DX) */
    example(value) {
        if (!this._examples)
            this._examples = [];
        this._examples.push(value);
        return this;
    }
    /** Set multiple LLM few-shot examples */
    examples(...values) { this._examples = values; return this; }
    toDescriptor() {
        const desc = this._default !== undefined
            ? (this._desc ? `${this._desc} (default: '${this._default}')` : `(default: '${this._default}')`)
            : this._desc;
        return {
            enum: this._values,
            ...(desc && { description: desc }),
            ...(this._optional && { optional: true }),
            ...(this._examples && { examples: this._examples }),
        };
    }
}
export class FluentArray {
    [FLUENT_DESCRIPTOR] = true;
    _desc;
    /** @internal */ _optional = false;
    _min;
    _max;
    /** @internal */ _itemType;
    constructor(itemType) {
        this._itemType = itemType;
    }
    /** Set minimum array length */
    min(n) { this._min = n; return this; }
    /** Set maximum array length */
    max(n) { this._max = n; return this; }
    /** Mark as optional */
    optional() {
        this._optional = true;
        return this;
    }
    /** Set human-readable description */
    describe(text) { this._desc = text; return this; }
    toDescriptor() {
        return {
            array: this._itemType,
            ...(this._desc && { description: this._desc }),
            ...(this._optional && { optional: true }),
            ...(this._min !== undefined && { min: this._min }),
            ...(this._max !== undefined && { max: this._max }),
        };
    }
}
/**
 * Resolve a FluentParamsMap to a plain ParamsMap by calling .toDescriptor()
 * on any FluentDescriptor values.
 * @internal
 */
export function resolveFluentParams(input) {
    const resolved = {};
    for (const [key, value] of Object.entries(input)) {
        resolved[key] = isFluentDescriptor(value)
            ? value.toDescriptor()
            : value;
    }
    return resolved;
}
