// @mcpfusion/core/dist/core/builder/FluentRouter.js
/**
 * FluentRouter — Prefix Grouping for Fluent Tools
 *
 * Eliminates repetitive `'users.'` prefixes across dozens of tools.
 * A router shares a common prefix, description, and middleware chain
 * across all child tools.
 *
 * @example
 * ```typescript
 * const f = initMCPFusion<AppContext>();
 *
 * const users = f.router('users')
 *     .describe('User management')
 *     .use(requireAuth);
 *
 * // Tool name: "users", action: "list"
 * const listUsers = users.query('list')
 *     .input({ limit: f.number() })
 *     .resolve(async ({ input }) => { ... });
 *
 * // Tool name: "users", action: "delete"
 * const deleteUser = users.mutation('delete')
 *     .input({ id: f.string() })
 *     .resolve(async ({ input }) => { ... });
 * ```
 *
 * @module
 */
import {} from '../types.js';
import { resolveMiddleware } from '../middleware/ContextDerivation.js';
import { FluentToolBuilder, QUERY_DEFAULTS, MUTATION_DEFAULTS, ACTION_DEFAULTS, } from './FluentToolBuilder.js';
/**
 * Fluent router that shares prefix, description, and middleware
 * across child tools created via `.query()`, `.mutation()`, `.action()`.
 *
 * @typeParam TContext - Base application context
 */
export class FluentRouter {
    /** @internal */ _prefix;
    /** @internal */ _description;
    /** @internal */ _middlewares = [];
    /** @internal */ _tags = [];
    /** @internal */ _interactive = false;
    constructor(prefix) {
        this._prefix = prefix;
    }
    /**
     * Set the shared description for all tools in this router.
     *
     * @param text - Human-readable description
     * @returns `this` for chaining
     */
    describe(text) {
        this._description = text;
        return this;
    }
    use(mw) {
        this._middlewares.push(resolveMiddleware(mw));
        return this;
    }
    /**
     * Set capability tags shared by all tools in this router.
     *
     * @param tags - Tag strings for filtering
     * @returns `this` for chaining
     */
    tags(...tags) {
        this._tags.push(...tags);
        return this;
    }
    /**
     * Enable human-in-the-loop interaction for all tools in this router.
     *
     * All child tools will inherit the `.interactive()` flag, meaning
     * their handlers can use `ask()` and `ask.redirect()` directly.
     *
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * const admin = f.router('admin')
     *     .use(requireAdmin)
     *     .interactive()   // ← all children can use ask()
     * ```
     */
    interactive() {
        this._interactive = true;
        return this;
    }
    /**
     * Create a read-only query tool under this router's prefix.
     *
     * @param action - Action name (e.g. `'list'` → tool name `'prefix.list'`)
     * @returns A `FluentToolBuilder` with the prefixed name and inherited config
     */
    query(action) {
        return this._createBuilder(action, QUERY_DEFAULTS);
    }
    /**
     * Create a destructive mutation tool under this router's prefix.
     *
     * @param action - Action name (e.g. `'delete'` → tool name `'prefix.delete'`)
     * @returns A `FluentToolBuilder` with the prefixed name and inherited config
     */
    mutation(action) {
        return this._createBuilder(action, MUTATION_DEFAULTS);
    }
    /**
     * Create a neutral action tool under this router's prefix.
     *
     * @param action - Action name (e.g. `'update'` → tool name `'prefix.update'`)
     * @returns A `FluentToolBuilder` with the prefixed name and inherited config
     */
    action(action) {
        return this._createBuilder(action, ACTION_DEFAULTS);
    }
    /**
     * Create a FluentToolBuilder with inherited router config.
     * @internal
     */
    _createBuilder(action, defaults) {
        const builder = new FluentToolBuilder(`${this._prefix}.${action}`, defaults);
        // Inherit router middleware
        builder._middlewares.push(...this._middlewares);
        // Inherit router tags
        if (this._tags.length > 0) {
            builder._tags.push(...this._tags);
        }
        // Inherit router description as fallback
        if (this._description) {
            builder._description = this._description;
        }
        // Inherit interactive flag
        if (this._interactive) {
            builder._interactive = true;
        }
        return builder;
    }
}
