// @mcpfusion/core/dist/core/builder/ParamDescriptors.js
/**
 * ParamDescriptors — JSON-to-Zod Converter with TypeScript Inference
 *
 * Converts plain JSON-like param descriptors into Zod schemas internally.
 * Supports string shorthands, object descriptors with constraints, enums,
 * arrays, and examples for few-shot LLM prompting.
 *
 * @example
 * ```typescript
 * // String shorthand
 * { name: 'string' }  →  z.object({ name: z.string() })
 *
 * // Object descriptor with constraints
 * { name: { type: 'string', min: 3, max: 100, description: 'Name' } }
 *
 * // Enum
 * { status: { enum: ['active', 'archived'], optional: true } }
 *
 * // Examples for LLM few-shot
 * { cron: { type: 'string', examples: ['0 12 * * *'] } }
 * ```
 *
 * @internal
 * @module
 */
import { z } from 'zod';
// ============================================================================
// Runtime Converter — JSON descriptors → Zod schema
// ============================================================================
/** Primitive string → Zod type. Single source of truth for all type resolution. */
const PRIMITIVE_ZOD = {
    string: () => z.string(),
    number: () => z.number(),
    boolean: () => z.boolean(),
};
/**
 * Resolve a primitive type string to the corresponding Zod type.
 * @throws {Error} When the primitive type string is unknown
 */
function primitiveToZod(type, context) {
    if (!(type in PRIMITIVE_ZOD)) {
        throw new Error(`Unknown ${context} type: "${type}". Use 'string', 'number', or 'boolean'.`);
    }
    return PRIMITIVE_ZOD[type]();
}
/**
 * Finalize a Zod type by applying description and optionality.
 * DRY helper — every descriptor branch converges here.
 */
function finalize(zodType, description, optional) {
    const described = description ? zodType.describe(description) : zodType;
    return optional === true ? described.optional() : described;
}
/**
 * Build a Zod description string, appending examples if present.
 * @internal
 */
function buildDescription(base, examples) {
    if (!base && (!examples || examples.length === 0))
        return undefined;
    const parts = [];
    if (base)
        parts.push(base);
    if (examples && examples.length > 0) {
        const exStr = examples.map(e => `'${String(e)}'`).join(', ');
        parts.push(`(e.g. ${exStr})`);
    }
    return parts.join(' ');
}
// ── Per-descriptor converters ────────────────────────────
function convertEnum(def) {
    const desc = buildDescription(def.description, def.examples);
    return finalize(z.enum(def.enum), desc, def.optional);
}
function convertArray(def) {
    let arrayType = z.array(primitiveToZod(def.array, 'array item'));
    if (def.min !== undefined)
        arrayType = arrayType.min(def.min);
    if (def.max !== undefined)
        arrayType = arrayType.max(def.max);
    return finalize(arrayType, def.description, def.optional);
}
function convertString(def) {
    let s = z.string();
    if (def.min !== undefined)
        s = s.min(def.min);
    if (def.max !== undefined)
        s = s.max(def.max);
    if (def.regex !== undefined)
        s = s.regex(new RegExp(def.regex));
    const desc = buildDescription(def.description, def.examples);
    return finalize(s, desc, def.optional);
}
function convertNumber(def) {
    let n = z.number();
    if (def.int === true)
        n = n.int();
    if (def.min !== undefined)
        n = n.min(def.min);
    if (def.max !== undefined)
        n = n.max(def.max);
    const desc = buildDescription(def.description, def.examples);
    return finalize(n, desc, def.optional);
}
function convertBoolean(def) {
    return finalize(z.boolean(), def.description, def.optional);
}
// ── Main converter (dispatch table) ──────────────────────
/**
 * Convert a single param descriptor to a Zod type.
 *
 * Dispatches to focused converter functions per descriptor kind.
 * Shorthand strings ('string', 'number', 'boolean') resolve directly.
 * @internal
 */
function descriptorToZod(value) {
    // Shorthand: 'string' | 'number' | 'boolean'
    if (typeof value === 'string')
        return primitiveToZod(value, 'shorthand');
    // Discriminated union dispatch
    if ('enum' in value)
        return convertEnum(value);
    if ('array' in value)
        return convertArray(value);
    // Object descriptors: { type: 'string' | 'number' | 'boolean', ... }
    switch (value.type) {
        case 'string': return convertString(value);
        case 'number': return convertNumber(value);
        case 'boolean': return convertBoolean(value);
        default:
            throw new Error(`Unknown param type: "${value.type}".`);
    }
}
/**
 * Convert a ParamsMap to a Zod object schema.
 *
 * @param params - Plain JSON param descriptors
 * @returns A ZodObject with full validation
 *
 * @example
 * ```typescript
 * const schema = convertParamsToZod({
 *     name: 'string',
 *     status: { enum: ['active', 'archived'], optional: true },
 *     limit: { type: 'number', min: 1, max: 100, optional: true },
 * });
 * // Equivalent to:
 * // z.object({
 * //     name: z.string(),
 * //     status: z.enum(['active','archived']).optional(),
 * //     limit: z.number().min(1).max(100).optional(),
 * // })
 * ```
 *
 * @internal
 */
export function convertParamsToZod(params) {
    const shape = {};
    for (const [key, value] of Object.entries(params)) {
        shape[key] = descriptorToZod(value);
    }
    return z.object(shape);
}
