// @mcpfusion/core/dist/fsm/StateMachineGate.js
// ── Lazy XState Loading ──────────────────────────────────
/** Cached XState module reference */
// eslint-disable-next-line @typescript-eslint/consistent-type-imports -- typeof import() needed: namespace import cannot be used as a type
let xstateModule = null;
/** Number of failed import attempts (retry up to 3 times) */
let xstateLoadAttempts = 0;
const MAX_XSTATE_LOAD_ATTEMPTS = 3;
/**
 * Lazily load the `xstate` module.
 *
 * Returns `null` if `xstate` is not installed — the framework
 * degrades gracefully (all tools remain visible, no gating).
 *
 * only caches successful imports. Failed imports
 * are retried up to {@link MAX_XSTATE_LOAD_ATTEMPTS} times to
 * handle transient filesystem errors on edge/serverless cold starts.
 *
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-imports -- typeof import() needed: namespace import cannot be used as a type
async function loadXState() {
    if (xstateModule != null)
        return xstateModule; // cached success
    if (xstateLoadAttempts >= MAX_XSTATE_LOAD_ATTEMPTS)
        return null; // max retries exceeded
    xstateLoadAttempts++;
    try {
        xstateModule = await import('xstate');
        return xstateModule;
    }
    catch {
        /* xstate not installed — allow retry on next call */
        return null;
    }
}
/**
 * Reset the XState module cache so that `loadXState()` will
 * re-attempt the dynamic import on next call.
 *
 * Intended for test environments where `xstate` availability
 * may change between test suites via dynamic mocking.
 *
 * @public
 */
export function resetXStateCache() {
    xstateLoadAttempts = 0;
    xstateModule = null;
}
/**
 * Pre-load `xstate` at boot time (optional optimization).
 *
 * Call during server startup to avoid the first-use dynamic import latency.
 * Returns `true` if xstate is available, `false` otherwise.
 *
 * @example
 * ```typescript
 * import { initFsmEngine } from '@mcpfusion/core';
 * const available = await initFsmEngine();
 * if (!available) console.warn('xstate not installed — FSM gating disabled');
 * ```
 */
export async function initFsmEngine() {
    return (await loadXState()) !== null;
}
// ── StateMachineGate ─────────────────────────────────────
/**
 * Temporal Anti-Hallucination Engine.
 *
 * Wraps an XState v5 finite state machine and controls which MCP tools
 * are visible to the LLM based on the current workflow state.
 *
 * **Hard constraint**: Tools not bound to the current state are removed
 * from `tools/list` entirely — the LLM physically cannot call them.
 *
 * **Soft constraint**: `suggestActions` (HATEOAS) continues to recommend
 * the best next action within the visible set. Zero conflict.
 */
export class StateMachineGate {
    _config;
    _bindings = new Map();
    _transitionCallbacks = [];
    _actor = null;
    _currentState;
    _initialized = false;
    _initPromise = null;
    /**
     * @param config - FSM definition (states, transitions, initial state)
     */
    constructor(config) {
        this._config = config;
        this._currentState = config.initial;
    }
    // ── Initialization ───────────────────────────────────
    /**
     * Initialize the XState actor (lazy-loaded).
     *
     * Called automatically on first use. Can be called explicitly
     * at boot time for eager initialization.
     *
     * @returns `true` if XState is available and the actor started
     */
    async init() {
        if (this._initialized)
            return this._actor !== null;
        // Serialize concurrent init() calls via a shared promise.
        // Without this, two concurrent transition() calls can both enter init(),
        // creating two XState actors — the first is leaked (never stopped).
        if (this._initPromise)
            return this._initPromise;
        this._initPromise = this._doInit();
        return this._initPromise;
    }
    /** @internal */
    async _doInit() {
        const xstate = await loadXState();
        if (xstate == null) {
            this._initialized = true;
            return false;
        }
        try {
            // Use _currentState as initial — it may have been set by restore()
            // before init() was called (serverless/edge: restore → transition → init)
            const machineConfig = this._currentState !== this._config.initial
                ? { ...this._config, initial: this._currentState }
                : this._config;
            const machine = xstate.createMachine(machineConfig);
            this._actor = xstate.createActor(machine);
            this._actor.subscribe((snapshot) => {
                const newState = typeof snapshot.value === 'string'
                    ? snapshot.value
                    : Object.keys(snapshot.value)[0] ?? this._config.initial;
                if (newState !== this._currentState) {
                    this._currentState = newState;
                    for (const cb of this._transitionCallbacks) {
                        cb();
                    }
                }
            });
            this._actor.start();
            this._initialized = true;
            return true;
        }
        catch {
            /* XState machine creation failed — degrade to manual transitions */
            this._initialized = true;
            return false;
        }
    }
    // ── Tool Binding ─────────────────────────────────────
    /**
     * Bind a tool to specific FSM states.
     *
     * The tool will only appear in `tools/list` when the FSM
     * is in one of the specified states.
     *
     * @param toolName - MCP tool name (flat: `cart_add_item`, grouped: `cart`)
     * @param allowedStates - FSM states where this tool is visible
     * @param transitionEvent - Event to send on successful execution (optional)
     * @returns `this` for chaining
     *
     * @example
     * ```typescript
     * gate.bindTool('cart_add_item', ['empty', 'has_items'], 'ADD_ITEM');
     * gate.bindTool('cart_checkout', ['has_items'], 'CHECKOUT');
     * ```
     */
    bindTool(toolName, allowedStates, transitionEvent) {
        const binding = {
            allowedStates: new Set(allowedStates),
        };
        if (transitionEvent !== undefined)
            binding.transitionEvent = transitionEvent;
        this._bindings.set(toolName, binding);
        return this;
    }
    // ── State Queries ────────────────────────────────────
    /**
     * Get the current FSM state.
     */
    get currentState() {
        return this._currentState;
    }
    /**
     * Get the initial FSM state (for progressive disclosure comparison).
     *
     * Used by ServerAttachment to determine compact mode:
     * `currentState === initialState` → compact descriptions (discovery).
     * Any other state → full expert descriptions.
     */
    get initialState() {
        return this._config.initial;
    }
    /**
     * Check if a specific tool is allowed in the current FSM state.
     *
     * Tools **not** registered via `bindTool()` are always visible
     * (ungated — they don't participate in FSM gating).
     *
     * @param toolName - MCP tool name to check
     * @returns `true` if the tool should appear in `tools/list`
     */
    isToolAllowed(toolName) {
        const binding = this._bindings.get(toolName);
        if (!binding)
            return true; // Ungated tools are always visible
        return binding.allowedStates.has(this._currentState);
    }
    /**
     * Filter a list of tool names by the current FSM state.
     *
     * @param toolNames - All registered tool names
     * @returns Only the tools allowed in the current state
     */
    getVisibleToolNames(toolNames) {
        return toolNames.filter(name => this.isToolAllowed(name));
    }
    /**
     * Get the transition event for a tool (if any).
     *
     * @param toolName - MCP tool name
     * @returns The event string, or `undefined` if no transition is bound
     */
    getTransitionEvent(toolName) {
        return this._bindings.get(toolName)?.transitionEvent;
    }
    /**
     * Check if any tools are bound to this FSM gate.
     *
     * @returns `true` if at least one tool is state-gated
     */
    get hasBindings() {
        return this._bindings.size > 0;
    }
    // ── State Transitions ────────────────────────────────
    /**
     * Send an event to the FSM, potentially triggering a state transition.
     *
     * @param eventType - The event to send (e.g. `'ADD_ITEM'`, `'CHECKOUT'`)
     * @returns Result indicating whether the state changed
     */
    async transition(eventType) {
        const previousState = this._currentState;
        // Ensure initialized
        if (!this._initialized) {
            await this.init();
        }
        if (this._actor) {
            // XState actor handles the transition.
            // Bug fix: await a microtask after send() so the subscribe() callback
            // that updates _currentState is guaranteed to have run before we read it.
            // XState v5 dispatches subscribe() synchronously in most cases, but
            // this is not contractual — a microtask yield makes it safe for all versions.
            this._actor.send({ type: eventType });
            await Promise.resolve();
        }
        else {
            // Fallback: manual transition without XState
            this._transitionManual(eventType);
        }
        return {
            changed: this._currentState !== previousState,
            previousState,
            currentState: this._currentState,
        };
    }
    /**
     * Manual state transition fallback when XState is not installed.
     *
     * Reads the FSM config directly to determine the next state.
     * This provides basic FSM gating even without `xstate` installed,
     * though without XState's guards, actions, and advanced features.
     *
     * @internal
     */
    _transitionManual(eventType) {
        const stateConfig = this._config.states[this._currentState];
        if (!stateConfig?.on)
            return;
        const target = stateConfig.on[eventType];
        if (target && this._config.states[target]) {
            const previousState = this._currentState;
            this._currentState = target;
            if (previousState !== this._currentState) {
                for (const cb of this._transitionCallbacks) {
                    cb();
                }
            }
        }
    }
    // ── Transition Callbacks ─────────────────────────────
    /**
     * Register a callback that fires when the FSM state changes.
     *
     * Used by `ServerAttachment` to emit `notifications/tools/list_changed`
     * when a state transition makes tools appear or disappear.
     *
     * @param callback - Function to call on state change
     * @returns Unsubscribe function
     */
    onTransition(callback) {
        this._transitionCallbacks.push(callback);
        return () => {
            const idx = this._transitionCallbacks.indexOf(callback);
            if (idx >= 0)
                this._transitionCallbacks.splice(idx, 1);
        };
    }
    // ── Persistence (Serverless/Edge) ────────────────────
    /**
     * Create a serializable snapshot of the current FSM state.
     *
     * Used with `FsmStateStore` for serverless deployments where
     * FSM state must survive across request boundaries.
     *
     * @returns Serializable snapshot
     */
    snapshot() {
        return {
            state: this._currentState,
            updatedAt: Date.now(),
        };
    }
    /**
     * Restore FSM state from a persisted snapshot.
     *
     * Resets the XState actor (if running) so the next `transition()`
     * re-initializes the machine starting from the restored state.
     * This ensures restore → transition works correctly in
     * serverless/edge deployments (Vercel, Cloudflare Workers).
     *
     * @param snap - Previously saved snapshot
     */
    restore(snap) {
        if (this._config.states[snap.state]) {
            this._currentState = snap.state;
            // If the actor is already running, stop it and mark as
            // uninitialized so the next transition() call will re-create
            // the machine starting from the restored state.
            if (this._actor) {
                this._actor.stop();
                this._actor = null;
                this._initialized = false;
                this._initPromise = null;
            }
        }
    }
    // ── Clone (Serverless Isolation) ─────────────────────
    /**
     * Create a lightweight clone of this gate with the same config
     * and bindings but independent mutable state.
     *
     * Used in serverless/edge deployments where concurrent requests
     * must not share `_currentState`. Each request gets its own clone,
     * restores session state into it, transitions, and saves — without
     * interfering with other concurrent requests.
     *
     * The clone starts **uninitialized** (no XState actor) so the first
     * `transition()` call will create a fresh actor from the cloned state.
     *
     * @returns A new `StateMachineGate` with identical config and bindings
     */
    clone() {
        const copy = new StateMachineGate(this._config);
        for (const [toolName, binding] of this._bindings) {
            copy._bindings.set(toolName, {
                allowedStates: new Set(binding.allowedStates),
                ...(binding.transitionEvent !== undefined ? { transitionEvent: binding.transitionEvent } : {}),
            });
        }
        copy._currentState = this._currentState;
        return copy;
    }
    // ── Cleanup ──────────────────────────────────────────
    /**
     * Stop the XState actor and release resources.
     */
    dispose() {
        if (this._actor) {
            this._actor.stop();
            this._actor = null;
        }
        this._transitionCallbacks.length = 0;
    }
}
