// @mcpfusion/core/dist/cli/MarketplaceManifest.js
/**
 * MarketplaceManifest — Listing Metadata from Code
 *
 * Reads `mcpfusion.marketplace.json` from the project root and normalizes
 * the data into a snake_case payload ready for the Vinkius Cloud API.
 *
 * Features:
 *   - `file:` references — resolve local file contents (e.g. `"file:README.md"`)
 *   - i18n support — `string | Record<locale, string>` for descriptions & FAQs
 *   - Changelog extraction — parses CHANGELOG.md for latest version section
 *   - Publisher type — `official | partner | community` (admin-confirmed)
 *
 * @module
 */
import { resolve } from 'node:path';
import { existsSync, readFileSync } from 'node:fs';
// ============================================================================
// Constants
// ============================================================================
/** Default manifest filename, committed alongside code (not gitignored). */
export const MARKETPLACE_MANIFEST_FILE = 'mcpfusion.marketplace.json';
// ============================================================================
// Reader
// ============================================================================
/**
 * Read `mcpfusion.marketplace.json` from the project root.
 *
 * - Returns `null` if the file doesn't exist (marketplace is optional).
 * - Resolves all `file:` references to actual file contents.
 * - Throws on malformed JSON or unresolvable file references.
 *
 * @param cwd - Project root directory
 */
export function readMarketplaceManifest(cwd) {
    const manifestPath = resolve(cwd, MARKETPLACE_MANIFEST_FILE);
    if (!existsSync(manifestPath))
        return null;
    const raw = readFileSync(manifestPath, 'utf-8');
    const parsed = JSON.parse(raw);
    // Validate required fields
    if (!parsed.title || typeof parsed.title !== 'string') {
        throw new Error(`${MARKETPLACE_MANIFEST_FILE}: "title" is required and must be a string.`);
    }
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- runtime validation: JSON.parse output is untrusted
    if (parsed.shortDescription == null) {
        throw new Error(`${MARKETPLACE_MANIFEST_FILE}: "shortDescription" is required.`);
    }
    // Resolve file: references in i18n fields
    const resolved = { ...parsed };
    if (parsed.longDescription != null) {
        resolved['longDescription'] = resolveI18nFileRefs(parsed.longDescription, cwd);
    }
    if (parsed.changelog) {
        try {
            resolved['changelog'] = resolveFileRef(parsed.changelog, cwd);
        }
        catch {
            // Changelog is optional — skip silently when the referenced file is missing.
            delete resolved['changelog'];
        }
    }
    // Resolve file: references in FAQ answers
    if (parsed.faqs && Array.isArray(parsed.faqs)) {
        resolved['faqs'] = parsed.faqs.map(faq => ({
            question: faq.question,
            answer: resolveI18nFileRefs(faq.answer, cwd),
        }));
    }
    // Resolve file: references in prompt examples
    if (parsed.promptExamples && Array.isArray(parsed.promptExamples)) {
        resolved['promptExamples'] = parsed.promptExamples.map(ex => ({
            prompt: resolveI18nFileRefs(ex.prompt, cwd),
            response: resolveI18nFileRefs(ex.response, cwd),
        }));
    }
    // Resolve file: references in setup instructions
    if (parsed.authentication?.setupInstructions != null) {
        resolved['authentication'] = {
            ...parsed.authentication,
            setupInstructions: resolveI18nFileRefs(parsed.authentication.setupInstructions, cwd)
        };
    }
    return resolved;
}
// ============================================================================
// Normalizer (camelCase → snake_case API payload)
// ============================================================================
/**
 * Normalize a `MarketplaceManifest` into the snake_case payload
 * expected by the Vinkius Cloud `EdgeDeployController`.
 *
 * - Extracts canonical `en` value for non-i18n columns.
 * - Builds `_i18n` maps for multilingual fields.
 * - Extracts latest changelog section.
 */
export function normalizeMarketplacePayload(manifest) {
    const { canonical: shortDesc, i18n: shortDescI18n } = extractI18n(manifest.shortDescription);
    const { canonical: longDesc, i18n: longDescI18n } = manifest.longDescription != null
        ? extractI18n(manifest.longDescription)
        : { canonical: null, i18n: null };
    const payload = {
        title: manifest.title,
        short_description: shortDesc,
        long_description: longDesc,
    };
    // Only include i18n maps when they have more than just `en`
    if (shortDescI18n && Object.keys(shortDescI18n).length > 1) {
        payload['short_description_i18n'] = shortDescI18n;
    }
    if (longDescI18n && Object.keys(longDescI18n).length > 1) {
        payload['long_description_i18n'] = longDescI18n;
    }
    if (manifest.categories)
        payload['categories'] = manifest.categories;
    if (manifest.tags)
        payload['tags'] = manifest.tags;
    if (manifest.iconUrl)
        payload['icon_url'] = manifest.iconUrl;
    if (manifest.coverImageUrl)
        payload['cover_image_url'] = manifest.coverImageUrl;
    if (manifest.previewVideoUrl)
        payload['preview_video_url'] = manifest.previewVideoUrl;
    if (manifest.listingType)
        payload['listing_type'] = manifest.listingType;
    if (manifest.visibility)
        payload['visibility'] = manifest.visibility;
    if (manifest.publisherType)
        payload['publisher_type'] = manifest.publisherType;
    // Pricing
    if (manifest.pricing) {
        payload['price_cents'] = manifest.pricing.priceCents;
        payload['subscriber_request_limit'] = manifest.pricing.subscriberRequestLimit;
        if (manifest.pricing.trialRequests !== undefined) {
            payload['trial_requests'] = manifest.pricing.trialRequests;
        }
        if (manifest.pricing.maxSubscribers !== undefined) {
            payload['max_subscribers'] = manifest.pricing.maxSubscribers;
        }
    }
    // FAQs
    if (manifest.faqs && manifest.faqs.length > 0) {
        payload['faqs'] = manifest.faqs.map(faq => {
            const { canonical: q, i18n: qI18n } = extractI18n(faq.question);
            const { canonical: a, i18n: aI18n } = extractI18n(faq.answer);
            const entry = { question: q, answer: a };
            if (qI18n && Object.keys(qI18n).length > 1)
                entry['question_i18n'] = qI18n;
            if (aI18n && Object.keys(aI18n).length > 1)
                entry['answer_i18n'] = aI18n;
            return entry;
        });
    }
    // Prompt Examples
    if (manifest.promptExamples && manifest.promptExamples.length > 0) {
        payload['prompt_examples'] = manifest.promptExamples.map(ex => {
            const { canonical: p, i18n: pI18n } = extractI18n(ex.prompt);
            const { canonical: r, i18n: rI18n } = extractI18n(ex.response);
            const entry = { prompt: p, response: r };
            if (pI18n && Object.keys(pI18n).length > 1)
                entry['prompt_i18n'] = pI18n;
            if (rI18n && Object.keys(rI18n).length > 1)
                entry['response_i18n'] = rI18n;
            return entry;
        });
    }
    // Changelog — extract latest version section
    if (manifest.changelog) {
        payload['changelog_excerpt'] = extractLatestChangelog(manifest.changelog);
    }
    // SaaS integration identity
    if (manifest.integration) {
        const integration = {
            name: manifest.integration.name,
            url: manifest.integration.url,
        };
        if (manifest.integration.logoUrl) {
            integration['logo_url'] = manifest.integration.logoUrl;
        }
        if (manifest.integration.description != null) {
            const { canonical: desc } = extractI18n(manifest.integration.description);
            integration['description'] = desc;
        }
        payload['integration'] = integration;
    }
    if (manifest.documentationUrl)
        payload['documentation_url'] = manifest.documentationUrl;
    if (manifest.supportUrl)
        payload['support_url'] = manifest.supportUrl;
    if (manifest.sourceCodeUrl)
        payload['source_code_url'] = manifest.sourceCodeUrl;
    if (manifest.compatibility)
        payload['compatibility'] = manifest.compatibility;
    // Authentication
    if (manifest.authentication) {
        const auth = { ...manifest.authentication };
        if (manifest.authentication.setupInstructions != null) {
            const { canonical, i18n } = extractI18n(manifest.authentication.setupInstructions);
            auth['setup_instructions'] = canonical;
            if (i18n && Object.keys(i18n).length > 1) {
                auth['setup_instructions_i18n'] = i18n;
            }
            delete auth['setupInstructions']; // remove camelCase version
        }
        // map camelCase to snake_case
        if (auth['redirectUri'] != null) {
            auth['redirect_uri'] = auth['redirectUri'];
            delete auth['redirectUri'];
        }
        if (auth['docsUrl'] != null) {
            auth['docs_url'] = auth['docsUrl'];
            delete auth['docsUrl'];
        }
        payload['authentication'] = auth;
    }
    // SEO metadata — camelCase → snake_case
    if (manifest.seoMetadata) {
        const seo = {};
        if (manifest.seoMetadata.benefits)
            seo['benefits'] = manifest.seoMetadata.benefits;
        if (manifest.seoMetadata.targetAudience)
            seo['target_audience'] = manifest.seoMetadata.targetAudience;
        if (manifest.seoMetadata.useCases) {
            seo['use_cases'] = manifest.seoMetadata.useCases.map(uc => ({
                title: uc.title,
                description: uc.description,
            }));
        }
        if (manifest.seoMetadata.antiPatterns) {
            seo['anti_patterns'] = manifest.seoMetadata.antiPatterns.map(ap => ({
                name: ap.name,
                example: ap.example,
                fix: ap.fix,
            }));
        }
        if (manifest.seoMetadata.decisionFramework)
            seo['decision_framework'] = manifest.seoMetadata.decisionFramework;
        payload['seo_metadata'] = seo;
    }
    return payload;
}
// ============================================================================
// Changelog Parser
// ============================================================================
/**
 * Extract the most recent version section from a CHANGELOG.md string.
 *
 * Looks for `## [x.y.z]` or `## x.y.z` headings and returns everything
 * from the first heading to the start of the next heading (or EOF).
 *
 * @returns The latest changelog section, or the full text if no headings found.
 */
export function extractLatestChangelog(content) {
    const lines = content.split('\n');
    let start = -1;
    let end = lines.length;
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        // Match: ## [1.0.0], ## 1.0.0, ## [1.0.0] - 2026-03-27
        if (/^##\s+(\[?\d+\.\d+)/.test(line)) {
            if (start === -1) {
                start = i;
            }
            else {
                end = i;
                break;
            }
        }
    }
    if (start === -1) {
        // No version headings found — return first 500 chars
        return content.slice(0, 500).trim();
    }
    return lines.slice(start, end).join('\n').trim();
}
// ============================================================================
// Internals
// ============================================================================
/**
 * Resolve a single `file:` reference to file contents.
 * If the value doesn't start with `file:`, returns it unchanged.
 * @internal
 */
function resolveFileRef(value, cwd) {
    if (!value.startsWith('file:'))
        return value;
    const filePath = resolve(cwd, value.slice(5));
    if (!existsSync(filePath)) {
        throw new Error(`${MARKETPLACE_MANIFEST_FILE}: file reference "${value}" not found at ${filePath}`);
    }
    return readFileSync(filePath, 'utf-8');
}
/**
 * Resolve `file:` references in an I18nString (string or locale map).
 * @internal
 */
function resolveI18nFileRefs(value, cwd) {
    if (typeof value === 'string') {
        return resolveFileRef(value, cwd);
    }
    const resolved = {};
    for (const [locale, text] of Object.entries(value)) {
        resolved[locale] = resolveFileRef(text, cwd);
    }
    return resolved;
}
/**
 * Extract the canonical `en` value and the full i18n map from an I18nString.
 *
 * - Plain string → `{ canonical: value, i18n: null }`
 * - Locale map → `{ canonical: map.en, i18n: map }`
 *
 * @internal
 */
function extractI18n(value) {
    if (typeof value === 'string') {
        return { canonical: value, i18n: null };
    }
    const en = value['en'];
    if (!en) {
        throw new Error(`${MARKETPLACE_MANIFEST_FILE}: i18n object must include an "en" locale as canonical fallback.`);
    }
    return { canonical: en, i18n: value };
}
