// @mcpfusion/core/dist/cli/mcpfusion.js
#!/usr/bin/env node
/**
 * MCP Fusion CLI
 *
 * Slim entry point: parses args, dispatches to command modules.
 * All logic lives in focused modules under `./commands/`.
 *
 * @module
 */
import { parseArgs } from './args.js';
import { HELP } from './constants.js';
import { commandLock } from './commands/lock.js';
import { commandDev } from './commands/dev.js';
import { commandCreate } from './commands/create.js';
import { commandRemote } from './commands/remote.js';
import { commandDeploy } from './commands/deploy.js';
import { commandToken } from './commands/token.js';
import { commandVersion } from './commands/version.js';
import { commandUpdate } from './commands/update.js';
import { commandDoctor } from './commands/doctor.js';
import { commandValidate } from './commands/validate.js';
// ─── Re-exports (backward compat — tests import from mcpfusion.js) ──
export { parseArgs } from './args.js';
export { MCPFUSION_VERSION, HELP, ansi } from './constants.js';
export { ProgressTracker, createDefaultReporter } from './progress.js';
export { resolveRegistry } from './registry.js';
export { collectConfig } from './commands/create.js';
export { commandLock, commandDev, commandCreate, commandToken, commandVersion, commandUpdate, commandDoctor, commandValidate };
export { ask } from './utils.js';
// ─── Main ────────────────────────────────────────────────────────
async function main() {
    const args = parseArgs(process.argv);
    if (args.help || !args.command) {
        console.log(HELP);
        process.exit(args.help ? 0 : 1);
    }
    switch (args.command) {
        case 'create':
            await commandCreate(args);
            break;
        case 'dev':
            await commandDev(args);
            break;
        case 'lock':
            await commandLock(args);
            process.exit(0);
            break;
        case 'deploy':
            await commandDeploy(args);
            process.exit(0);
            break;
        case 'remote':
            commandRemote(args);
            break;
        case 'token':
            await commandToken(args);
            break;
        case 'inspect':
        case 'insp':
        case 'debug':
        case 'dbg': {
            const inspectArgv = process.argv.slice(3);
            try {
                const { runInspector } = await import('@mcpfusion/inspector');
                await runInspector(inspectArgv);
            }
            catch {
                console.error(`\x1b[31m\u2717\x1b[0m The inspector TUI requires the optional package:\n\n` +
                    `  npm install @mcpfusion/inspector\n`);
                process.exit(1);
            }
            break;
        }
        case 'version':
            commandVersion(args);
            break;
        case 'update':
            await commandUpdate(args);
            break;
        case 'doctor':
            await commandDoctor(args);
            break;
        case 'validate':
            await commandValidate(args);
            process.exit(0);
            break;
        case 'yaml': {
            try {
                const pkg = '@mcpfusion/yaml';
                const { commandYaml } = await import(pkg);
                await commandYaml();
            }
            catch {
                console.error(`\x1b[31m\u2717\x1b[0m The YAML engine requires the optional package:\n\n` +
                    `  npm install @mcpfusion/yaml\n`);
                process.exit(1);
            }
            break;
        }
        default:
            console.error(`Unknown command: "${args.command}"\n`);
            console.log(HELP);
            process.exit(1);
    }
}
/* c8 ignore next 9 — CLI entry-point guard */
function detectCLI() {
    if (typeof process === 'undefined' || !process.argv[1])
        return false;
    const base = process.argv[1].replace(/\\/g, '/').split('/').pop() ?? '';
    // Strip extension (.js, .cjs, .mjs, .cmd, .ps1, .exe, etc.)
    const name = base.replace(/\.[a-z0-9]+$/i, '');
    return name === 'mcpfusion';
}
if (detectCLI()) {
    main().catch((err) => {
        console.error(`Error: ${err.message}`);
        process.exit(1);
    });
}
