// @mcpfusion/core/dist/cli/progress.js
/**
 * CLI progress reporting — animated spinner, step tracking.
 *
 * Drives Composer/Yarn-style step output with a live spinner
 * that cycles while a step is running, so the user always sees
 * movement and knows the CLI hasn't frozen.
 *
 * @module
 */
// ─── Spinner ─────────────────────────────────────────────────────
const SPINNER_FRAMES = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
const SPINNER_INTERVAL_MS = 80;
// ─── Default Reporter ────────────────────────────────────────────
/** @internal exported for testing */
export function createDefaultReporter() {
    let spinnerTimer;
    let frameIndex = 0;
    let currentLine = '';
    function clearLine() {
        if (currentLine) {
            process.stderr.write(`\r\x1b[2K`);
            currentLine = '';
        }
    }
    function startSpinner(label) {
        stopSpinner();
        frameIndex = 0;
        const render = () => {
            const frame = SPINNER_FRAMES[frameIndex % SPINNER_FRAMES.length];
            currentLine = `  ${frame} ${label}...`;
            process.stderr.write(`\r\x1b[2K${currentLine}`);
            frameIndex++;
        };
        render();
        spinnerTimer = setInterval(render, SPINNER_INTERVAL_MS);
        // Don't let the spinner timer prevent process exit (important in tests)
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- defensive: setInterval returns number in browsers, object in Node.js
        if (typeof spinnerTimer === 'object' && spinnerTimer !== null && 'unref' in spinnerTimer) {
            spinnerTimer.unref();
        }
    }
    function stopSpinner() {
        if (spinnerTimer) {
            clearInterval(spinnerTimer);
            spinnerTimer = undefined;
        }
        clearLine();
    }
    return (step) => {
        switch (step.status) {
            case 'running':
                startSpinner(step.label);
                break;
            case 'done':
                stopSpinner();
                process.stderr.write(`  \x1b[32m✓\x1b[0m ${step.label}\n`);
                break;
            case 'failed': {
                stopSpinner();
                const detail = step.detail ? ` — ${step.detail}` : '';
                process.stderr.write(`  \x1b[31m✗\x1b[0m ${step.label}${detail}\n`);
                break;
            }
        }
    };
}
// ─── Tracker ─────────────────────────────────────────────────────
/** @internal exported for testing */
export class ProgressTracker {
    reporter;
    startTimes = new Map();
    constructor(reporter) {
        this.reporter = reporter ?? createDefaultReporter();
    }
    start(id, label) {
        this.startTimes.set(id, Date.now());
        this.reporter({ id, label, status: 'running' });
    }
    done(id, label, detail) {
        const durationMs = this.elapsed(id);
        this.reporter({
            id, label, status: 'done',
            ...(detail !== undefined ? { detail } : {}),
            ...(durationMs !== undefined ? { durationMs } : {}),
        });
    }
    fail(id, label, detail) {
        const durationMs = this.elapsed(id);
        this.reporter({
            id, label, status: 'failed',
            ...(detail !== undefined ? { detail } : {}),
            ...(durationMs !== undefined ? { durationMs } : {}),
        });
    }
    elapsed(id) {
        const start = this.startTimes.get(id);
        if (start === undefined)
            return undefined;
        this.startTimes.delete(id);
        return Date.now() - start;
    }
}
