// @mcpfusion/core/dist/resource/ResourceBuilder.js
/**
 * ResourceBuilder — Fluent Builder for MCP Resources
 *
 * Provides a fluent API for defining MCP Resources that expose
 * data to LLMs. Resources can optionally be subscribable, enabling
 * push notifications when data changes.
 *
 * @example
 * ```typescript
 * import { defineResource } from '@mcpfusion/core';
 *
 * const stockPrice = defineResource<AppContext>('stock_price', {
 *     uri: 'stock://prices/{symbol}',
 *     mimeType: 'application/json',
 *     description: 'Real-time stock price for a given symbol',
 *     subscribable: true,
 *     handler: async (uri, ctx) => {
 *         const symbol = uri.match(/stock:\/\/prices\/(.+)/)?.[1];
 *         const price = await ctx.stockApi.getPrice(symbol);
 *         return { text: JSON.stringify(price) };
 *     },
 * });
 * ```
 *
 * @see {@link ResourceRegistry} for registration and routing
 * @see {@link SubscriptionManager} for subscription tracking
 *
 * @module
 */
// ── Implementation ───────────────────────────────────────
class ResourceBuilderImpl {
    _name;
    _uri;
    _description;
    _mimeType;
    _tags;
    _subscribable;
    _annotations;
    _handler;
    constructor(name, config) {
        this._name = name;
        this._uri = config.uri;
        this._description = config.description;
        this._mimeType = config.mimeType ?? 'application/json';
        this._tags = config.tags ?? [];
        this._subscribable = config.subscribable ?? false;
        this._annotations = config.annotations;
        this._handler = config.handler;
    }
    getName() { return this._name; }
    getUri() { return this._uri; }
    getTags() { return this._tags; }
    isSubscribable() { return this._subscribable; }
    buildResourceDefinition() {
        const def = {
            uri: this._uri,
            name: this._name,
            ...(this._description ? { description: this._description } : {}),
            ...(this._mimeType ? { mimeType: this._mimeType } : {}),
            ...(this._annotations ? { annotations: this._annotations } : {}),
        };
        return def;
    }
    async read(uri, ctx) {
        return this._handler(uri, ctx);
    }
}
// ── Factory ──────────────────────────────────────────────
/**
 * Define a new MCP Resource.
 *
 * @param name - Unique resource name (used as identifier)
 * @param config - Resource configuration
 * @returns A resource builder for registration in {@link ResourceRegistry}
 *
 * @example
 * ```typescript
 * const deployStatus = defineResource<AppContext>('deploy_status', {
 *     uri: 'deploy://status/{environment}',
 *     description: 'Real-time deploy pipeline status',
 *     subscribable: true,
 *     handler: async (uri, ctx) => {
 *         const env = uri.match(/deploy:\/\/status\/(.+)/)?.[1];
 *         const status = await ctx.deployService.getStatus(env);
 *         return { text: JSON.stringify(status) };
 *     },
 * });
 * ```
 */
export function defineResource(name, config) {
    return new ResourceBuilderImpl(name, config);
}
