// @mcpfusion/core/dist/resource/SubscriptionManager.js
/**
 * SubscriptionManager — Push Notification Tracking
 *
 * Tracks which resource URIs are subscribed and routes
 * `notifications/resources/updated` to the MCP transport layer.
 *
 * Designed to be embedded within {@link ResourceRegistry}.
 * Supports both in-process usage (framework) and external
 * delegation (runtime Redis bridge).
 *
 * @see {@link ResourceRegistry} for the public API
 *
 * @module
 */
// ── Manager ──────────────────────────────────────────────
export class SubscriptionManager {
    _subscribed = new Set();
    _sink;
    /**
     * Subscribe to push notifications for a resource URI.
     *
     * @param uri - The resource URI to subscribe to
     */
    subscribe(uri) {
        this._subscribed.add(uri);
    }
    /**
     * Unsubscribe from push notifications for a resource URI.
     *
     * @param uri - The resource URI to unsubscribe from
     */
    unsubscribe(uri) {
        this._subscribed.delete(uri);
    }
    /**
     * Check if a URI is currently subscribed.
     *
     * @param uri - The URI to check
     */
    isSubscribed(uri) {
        return this._subscribed.has(uri);
    }
    /**
     * Get all currently subscribed URIs.
     */
    getSubscriptions() {
        return this._subscribed;
    }
    /**
     * Emit a `notifications/resources/updated` for a URI.
     *
     * Only emits if the URI is subscribed AND a sink is configured.
     * Errors in the sink are swallowed (best-effort delivery).
     *
     * @param uri - The URI of the resource that changed
     */
    async notify(uri) {
        if (!this._sink || !this._subscribed.has(uri))
            return;
        try {
            const result = this._sink(uri);
            if (result instanceof Promise) {
                await result;
            }
        }
        catch {
            /* best-effort — sink must not break the pipeline */
        }
    }
    /**
     * Set the notification sink.
     *
     * @param sink - Callback that emits the MCP notification
     * @internal
     */
    setSink(sink) {
        this._sink = sink;
    }
    /** Number of active subscriptions. */
    get size() { return this._subscribed.size; }
    /** Remove all subscriptions. */
    clear() {
        this._subscribed.clear();
    }
}
