// @mcpfusion/core/dist/server/DevServer.js
/**
 * DevServer — HMR-enabled MCP Development Server
 *
 * The "killer feature" for MCP DX: `mcpfusion dev` starts a
 * development server that watches for file changes and performs
 * automatic Hot Module Replacement without requiring the LLM client
 * (Claude Desktop, Cursor, etc.) to be restarted.
 *
 * ## How It Works
 *
 * 1. The dev server starts an MCP stdio server with your tools
 * 2. When a `.ts`/`.js` file changes, the module cache is invalidated
 * 3. Tools are re-registered from the updated modules
 * 4. The MCP `notifications/tools/list_changed` notification is sent
 * 5. The LLM client picks up the new tool definitions transparently
 *
 * ## Usage
 *
 * ```bash
 * # Auto-detects src/server.ts
 * mcpfusion dev
 *
 * # Explicit entrypoint + custom watch dir
 * mcpfusion dev --server ./src/server.ts --dir ./src/tools
 * ```
 *
 * @example
 * ```typescript
 * // Using the programmatic API
 * import { createDevServer } from '@mcpfusion/core/dev';
 *
 * const devServer = createDevServer({
 *   dir: './src/tools',
 *   setup: async (registry) => {
 *     // Your setup logic — called on every reload
 *     await autoDiscover(registry, './src/tools');
 *   },
 * });
 *
 * await devServer.start();
 * ```
 *
 * @module
 */
import { watch } from 'node:fs';
import { resolve, join, relative } from 'node:path';
import { toErrorMessage } from '../core/ErrorUtils.js';
import { pathToFileURL } from 'node:url';
// ── Module Cache Invalidation ────────────────────────────
/**
 * Invalidate Node.js ESM module cache for a given file.
 *
 * ESM modules are cached by URL. We use an import timestamp trick
 * to force re-evaluation on next import.
 *
 * For CJS, we clear `require.cache` if available.
 *
 * @internal
 */
function invalidateModule(filePath) {
    const absolutePath = resolve(filePath);
    // CJS cache invalidation (when running in CJS mode)
    if (typeof require !== 'undefined') {
        invalidateCjsTree(absolutePath);
    }
    // ESM modules can't be uncached directly — use a cache-busting
    // query parameter so the next dynamic import() re-evaluates the module.
    lastCacheBustUrl = cacheBustUrl(absolutePath);
}
/**
 * The last cache-busted URL produced by `invalidateModule()`.
 * The setup callback can use this to re-import an ESM module that was invalidated.
 * @internal
 */
let lastCacheBustUrl;
/**
 * Recursively invalidate a CJS module and all modules that depend on it.
 *
 * Walks `require.cache` to find modules whose `children` include the
 * changed module (i.e., importers / dependents), and removes them so
 * the next `require()` re-evaluates the entire dependency chain.
 *
 * @internal
 */
function invalidateCjsTree(entryPath) {
    const visited = new Set();
    const queue = [entryPath];
    while (queue.length > 0) {
        const current = queue.pop();
        if (visited.has(current))
            continue;
        visited.add(current);
        delete require.cache[current];
        // Find all cached modules that imported `current`
        for (const [key, mod] of Object.entries(require.cache)) {
            if (!mod)
                continue;
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- require.cache module type lacks .children in @types/node
            const children = mod.children ?? [];
            if (children.some(c => c.id === current)) {
                queue.push(key);
            }
        }
    }
}
/**
 * Create a cache-busting import URL for ESM modules.
 * Appends a timestamp query to force the module to be re-evaluated.
 *
 * Use this inside a DevServer `setup` callback to re-import ESM modules:
 * ```ts
 * const mod = await import(cacheBustUrl('./src/tools.ts'));
 * ```
 */
export function cacheBustUrl(filePath) {
    const url = pathToFileURL(resolve(filePath));
    url.searchParams.set('t', String(Date.now()));
    return url.href;
}
// ── Dev Server Factory ───────────────────────────────────
/**
 * Create an HMR-enabled MCP development server.
 *
 * Watches a directory for file changes and automatically reloads
 * tools, then notifies the connected MCP client via the native
 * `notifications/tools/list_changed` notification.
 *
 * @param config - Dev server configuration
 * @returns A {@link DevServer} instance with start/stop/reload controls
 *
 * @example
 * ```typescript
 * import { createDevServer, autoDiscover, ToolRegistry } from '@mcpfusion/core';
 *
 * const devServer = createDevServer({
 *   dir: './src/tools',
 *   setup: async (registry) => {
 *     await autoDiscover(registry, './src/tools');
 *   },
 *   onReload: (file) => console.log(`[HMR] Reloaded: ${file}`),
 * });
 *
 * await devServer.start();
 * // File changes → auto-reload → LLM client gets notification
 * ```
 */
export function createDevServer(config) {
    const { dir, extensions = ['.ts', '.js', '.mjs', '.mts'], debounce = 300, setup, onReload, server, } = config;
    const absoluteDir = resolve(dir);
    let watcher;
    let debounceTimer;
    let reloadCount = 0;
    /** Whether we've warned about ESM cache-bust already */
    let esmWarningEmitted = false;
    /**
     * Perform a full reload: clear caches, re-run setup, notify MCP client.
     */
    async function performReload(changedFile) {
        reloadCount++;
        // Invalidate CJS cache for all watched files
        invalidateModule(changedFile);
        // Capture the cache-busted URL before calling setup
        const currentCacheBustUrl = lastCacheBustUrl;
        // Re-run the user's setup callback
        // Provide a duck-typed registry that satisfies the interface.
        // The user's setup may use this or close over their own registry.
        const builders = [];
        const reloadRegistry = {
            register(builder) { builders.push(builder); },
            getBuilders() { return builders; },
        };
        // Pass both registry and cacheBustUrl to the setup callback.
        // The setup function can destructure the context or treat it
        // as a plain registry (backward-compatible duck typing).
        const setupCtx = {
            registry: reloadRegistry,
            cacheBustUrl: currentCacheBustUrl,
            // Duck-type compatibility: expose register/getBuilders at top level
            // so legacy callbacks that expect (registry) still work.
        };
        // Assign duck-type methods so old `setup(registry)` signatures work
        const ctxWithDuck = Object.assign(setupCtx, {
            register: reloadRegistry.register.bind(reloadRegistry),
            getBuilders: reloadRegistry.getBuilders.bind(reloadRegistry),
        });
        try {
            await setup(ctxWithDuck);
        }
        catch (err) {
            const message = toErrorMessage(err);
            console.error(`[mcpfusion dev] Reload failed: ${message}`);
            return;
        }
        // Warn once if ESM modules may not benefit from cache invalidation
        if (!esmWarningEmitted && changedFile !== '(initial)' && changedFile !== '(manual)') {
            const isEsm = extensions.some(ext => ext === '.mjs' || ext === '.mts') ||
                changedFile.endsWith('.mjs') || changedFile.endsWith('.mts') ||
                extensions.includes('.ts') || extensions.includes('.js');
            if (isEsm) {
                console.warn('[mcpfusion dev] ESM hot-reload requires cache-busted imports. ' +
                    'Use the `cacheBustUrl` from the setup context or `cacheBustUrl()` from \'@mcpfusion/core\' ' +
                    'when dynamically importing modules. Plain `import(\'./path\')` will use stale cache.');
                esmWarningEmitted = true;
            }
        }
        // Transfer collected builders to the real registry ( fix)
        if (builders.length > 0 && config.registry) {
            if (typeof config.registry.clear === 'function') {
                config.registry.clear();
            }
            for (const b of builders) {
                config.registry.register(b);
            }
        }
        // Notify MCP client about tool list changes
        if (server) {
            const notification = { method: 'notifications/tools/list_changed' };
            try {
                if (typeof server.sendNotification === 'function') {
                    await server.sendNotification(notification);
                }
                else if (typeof server.notification === 'function') {
                    await server.notification(notification);
                }
            }
            catch {
                // Connection might not be established yet — ignore
            }
        }
        // User callback
        onReload?.(changedFile);
    }
    return {
        async start() {
            // Perform initial load
            await performReload('(initial)');
            // Start watching
            const watchOptions = { recursive: true };
            // warn on Linux with Node.js < 20 where recursive watch is unsupported
            if (process.platform === 'linux' && parseInt(process.versions.node, 10) < 20) {
                console.warn('[mcpfusion dev] Warning: recursive fs.watch() is not fully supported on Linux with Node.js < 20. ' +
                    'File changes in subdirectories may not be detected. Consider upgrading to Node.js 20+.');
            }
            watcher = watch(absoluteDir, watchOptions, (_eventType, filename) => {
                if (!filename)
                    return;
                // Filter by extension
                const ext = '.' + filename.split('.').pop();
                if (!extensions.includes(ext))
                    return;
                // Skip test/spec files
                if (/\.(test|spec|d)\./.test(filename))
                    return;
                // Debounce rapid changes
                if (debounceTimer)
                    clearTimeout(debounceTimer);
                debounceTimer = setTimeout(() => {
                    const fullPath = join(absoluteDir, filename);
                    void performReload(relative(process.cwd(), fullPath));
                }, debounce);
            });
            watcher.on('error', (err) => {
                console.error(`[mcpfusion dev] Watcher error: ${toErrorMessage(err)}`);
            });
            console.log(`[mcpfusion dev] Watching ${relative(process.cwd(), absoluteDir)} for changes...`);
        },
        stop() {
            if (debounceTimer)
                clearTimeout(debounceTimer);
            watcher?.close();
            watcher = undefined;
            console.log(`[mcpfusion dev] Stopped. ${reloadCount} reload(s) performed.`);
        },
        async reload(reason) {
            await performReload(reason ?? '(manual)');
        },
    };
}
