// @mcpfusion/core/dist/server/ServerAttachment.js
/**
 * ServerAttachment — MCP Server Integration Strategy
 *
 * Handles attaching a ToolRegistry to an MCP Server by registering
 * request handlers for tools/list and tools/call.
 *
 * Supports both Server (low-level) and McpServer (high-level) via duck-typing.
 *
 * Pure-function module: receives dependencies, returns detach function.
 */
import { ListToolsRequestSchema, CallToolRequestSchema, ListPromptsRequestSchema, GetPromptRequestSchema, ListResourcesRequestSchema, ReadResourceRequestSchema, SubscribeRequestSchema, UnsubscribeRequestSchema, } from '@modelcontextprotocol/sdk/types.js';
import {} from '@modelcontextprotocol/sdk/types.js';
import { error, toolError, isHandoffResponse } from '../core/response.js';
import {} from '../core/types.js';
import {} from '../core/execution/ProgressHelper.js';
import { resolveServer } from './ServerResolver.js';
import {} from '../observability/DebugObserver.js';
import {} from '../observability/Tracing.js';
import { StateSyncLayer } from '../state-sync/StateSyncLayer.js';
import {} from '../state-sync/types.js';
import {} from '../introspection/types.js';
import { registerIntrospectionResource } from '../introspection/IntrospectionResource.js';
import { compileManifest, cloneManifest } from '../introspection/ManifestCompiler.js';
import { AttestationError } from '../introspection/CryptoAttestation.js';
import { enrichValidationError } from '../introspection/ContractAwareSelfHealing.js';
import { compileContracts } from '../introspection/ToolContract.js';
import { computeServerDigest } from '../introspection/BehaviorDigest.js';
import {} from '../exposition/types.js';
import { compileExposition } from '../exposition/ExpositionCompiler.js';
import {} from '../resource/ResourceRegistry.js';
import {} from '../prompt/PromptRegistry.js';
import {} from '../prompt/types.js';
import { randomUUID } from 'node:crypto';
import { _elicitStore } from '../core/elicitation/index.js';
// ── Observability Propagation ────────────────────────────
/**
 * Propagate debug and tracing observers to all registered builders.
 * Zero overhead when neither is configured.
 */
function propagateObservability(registry, debug, tracing, telemetry) {
    if (debug && registry.enableDebug) {
        registry.enableDebug(debug);
    }
    if (tracing && registry.enableTracing) {
        registry.enableTracing(tracing);
    }
    if (telemetry && registry.enableTelemetry) {
        registry.enableTelemetry(telemetry);
    }
}
// ── Missing Context Guard ────────────────────────────────
/**
 * Proxy sentinel used when `contextFactory` is not provided.
 *
 * Instead of `undefined` (which causes cryptic `TypeError: Cannot read
 * properties of undefined`), this proxy throws a clear, actionable error
 * the moment a handler accesses any property on `ctx`.
 *
 * For `void` contexts where handlers never touch `ctx`, the proxy is
 * never triggered — zero false positives.
 *
 * @internal — exported for reuse by `startServer.ts` edge handler.
 */
export const _missingContextProxy = new Proxy(Object.freeze({}), {
    get(_target, prop) {
        // Allow symbol access (e.g. Symbol.toPrimitive, Symbol.toStringTag) and
        // JSON.stringify probing ('toJSON') to avoid breaking framework internals.
        if (typeof prop === 'symbol')
            return undefined;
        throw new Error(`[mcpfusion] Attempted to access "ctx.${String(prop)}" but no contextFactory was provided. ` +
            `Add contextFactory to your attachToServer() options:\n\n` +
            `  registry.attachToServer(server, {\n` +
            `      contextFactory: (extra) => createAppContext(extra),\n` +
            `  });\n`);
    },
});
// ── FSM Session Helpers ──────────────────────────────────
/**
 * Resolve the session ID from the MCP extra argument,
 * falling back to the handler context's stable ID for
 * session-less transports (e.g. stdio).
 *
 * Extracted to eliminate 4× repetition of the same expression.
 */
function resolveSessionId(extra, hCtx) {
    return extractSessionId(extra) ?? hCtx.fallbackSessionId;
}
/**
 * Clone the FSM and restore the session-specific snapshot.
 *
 * Shared by both `tools/list` and `tools/call` handlers.
 * Eliminates the duplicated clone+restore logic ( + ).
 */
async function cloneAndRestoreFsm(hCtx, extra) {
    let fsm = hCtx.fsm;
    if (!fsm)
        return undefined;
    fsm = fsm.clone();
    const sessionId = resolveSessionId(extra, hCtx);
    if (hCtx.fsmStore) {
        const snap = await hCtx.fsmStore.load(sessionId);
        if (snap)
            fsm.restore(snap);
    }
    else {
        const snap = hCtx.fsmMemorySnapshots?.get(sessionId);
        if (snap)
            fsm.restore(snap);
    }
    return fsm;
}
// ── Handler Factories ────────────────────────────────────
/**
 * Create the `tools/list` request handler.
 *
 * In flat mode, re-compiles exposition from the current registry state.
 * In grouped mode, delegates to the registry's tag-filtered listing.
 */
function createToolListHandler(hCtx) {
    return async (_request, extra) => {
        // Per-request FSM clone for serverless isolation ( +  fix).
        const fsm = await cloneAndRestoreFsm(hCtx, extra);
        // FHP: if a handoff tunnel is active, proxy the upstream's tools list
        if (hCtx.swarmGateway) {
            const sessionId = resolveSessionId(extra, hCtx);
            const proxied = await hCtx.swarmGateway.proxyToolsList(sessionId);
            if (proxied != null) {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any -- McpTool shape
                return { tools: proxied };
            }
        }
        let tools;
        if (hCtx.isFlat) {
            // FSM Progressive Disclosure: determine compact mode from FSM state.
            // Initial state = compact descriptions. After first call triggers
            // transition, FSM moves to a non-initial state = full expert descriptions.
            // listChanged notification forces client re-fetch with new descriptions.
            const fsmCompactMode = fsm ? fsm.currentState === fsm.initialState : false;
            const exposition = hCtx.recompile(fsmCompactMode);
            tools = hCtx.filter
                ? filterFlatTools(exposition.tools, exposition.routingMap, hCtx.filter)
                : exposition.tools;
        }
        else {
            tools = hCtx.filter
                ? hCtx.registry.getTools(hCtx.filter)
                : hCtx.registry.getAllTools();
        }
        // FSM State Gate: remove tools not allowed in the current state
        if (fsm && fsm.hasBindings) {
            tools = tools.filter(tool => fsm.isToolAllowed(tool.name));
        }
        return { tools: hCtx.syncLayer ? hCtx.syncLayer.decorateTools(tools) : tools };
    };
}
/**
 * Create the `tools/call` request handler.
 *
 * Handles both flat (O(1) dispatch) and grouped (registry routing) modes.
 * Wires progress notifications when the client opts in via `_meta.progressToken`.
 */
function createToolCallHandler(hCtx) {
    return async (request, extra) => {
        const { name, arguments: args = {} } = request.params;
        const ctx = hCtx.contextFactory
            ? await hCtx.contextFactory(extra)
            : _missingContextProxy;
        const progressSink = createProgressSink(extra);
        const signal = extractSignal(extra);
        const elicitSink = extractElicitSink(extra);
        const emit = hCtx.telemetry;
        // ── Telemetry: route event ──────────────────────────
        // Resolve group/action from the routing map instead of naive
        // split('_') — avoids misattributing tools with underscores
        // in their names (e.g. 'user_accounts_list' → group='user_accounts', action='list').
        //
        // Lazy evaluation: recompile() is deferred until the first consumer
        // needs it (telemetry, FSM gate, or routing). Zero overhead when
        // none of these features are active.
        let _cachedExposition;
        const getExposition = () => {
            if (!hCtx.isFlat)
                return undefined;
            if (!_cachedExposition)
                _cachedExposition = hCtx.recompile();
            return _cachedExposition;
        };
        const exposition = getExposition();
        const flatRoute = exposition?.routingMap.get(name);
        const toolGroup = flatRoute ? flatRoute.builder.getName() : name;
        const action = flatRoute ? flatRoute.actionKey : name;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any -- TelemetrySink accepts extensible event shapes
        emit?.({ type: 'route', tool: toolGroup, action, args, timestamp: Date.now() });
        // Per-request FSM clone for serverless isolation ( +  fix).
        const fsm = await cloneAndRestoreFsm(hCtx, extra);
        // Resolve the canonical FSM tool name:
        // - Flat mode: `name` is already the action-qualified key
        // - Grouped mode: compose `group.action` to match FSM bindings
        //   (grouped `name` is the builder name, e.g. 'billing', not the
        //    action-qualified key 'billing.get_invoice')
        const fsmToolName = flatRoute ? name : `${toolGroup}.${action}`;
        // enforce FSM gate on tools/call — not just tools/list.
        // Without this, a client that knows a tool's name can bypass the gate.
        if (fsm && fsm.hasBindings && !fsm.isToolAllowed(fsmToolName)) {
            return toolError('FORBIDDEN', {
                message: `Tool "${name}" is not available in the current FSM state ("${fsm.currentState}").`,
                suggestion: 'This tool is gated by the FSM State Gate. Call an allowed tool to advance the state first.',
                availableActions: fsm.getVisibleToolNames([...new Set((exposition?.tools ?? hCtx.registry.getAllTools()).map(t => t.name))]),
                severity: 'error',
                details: { currentState: fsm.currentState, blockedTool: name },
            });
        }
        // ── FHP: proxy tools/call when handoff tunnel is active ──────────────────
        // This check MUST happen before the local tool execution block below.
        // When a session has an active handoff, the LLM will call upstream tools
        // (e.g. 'finance.refund') that do not exist in the gateway registry.
        // Checking for an active handoff FIRST short-circuits the local execution
        // entirely, giving clean telemetry and avoiding the unnecessary registry lookup.
        if (hCtx.swarmGateway) {
            const sessionId = resolveSessionId(extra, hCtx);
            if (hCtx.swarmGateway.hasActiveHandoff(sessionId)) {
                // Handle return_to_triage
                if (name.endsWith('.return_to_triage')) {
                    await hCtx.swarmGateway.returnToGateway(sessionId);
                    hCtx.notifyToolListChanged?.();
                    return { content: [{ type: 'text', text: '[RETURN] Specialised session ended. Gateway tools restored.' }] };
                }
                const callSignal = extractSignal(extra) ?? new AbortController().signal;
                const proxied = await hCtx.swarmGateway.proxyToolsCall(sessionId, name, args, callSignal);
                if (proxied !== null)
                    return proxied;
                // proxied === null means the gateway deferred (e.g. unknown tool in upstream);
                // fall through to local execution so gateway-native tools still work.
            }
        }
        let result;
        const t0 = Date.now();
        // Core execution — optionally wrapped in elicitation context
        // so ask() can resolve its transport via AsyncLocalStorage.
        const executeLocal = async () => {
            try {
                if (hCtx.isFlat) {
                    // Reuse exposition compiled above for telemetry (avoid double recompile)
                    if (flatRoute) {
                        const enrichedArgs = { ...args, [flatRoute.discriminator]: flatRoute.actionKey };
                        let r = await flatRoute.builder.execute(ctx, enrichedArgs, progressSink, signal);
                        r = decorateIfSync(hCtx.syncLayer, flatRoute, r);
                        return r;
                    }
                    else {
                        let r = await hCtx.registry.routeCall(ctx, name, args, progressSink, signal);
                        r = hCtx.syncLayer ? hCtx.syncLayer.decorateResult(name, r) : r;
                        return r;
                    }
                }
                else {
                    let r = await hCtx.registry.routeCall(ctx, name, args, progressSink, signal);
                    r = hCtx.syncLayer ? hCtx.syncLayer.decorateResult(name, r) : r;
                    return r;
                }
            }
            catch (err) {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any -- TelemetrySink accepts extensible event shapes
                emit?.({ type: 'error', tool: toolGroup, action, error: String(err), timestamp: Date.now() });
                throw err;
            }
        };
        // Bind elicitation transport via AsyncLocalStorage.
        // Zero overhead when elicitSink is undefined — no ALS context created.
        result = elicitSink
            ? await _elicitStore.run(elicitSink, executeLocal)
            : await executeLocal();
        // ── Self-Healing: enrich validation errors with contract deltas ( fix) ──
        if (result.isError && hCtx.selfHealing) {
            const firstContent = result.content[0];
            const text = firstContent?.type === 'text' ? firstContent.text : '';
            if (text) {
                const healing = enrichValidationError(text, toolGroup, action, hCtx.selfHealing);
                if (healing.injected) {
                    result = { ...result, content: [{ type: 'text', text: healing.enrichedError }] };
                }
            }
        }
        // ── FHP: detect HandoffResponse and activate SwarmGateway tunnel ──────────
        if (isHandoffResponse(result) && hCtx.swarmGateway) {
            const sessionId = resolveSessionId(extra, hCtx);
            // Reuse the signal already extracted at the top of this handler (line ~656).
            // Using a new name avoids shadowing the outer `signal` variable.
            const handoffSignal = signal ?? new AbortController().signal;
            const gateway = hCtx.swarmGateway;
            const payload = result.payload;
            // Activate tunnel asynchronously — ACK is returned immediately to the LLM
            void gateway.activateHandoff(payload, sessionId, handoffSignal)
                .then(() => hCtx.notifyToolListChanged?.())
                .catch(() => hCtx.notifyToolListChanged?.());
            // Cognitive anchor: visible ACK prevents LLM anxiety loop while tools reload.
            // Uses HANDOFF_CONNECTING (not HANDOFF_UPSTREAM_UNAVAILABLE): the latter
            // signals a failure to the LLM, whereas here the upstream is actively connecting.
            return toolError('HANDOFF_CONNECTING', {
                message: `[HANDOFF] ${payload.reason ?? 'Specialist selected'}. Your tools are being updated — please wait.`,
                suggestion: 'Wait for the tools list to refresh, then proceed with the specialised tools.',
                severity: 'warning',
            });
        }
        // ── Telemetry: execute event ─────────────────────────
        // Single Date.now() capture ensures durationMs and timestamp are coherent.
        const t1 = Date.now();
        emit?.({
            type: 'execute', tool: toolGroup, action,
            durationMs: t1 - t0,
            isError: !!result.isError,
            timestamp: t1,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- TelemetrySink accepts extensible event shapes
        });
        // FSM State Gate: auto-transition on successful execution
        if (fsm && !result.isError) {
            const transitionEvent = fsm.getTransitionEvent(fsmToolName);
            if (transitionEvent) {
                const fromState = fsm.currentState;
                const transition = await fsm.transition(transitionEvent);
                if (transition.changed) {
                    // Emit fsm.transition telemetry event
                    if (hCtx.telemetry) {
                        hCtx.telemetry({
                            type: 'fsm.transition',
                            tool: name,
                            action: transitionEvent,
                            from: fromState,
                            to: fsm.currentState,
                            timestamp: Date.now(),
                            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- TelemetrySink accepts extensible event shapes
                        });
                    }
                    // Persist new state to external store (serverless/edge)
                    // Use fallback session ID for transports without sessions (e.g., stdio) ( fix)
                    const sessionId = resolveSessionId(extra, hCtx);
                    if (hCtx.fsmStore) {
                        await hCtx.fsmStore.save(sessionId, fsm.snapshot());
                    }
                    else if (hCtx.fsmMemorySnapshots) {
                        // persist to in-memory session map
                        hCtx.fsmMemorySnapshots.set(sessionId, fsm.snapshot());
                    }
                    // Notify client to re-fetch tools/list
                    hCtx.notifyToolListChanged?.();
                }
            }
        }
        return result;
    };
}
/**
 * Decorate a flat-route result with state-sync metadata when applicable.
 * Uses the canonical dot-notation key for policy matching.
 */
function decorateIfSync(syncLayer, flatRoute, result) {
    if (!syncLayer)
        return result;
    const canonicalKey = `${flatRoute.builder.getName()}.${flatRoute.actionKey}`;
    return syncLayer.decorateResult(canonicalKey, result);
}
/**
 * Register `prompts/list` and `prompts/get` handlers on the server.
 *
 * Wires the prompt lifecycle notification sink and the internal
 * loopback dispatcher that allows prompts to invoke tools in-memory.
 */
function registerPromptHandlers(resolved, server, prompts, registry, filter, contextFactory) {
    // Wire lifecycle sync
    const serverAny = server;
    const sendFn = serverAny['sendPromptListChanged'];
    if (typeof sendFn === 'function') {
        prompts.setNotificationSink(() => { sendFn.call(server); });
    }
    // prompts/list
    resolved.setRequestHandler(ListPromptsRequestSchema, async (request) => {
        const params = {};
        if (filter)
            params.filter = filter;
        if (request.params?.cursor)
            params.cursor = request.params.cursor;
        return await prompts.listPrompts(params);
    });
    // prompts/get — with loopback dispatcher and signal propagation
    resolved.setRequestHandler(GetPromptRequestSchema, async (request, extra) => {
        const { name, arguments: args = {} } = request.params;
        const ctx = contextFactory
            ? await contextFactory(extra)
            : _missingContextProxy;
        const signal = extractSignal(extra);
        const enrichedCtx = injectLoopbackDispatcher(ctx, registry, signal);
        return prompts.routeGet(enrichedCtx, name, args);
    });
}
/**
 * Inject `invokeTool()` into the context so prompt handlers can call
 * tools in-memory. Runs the Tool's full pipeline with RBAC enforced.
 * Propagates the cancellation signal from the parent request.
 */
function injectLoopbackDispatcher(ctx, registry, signal) {
    // Protect the original context from mutation — use prototype-based proxy
    // for object contexts (safe — no property copy). For non-object contexts
    // (primitives or null), start with an empty wrapper.
    let wrapped;
    if (ctx != null && typeof ctx === 'object') {
        wrapped = Object.create(ctx);
    }
    else {
        // ctx is null, undefined, or a primitive — start fresh.
        // No prototype pollution risk here: no properties to copy.
        wrapped = {};
    }
    wrapped['invokeTool'] = async (toolName, toolArgs = {}) => {
        const response = await registry.routeCall(wrapped, toolName, toolArgs, undefined, signal);
        const text = response.content
            .filter((c) => c.type === 'text')
            .map((c) => c.text)
            .join('\n');
        return {
            text,
            isError: response.isError ?? false,
            raw: response,
        };
    };
    return wrapped;
}
/**
 * Register `resources/list`, `resources/read`, `resources/subscribe`,
 * and `resources/unsubscribe` handlers on the MCP server.
 *
 * Wires the ResourceRegistry notification sink to the MCP server's
 * notification method for push delivery via SSE/Streamable HTTP.
 *
 * @internal
 */
function registerResourceHandlers(resolved, server, resources, contextFactory, introspection) {
    const resourceServer = resolved;
    // Pre-compute introspection manifest URI for merge.
    const manifestUri = introspection?.config.uri ?? 'mcpfusion://manifest.json';
    // Wire notification sink for `notifications/resources/updated`
    const serverAny = server;
    const sendResourceUpdated = serverAny['sendResourceUpdated'];
    const sendNotification = serverAny['notification'];
    if (typeof sendResourceUpdated === 'function') {
        resources.setNotificationSink((uri) => {
            sendResourceUpdated.call(server, uri);
        });
    }
    else if (typeof sendNotification === 'function') {
        resources.setNotificationSink((uri) => {
            void sendNotification.call(server, {
                method: 'notifications/resources/updated',
                params: { uri },
            });
        });
    }
    // Wire lifecycle sync for `notifications/resources/list_changed`
    const sendListChanged = serverAny['sendResourceListChanged'];
    if (typeof sendListChanged === 'function') {
        resources.setListChangedSink(() => { sendListChanged.call(server); });
    }
    // resources/list — merge with introspection resources if present
    // The handler is a per-request arrow function (not an IIFE) so that
    // resources registered or removed dynamically after attachToServer() are
    // always reflected in the response.
    resourceServer.setRequestHandler(ListResourcesRequestSchema, (() => {
        return () => {
            const list = resources.listResources();
            if (introspection) {
                list.push({
                    uri: manifestUri,
                    name: 'mcpfusion:manifest',
                    description: 'Dynamic introspection manifest exposing all registered tools, actions, and presenters. RBAC-filtered per session context.',
                    mimeType: 'application/json',
                });
            }
            return { resources: list };
        };
    })());
    // resources/read — with introspection manifest delegation ( fix)
    resourceServer.setRequestHandler(ReadResourceRequestSchema, (async (request, extra) => {
        // Handle introspection manifest URI before ResourceRegistry
        if (introspection && request.params.uri === manifestUri) {
            const fullManifest = compileManifest(introspection.serverName, introspection.builders.values());
            let manifest = fullManifest;
            if (introspection.config.filter && contextFactory) {
                const ctx = await contextFactory(extra);
                manifest = introspection.config.filter(cloneManifest(fullManifest), ctx);
            }
            return {
                contents: [{
                        uri: manifestUri,
                        mimeType: 'application/json',
                        text: JSON.stringify(manifest, null, 2),
                    }],
            };
        }
        const ctx = contextFactory
            ? await contextFactory(extra)
            : _missingContextProxy;
        return resources.readResource(request.params.uri, ctx);
    }));
    // resources/subscribe
    resourceServer.setRequestHandler(SubscribeRequestSchema, ((request) => {
        const accepted = resources.subscribe(request.params.uri);
        if (!accepted) {
            return {
                _meta: {},
            };
        }
        return { _meta: {} };
    }));
    // resources/unsubscribe
    resourceServer.setRequestHandler(UnsubscribeRequestSchema, ((request) => {
        resources.unsubscribe(request.params.uri);
        return { _meta: {} };
    }));
}
/**
 * Create the detach function that replaces all handlers with no-ops.
 */
function createDetachFn(resolved, hasPrompts) {
    return () => {
        resolved.setRequestHandler(ListToolsRequestSchema, () => ({ tools: [] }));
        resolved.setRequestHandler(CallToolRequestSchema, () => error('Tool handlers have been detached'));
        if (hasPrompts) {
            resolved.setRequestHandler(ListPromptsRequestSchema, () => ({ prompts: [] }));
            resolved.setRequestHandler(GetPromptRequestSchema, () => ({
                messages: [{ role: 'user', content: { type: 'text', text: 'Prompt handlers have been detached' } }],
            }));
        }
    };
}
// ── Public API ───────────────────────────────────────────
/**
 * Attach a registry to an MCP Server.
 *
 * Resolves the server type, registers tools/list and tools/call handlers,
 * and returns a detach function to remove the handlers.
 *
 * @param server - Server or McpServer instance (duck-typed)
 * @param registry - Delegate providing tool listing and routing
 * @param options - Filter and context factory options
 * @returns A detach function to remove the handlers
 */
export async function attachToServer(server, registry, options = {}) {
    const resolved = resolveServer(server);
    const { filter, contextFactory, debug, tracing, stateSync, introspection, serverName, toolExposition = 'flat', actionSeparator = '_', prompts, zeroTrust, selfHealing, swarmGateway, } = options;
    // 1. Propagate observability to all registered builders
    propagateObservability(registry, debug, tracing, options.telemetry);
    // 2. Create State Sync layer (zero overhead when not configured)
    //    Merge manual policies with fluent hints from builders (.invalidates(), .cached())
    const mergedSyncConfig = mergeStateSyncConfig(stateSync, registry.getBuilders());
    const syncLayer = mergedSyncConfig ? new StateSyncLayer(mergedSyncConfig) : undefined;
    // 3. Register introspection resource (zero overhead when disabled)
    //    When `resources` is also configured, introspection is merged
    //    into registerResourceHandlers to avoid setRequestHandler overwrite.
    if (introspection?.enabled && !options.resources) {
        registerIntrospectionResource(resolved, introspection, serverName ?? 'mcpfusion-server', { values: () => registry.getBuilders() }, contextFactory);
    }
    // 3b. Zero-Trust: compile contracts, compute digest, verify attestation
    //     Zero overhead when not configured — no crypto operations run.
    if (zeroTrust) {
        const contracts = await compileContracts(registry.getBuilders());
        const serverDigest = await computeServerDigest(contracts);
        // Synchronous digest comparison (no signer needed for pinning)
        if (zeroTrust.expectedDigest && serverDigest.digest !== zeroTrust.expectedDigest) {
            if (zeroTrust.failOnMismatch ?? true) {
                throw new AttestationError(`[mcpfusion] Zero-Trust attestation failed: computed digest ${serverDigest.digest} does not match expected ${zeroTrust.expectedDigest}`, {
                    valid: false,
                    computedDigest: serverDigest.digest,
                    expectedDigest: zeroTrust.expectedDigest,
                    signature: null,
                    signerName: typeof zeroTrust.signer === 'string' ? zeroTrust.signer : zeroTrust.signer.name,
                    attestedAt: new Date().toISOString(),
                    error: `Digest mismatch: ${serverDigest.digest} !== ${zeroTrust.expectedDigest}`,
                });
            }
        }
    }
    // 4. Build handler context (shared state for all handler factories)
    // FSM State Gate: auto-bind tool bindings from builders
    const { fsm, fsmStore } = options;
    if (fsm) {
        autoBindFsmFromBuilders(fsm, registry.getBuilders(), toolExposition, actionSeparator);
    }
    // Wire the notification sink for list_changed (FSM transitions + FHP swarm)
    let notifyToolListChanged;
    if (fsm || swarmGateway) {
        const serverAny = server;
        const sendFn = serverAny['sendToolListChanged'] ?? serverAny['notification'];
        if (typeof sendFn === 'function') {
            notifyToolListChanged = () => {
                try {
                    void sendFn.call(server, { method: 'notifications/tools/list_changed' });
                }
                catch {
                    // Connection might not be established — ignore
                }
            };
        }
    }
    // exposition dirty flag — O(1) cache validation
    // Starts dirty to force initial compile, then stays clean until invalidated.
    let _expositionDirty = true;
    const hCtx = {
        registry,
        ...(filter ? { filter } : {}),
        ...(contextFactory ? { contextFactory } : {}),
        ...(syncLayer ? { syncLayer } : {}),
        toolExposition, actionSeparator,
        isFlat: toolExposition === 'flat',
        // O(1) exposition cache with dirty flag + builder-count safety net.
        // The dirty flag provides O(1) fast-path invalidation (set whenever a builder
        // registers or the filter changes). The builder-count safety net catches
        // late-registered builders that might have been missed by the dirty flag.
        //
        //  (Performance) fix: previous safety net iterated `registry.getBuilders()`
        // with a for...of loop (O(n)) on every tools/call request even on the cache hit path.
        // `ToolRegistry.size` is a Map.size getter — always O(1). No loop needed.
        recompile: (() => {
            let cachedResult;
            let cachedBuilderCount = -1;
            let cachedCompactMode;
            return (fsmCompactMode) => {
                // O(1) size check: detect late-registered builders without iterating.
                const currentCount = registry.size;
                // Invalidate cache when FSM compact mode changes (progressive disclosure transition)
                const compactModeChanged = fsmCompactMode !== cachedCompactMode;
                if (!_expositionDirty && cachedResult && currentCount === cachedBuilderCount && !compactModeChanged) {
                    return cachedResult;
                }
                _expositionDirty = false;
                cachedBuilderCount = currentCount;
                cachedCompactMode = fsmCompactMode;
                const builders = [...registry.getBuilders()];
                // route diagnostic warnings through debug observer
                const warnFn = debug
                    ? (msg) => debug({ type: 'error', tool: '', action: '', error: msg, step: 'route', timestamp: Date.now() })
                    : undefined;
                cachedResult = compileExposition(builders, toolExposition, actionSeparator, warnFn, fsmCompactMode);
                return cachedResult;
            };
        })(),
        ...(fsm ? { fsm } : {}),
        ...(fsmStore ? { fsmStore } : {}),
        // in-memory FSM snapshot store when no external fsmStore
        // bounded LRU eviction (max 10,000 entries) to prevent
        // unbounded memory growth proportional to unique session count.
        ...(fsm && !fsmStore ? { fsmMemorySnapshots: createBoundedSnapshotMap(10_000) } : {}),
        ...(notifyToolListChanged ? { notifyToolListChanged } : {}),
        ...(options.telemetry ? { telemetry: options.telemetry } : {}),
        ...(selfHealing ? { selfHealing } : {}),
        ...(swarmGateway ? { swarmGateway } : {}),
        // per-attachment UUID — never use static key for session-scoped mutable state
        fallbackSessionId: randomUUID(),
    };
    // 5. Register tool handlers
    resolved.setRequestHandler(ListToolsRequestSchema, createToolListHandler(hCtx));
    resolved.setRequestHandler(CallToolRequestSchema, createToolCallHandler(hCtx));
    // 6. Register prompt handlers (zero overhead when omitted)
    if (prompts) {
        registerPromptHandlers(resolved, server, prompts, registry, filter, contextFactory);
    }
    // 7. Register resource handlers (zero overhead when omitted)
    const { resources } = options;
    if (resources) {
        // pass introspection config so manifest resource is merged
        // into ResourceRegistry handlers instead of being overwritten.
        registerResourceHandlers(resolved, server, resources, contextFactory, introspection?.enabled ? {
            config: introspection,
            serverName: serverName ?? 'mcpfusion-server',
            builders: { values: () => registry.getBuilders() },
        } : undefined);
    }
    // 8. Return detach function
    return createDetachFn(resolved, prompts !== undefined);
}
// ── Flat Tool Filtering ──────────────────────────────────
/**
 * Filter flat tools by tag criteria.
 *
 * Maps each flat tool back to its originating builder to check tags,
 * then applies the standard tag filter logic.
 */
function filterFlatTools(tools, routeMap, filter) {
    const requiredTags = filter.tags && filter.tags.length > 0 ? new Set(filter.tags) : undefined;
    const anyTags = filter.anyTag && filter.anyTag.length > 0 ? new Set(filter.anyTag) : undefined;
    const excludeTags = filter.exclude && filter.exclude.length > 0 ? new Set(filter.exclude) : undefined;
    if (!requiredTags && !anyTags && !excludeTags)
        return tools;
    return tools.filter(tool => {
        const route = routeMap.get(tool.name);
        if (!route)
            return true; // Non-flat tool, include by default
        const builderTags = route.builder.getTags();
        // AND logic: builder must have ALL required tags
        if (requiredTags && !Array.from(requiredTags).every(t => builderTags.includes(t))) {
            return false;
        }
        // OR logic: builder must have at least ONE of these tags
        if (anyTags && !builderTags.some(t => anyTags.has(t))) {
            return false;
        }
        // Exclude: builder must NOT have ANY of these tags
        if (excludeTags && builderTags.some(t => excludeTags.has(t))) {
            return false;
        }
        return true;
    });
}
// ── Progress Sink Factory ────────────────────────────────
/**
 * Duck-type check: the extra object from MCP SDK has _meta and sendNotification.
 */
function isMcpExtra(extra) {
    return (typeof extra === 'object' &&
        extra !== null &&
        'sendNotification' in extra &&
        typeof extra.sendNotification === 'function');
}
/**
 * Create a ProgressSink from the MCP request `extra` object.
 *
 * When the client includes `_meta.progressToken` in its `tools/call` request,
 * this factory returns a ProgressSink that maps each internal ProgressEvent
 * to the MCP `notifications/progress` protocol wire format.
 *
 * When no progressToken is present (client didn't opt in),
 * returns `undefined` — zero overhead.
 *
 * @param extra - The MCP request handler's extra argument (duck-typed)
 * @returns A ProgressSink or undefined
 */
function createProgressSink(extra) {
    if (!isMcpExtra(extra))
        return undefined;
    const token = extra._meta?.progressToken;
    if (token === undefined)
        return undefined;
    const sendNotification = extra.sendNotification;
    return (event) => {
        // Fire-and-forget: progress notifications are best-effort.
        // We intentionally do not await to avoid blocking the handler pipeline.
        void sendNotification({
            method: 'notifications/progress',
            params: {
                progressToken: token,
                progress: event.percent,
                total: 100,
                message: event.message,
            },
        });
    };
}
// ── Signal Extraction ────────────────────────────────────
/**
 * Extract the AbortSignal from the MCP SDK `extra` object.
 *
 * The SDK fires this signal when the client sends `notifications/cancelled`
 * or when the transport connection drops. By extracting and propagating it,
 * the framework enables cooperative cancellation at every pipeline layer.
 *
 * Returns `undefined` when not available — zero overhead.
 *
 * @param extra - The MCP request handler's extra argument (duck-typed)
 * @returns The AbortSignal or undefined
 */
function extractSignal(extra) {
    if (!isMcpExtra(extra))
        return undefined;
    return extra.signal;
}
// ── Elicitation Sink Extraction ────────────────────────
/**
 * Extract the elicitation transport function from the MCP SDK `extra` object.
 *
 * When the MCP SDK provides `sendRequest`, this returns a function that
 * sends `elicitation/create` requests to the client for human-in-the-loop
 * workflows. Bound to `AsyncLocalStorage` so `ask()` works as a standalone.
 *
 * Returns `undefined` when not available — zero overhead.
 *
 * @param extra - The MCP request handler's extra argument (duck-typed)
 * @returns An ElicitSink or undefined
 */
function extractElicitSink(extra) {
    if (!isMcpExtra(extra))
        return undefined;
    const sendRequest = extra.sendRequest;
    if (typeof sendRequest !== 'function')
        return undefined;
    return sendRequest;
}
// ── State Sync Hint Collection ──────────────────────────────
/**
 * Collect per-builder state sync hints and merge with manual config.
 *
 * Three scenarios:
 * 1. Manual `stateSync` only — returns it unchanged
 * 2. Fluent hints only — generates policies automatically
 * 3. Both — fluent-generated policies are appended AFTER manual ones
 *    (first-match-wins, so manual policies take precedence)
 *
 * Zero overhead when neither is configured.
 */
function mergeStateSyncConfig(manual, builders) {
    const hintPolicies = collectHintPolicies(builders);
    if (hintPolicies.length === 0)
        return manual;
    if (!manual)
        return { policies: hintPolicies };
    // Merge: manual first (higher precedence), then auto-generated
    return {
        ...manual,
        policies: [...manual.policies, ...hintPolicies],
    };
}
/**
 * Walk all builders and convert their StateSyncHints into SyncPolicy[].
 *
 * For each builder with hints:
 * - `'*'` key → tool-level policy matching `{toolName}.*`
 * - Named action keys → action-level policy matching `{toolName}.{actionKey}`
 */
function collectHintPolicies(builders) {
    const policies = [];
    for (const builder of builders) {
        if (!builder.getStateSyncHints)
            continue;
        const hints = builder.getStateSyncHints();
        if (hints.size === 0)
            continue;
        const toolName = builder.getName();
        for (const [key, hint] of hints) {
            const match = key === '*' ? `${toolName}.*` : `${toolName}.${key}`;
            policies.push({
                match,
                ...(hint.cacheControl ? { cacheControl: hint.cacheControl } : {}),
                ...(hint.invalidates != null && hint.invalidates.length > 0 ? { invalidates: [...hint.invalidates] } : {}),
            });
        }
    }
    return policies;
}
// ── Session ID Extraction ──────────────────────────────
/**
 * Create a bounded Map for in-memory FSM snapshots with LRU eviction.
 *
 * When the map exceeds `maxSize`, the oldest entry (first inserted) is evicted.
 * Uses native `Map` iteration order guarantee (insertion order) as the LRU proxy.
 * On `get()`, the accessed entry is re-inserted to refresh its position.
 *
 * prevents unbounded memory growth proportional to unique sessions.
 */
function createBoundedSnapshotMap(maxSize) {
    const map = new Map();
    const originalSet = map.set.bind(map);
    const originalGet = map.get.bind(map);
    const originalHas = map.has.bind(map);
    const originalDelete = map.delete.bind(map);
    map.get = (key) => {
        const value = originalGet(key);
        if (value !== undefined) {
            // Refresh position: delete and re-insert to make it "most recently used"
            originalDelete(key);
            originalSet(key, value);
        }
        return value;
    };
    map.set = (key, value) => {
        // If key already exists, delete first to refresh position
        if (originalHas(key)) {
            originalDelete(key);
        }
        originalSet(key, value);
        // Evict oldest entry if over capacity
        if (map.size > maxSize) {
            const oldest = map.keys().next().value;
            if (oldest !== undefined)
                originalDelete(oldest);
        }
        return map;
    };
    return map;
}
/**
 * Extract the MCP session identifier from the request `extra` object.
 *
 * For Streamable HTTP transport, the session ID comes from the
 * `Mcp-Session-Id` header. For stdio/SSE transports with persistent
 * connections, a stable session ID may be available from the SDK.
 *
 * Returns `undefined` when not available (stdio transport without session tracking).
 *
 * @param extra - The MCP request handler's extra argument (duck-typed)
 * @returns Session ID string or undefined
 */
function extractSessionId(extra) {
    if (typeof extra !== 'object' || extra === null)
        return undefined;
    const ex = extra;
    // Standard MCP SDK session ID
    if (typeof ex['sessionId'] === 'string')
        return ex['sessionId'];
    // Streamable HTTP: from request headers  
    const headers = ex['headers'];
    if (headers && typeof headers['mcp-session-id'] === 'string') {
        return headers['mcp-session-id'];
    }
    return undefined;
}
// ── FSM Auto-Binding ─────────────────────────────────
/**
 * Auto-bind FSM tool bindings from all registered builders.
 *
 * Walks all builders, checks for `.bindState()` metadata, and registers
 * the bindings on the `StateMachineGate`. This allows the dev to use
 * `.bindState()` on FluentToolBuilder without manually calling
 * `gate.bindTool()` for each tool.
 *
 * In flat exposition mode, tool names are `{toolName}{separator}{actionKey}`.
 * In grouped mode, tool names are just the builder's name.
 */
function autoBindFsmFromBuilders(gate, builders, exposition, separator) {
    for (const builder of builders) {
        // Duck-type: check if builder has getFsmBinding
        const getFsm = builder['getFsmBinding'];
        if (typeof getFsm !== 'function')
            continue;
        const binding = getFsm.call(builder);
        if (!binding)
            continue;
        const toolName = builder.getName();
        if (exposition === 'flat') {
            // In flat mode, each action becomes a separate tool: toolName_actionKey
            // We need to bind each flat tool name to the FSM
            const actions = builder['getActions'];
            if (typeof actions === 'function') {
                const actionList = actions.call(builder);
                const isSingleAction = actionList.length === 1;
                for (const action of actionList) {
                    // single-action default tools use bare name
                    // (matching ExpositionCompiler.compileFlat behavior)
                    const flatName = (isSingleAction && action.key === 'default')
                        ? toolName
                        : `${toolName}${separator}${action.key}`;
                    gate.bindTool(flatName, binding.states, binding.transition);
                }
            }
            else {
                // Fallback: bind the base tool name
                gate.bindTool(toolName, binding.states, binding.transition);
            }
        }
        else {
            // Grouped mode: tool name is just the builder name
            gate.bindTool(toolName, binding.states, binding.transition);
        }
    }
}
