// @mcpfusion/core/dist/server/startServer.js
/**
 * startServer — One-Liner Bootstrap for MCP mcpfusion Servers
 *
 * Abstracts the entire server startup boilerplate into a single call:
 *   1. Creates the MCP Server instance
 *   2. Attaches the tool registry with telemetry
 *   3. Builds the topology for Inspector TUI auto-discovery
 *   4. Starts the Telemetry Bus (IPC)
 *   5. Connects the transport (stdio or Streamable HTTP)
 *
 * Supports both `stdio` and `http` transports:
 * - `stdio` (default): connects via stdin/stdout
 * - `http`: starts an HTTP server with session management
 *
 * @module
 */
import { createServer as createHttpServer } from 'node:http';
import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { StreamableHTTPServerTransport } from '@modelcontextprotocol/sdk/server/streamableHttp.js';
import { attachToServer as _attachToServer, _missingContextProxy } from './ServerAttachment.js';
import { createTelemetryBus } from '../observability/TelemetryBus.js';
import { compileServerCard, SERVER_CARD_PATH } from '../introspection/ServerCard.js';
/**
 * UUID v4 format validator for Mcp-Session-Id.
 * Rejects oversized or malformed session IDs before they reach the session Map.
 * @internal
 */
const UUID_V4_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
function isValidSessionId(id) {
    return UUID_V4_RE.test(id);
}
/** Dangerous keys that must never be assigned to user state objects. */
const POISONED_KEYS = new Set(['__proto__', 'constructor', 'prototype']);
/**
 * Token bucket rate limiter — tracks request counts per session per minute.
 * Resets automatically every 60 seconds.
 * @internal
 */
class RateLimitBucket {
    _limit;
    _buckets = new Map();
    constructor(limitPerMinute) {
        this._limit = Math.max(1, limitPerMinute);
    }
    /** Returns `true` if the request is allowed, `false` if rate-limited. */
    allow(sessionId) {
        const now = Date.now();
        let bucket = this._buckets.get(sessionId);
        if (!bucket || now >= bucket.resetAt) {
            bucket = { count: 0, resetAt: now + 60_000 };
            this._buckets.set(sessionId, bucket);
        }
        bucket.count++;
        return bucket.count <= this._limit;
    }
    /** Remove stale buckets for sessions that no longer exist. */
    prune(activeSessions) {
        for (const key of this._buckets.keys()) {
            if (!activeSessions.has(key))
                this._buckets.delete(key);
        }
    }
}
/**
 * Apply CORS headers to a response if CORS is configured.
 * Returns `true` if the request is a preflight OPTIONS that was fully handled.
 * @internal
 */
function applyCorsHeaders(req, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- duck-typed HTTP response
res, cors) {
    if (!cors)
        return false;
    const requestOrigin = req.headers['origin'] ?? '';
    const allowedOrigins = typeof cors.origin === 'string' ? [cors.origin] : cors.origin;
    const methods = cors.methods ?? ['GET', 'POST', 'DELETE', 'OPTIONS'];
    // Check if the request origin is allowed
    let matchedOrigin;
    if (allowedOrigins.includes('*')) {
        matchedOrigin = '*';
    }
    else if (requestOrigin && allowedOrigins.includes(requestOrigin)) {
        matchedOrigin = requestOrigin;
    }
    else {
        matchedOrigin = allowedOrigins[0] ?? '';
    }
    res.setHeader('Access-Control-Allow-Origin', matchedOrigin);
    res.setHeader('Access-Control-Allow-Methods', methods.join(', '));
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Mcp-Session-Id, Authorization');
    // Handle preflight
    if (req.method === 'OPTIONS') {
        res.writeHead(204).end();
        return true;
    }
    return false;
}
// ============================================================================
// Implementation
// ============================================================================
/**
 * Start an MCP mcpfusion server with a single call.
 *
 * Handles all bootstrap boilerplate: Server creation, registry attachment,
 * telemetry bus, and stdio transport connection.
 *
 * @example
 * ```typescript
 * import { startServer, autoDiscover } from '@mcpfusion/core';
 *
 * const registry = f.registry();
 * await autoDiscover(registry, new URL('./tools', import.meta.url));
 *
 * await startServer({
 *     name: 'my-server',
 *     registry,
 *     contextFactory: () => createContext(),
 * });
 * ```
 */
export async function startServer(options) {
    const { name, version = '1.0.0', registry, prompts, contextFactory, telemetry = true, transport = 'stdio', port = 3001, attach = {}, state, credentials, serverCard: serverCardOpt, } = options;
    // ── Vinkius Cloud Edge Detection ─────────────────────────────────────
    // When running inside a V8 Isolate, the host injects
    // __vinkius_edge_interceptor into globalThis. If present:
    //   1. Serialize tool definitions to host via IPC
    //   2. Expose __vinkius_edge_dispatch for tool invocations
    //   3. Abort normal server startup (no stdio transport)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const g = globalThis;
    if (typeof g.__vinkius_edge_interceptor !== 'undefined') {
        const tools = [];
        for (const b of registry.getBuilders()) {
            tools.push({
                definition: b.buildToolDefinition(),
                name: b.getName(),
            });
        }
        // ── Prompt Definitions ───────────────────────────────
        // Serialize all registered prompts so the host can expose
        // them via MCP prompts/list + prompts/get.
        // McpPromptDef is already the serializable shape (name, description, arguments)
        const promptDefs = prompts ? prompts.getAllPrompts() : [];
        // ── FSM Config ───────────────────────────────────────
        // Serialize FSM config and tool bindings so the host can
        // implement state-gated tool visibility at edge.
        let fsmData;
        const fsmGate = attach['fsm'];
        if (fsmGate != null && fsmGate._config != null && fsmGate._bindings != null) {
            const bindings = [];
            for (const [toolName, binding] of fsmGate._bindings) {
                bindings.push({
                    tool: toolName,
                    states: [...binding.allowedStates],
                    ...(binding.transitionEvent ? { event: binding.transitionEvent } : {}),
                });
            }
            fsmData = { config: fsmGate._config, bindings };
        }
        // Send definitions to host via C++ IPC
        g.__vinkius_edge_interceptor.applySync(undefined, [
            JSON.stringify({
                serverName: name,
                version,
                tools,
                ...(promptDefs.length > 0 ? { prompts: promptDefs } : {}),
                ...(fsmData ? { fsm: fsmData } : {}),
                ...(credentials ? { credentials } : {}),
            }),
        ]);
        // Expose async dispatcher — returns JSON string to cross the
        // structured-clone boundary safely. The host (IsolateRunner)
        // calls JSON.parse() on the other side. This eliminates all
        // DataCloneError crashes (Promise, Function, Symbol, etc.).
        g.__vinkius_edge_dispatch = async (toolName, args) => {
            try {
                const ctx = contextFactory
                    ? await contextFactory(undefined)
                    : _missingContextProxy;
                return JSON.stringify(await registry.routeCall(ctx, toolName, args));
            }
            catch (e) {
                const err = e;
                return JSON.stringify({
                    isError: true,
                    content: [{
                            type: 'text',
                            text: String(err.stack ?? err.message ?? e),
                        }],
                });
            }
        };
        // ── Prompt Dispatcher ────────────────────────────────
        // Expose prompt handler for host-side prompt/get invocation.
        if (prompts) {
            g.__vinkius_edge_prompt_get = async (promptName, args) => {
                try {
                    const ctx = contextFactory
                        ? await contextFactory(undefined)
                        : _missingContextProxy;
                    return await prompts.routeGet(ctx, promptName, args);
                }
                catch (e) {
                    const err = e;
                    return {
                        messages: [{
                                role: 'user',
                                content: {
                                    type: 'text',
                                    text: `[ERROR] ${String(err.message ?? e)}`,
                                },
                            }],
                    };
                }
            };
        }
        // ── Hibernation State Hooks ───────────────────────────
        // When `state` is provided, expose serialization hooks so the
        // runtime can extract/inject mutable state across freeze cycles.
        // The state object is updated IN-PLACE to preserve closure references.
        if (state) {
            g.__vinkius_edge_getState = () => JSON.stringify(state);
            g.__vinkius_edge_setState = (json) => {
                const restored = JSON.parse(json);
                // Guard: must be a plain object
                if (typeof restored !== 'object' || restored === null || Array.isArray(restored))
                    return;
                // Clear existing keys (handles removed properties)
                for (const key of Object.keys(state)) {
                    delete state[key];
                }
                // Assign restored values — skip poisoned keys to prevent prototype pollution
                for (const [key, value] of Object.entries(restored)) {
                    if (POISONED_KEYS.has(key))
                        continue;
                    state[key] = value;
                }
            };
        }
        // Abort normal startup — no Server, no Transport ( fix)
        return { server: null, close: () => Promise.resolve() };
    }
    // ── CLI Introspection Mode ───────────────────────────────────────────
    // When `mcpfusion deploy` sets MCPFUSION_INTROSPECT=1, we capture the registry
    // and tool definitions without starting any transport. This allows the
    // CLI to generate a fresh lockfile manifest as the deploy's signature.
    if (process.env['MCPFUSION_INTROSPECT'] === '1') {
        const introspectResult = {
            serverName: name,
            version,
            registry,
            ...(credentials ? { credentials } : {}),
        };
        // Store result and resolve the waiting promise from deploy.ts
        g.__MCPFUSION_INTROSPECT_result = introspectResult;
        if (typeof g.__MCPFUSION_INTROSPECT_resolve === 'function') {
            g.__MCPFUSION_INTROSPECT_resolve(introspectResult);
        }
        return { server: null, close: () => Promise.resolve() };
    }
    // ── Normal Server Startup ────────────────────────────────────────────
    // 1. Telemetry Bus (optional, default on)
    //    Gracefully degrades on serverless platforms (Vercel, Cloudflare)
    //    where IPC sockets / Named Pipes are not available.
    let bus;
    if (telemetry) {
        try {
            // Build topology from registry builders
            const toolGroups = new Map();
            for (const b of registry.getBuilders()) {
                const group = b.getName();
                for (const actionKey of b.getActionNames()) {
                    const list = toolGroups.get(group) ?? [];
                    list.push(actionKey);
                    toolGroups.set(group, list);
                }
            }
            bus = await createTelemetryBus({
                onConnect: () => ({
                    type: 'topology',
                    serverName: name,
                    pid: process.pid,
                    tools: [...toolGroups.entries()].map(([n, actions]) => ({ name: n, actions })),
                    timestamp: Date.now(),
                }),
            });
            process.stderr.write(`📡 Telemetry bus ready (PID ${process.pid})\n`);
        }
        catch {
            // Serverless / sandboxed environments — telemetry unavailable
        }
    }
    // 2. MCP Server Instance
    //    Capabilities are declared based on what features are enabled.
    //    The SDK validates that setRequestHandler() is only called for
    //    capabilities that were declared here — missing declarations
    //    cause "Server does not support X" runtime errors.
    const attachRecord = attach;
    const needsResources = attach.introspection != null
        || attachRecord['resources'] != null;
    const server = new Server({ name, version }, { capabilities: {
            tools: {},
            ...(prompts ? { prompts: {} } : {}),
            ...(needsResources ? { resources: {} } : {}),
        } });
    // 3. Attach Registry
    await registry.attachToServer(server, {
        ...attach,
        ...(contextFactory ? { contextFactory } : {}),
        ...(prompts ? { prompts } : {}),
        ...(bus ? { telemetry: bus.emit } : {}),
    });
    // 4. Connect Transport
    if (transport === 'http') {
        // ── Server Card Compilation (zero overhead when disabled) ─────────
        // Pre-compiled as a static JSON string at startup — no per-request cost.
        let serverCardJson;
        if (serverCardOpt != null) {
            const cardConfig = {
                name,
                version,
                transport: 'streamable-http',
                ...(serverCardOpt === true ? {} : serverCardOpt),
            };
            const resources = attachRecord['resources'];
            const resourceList = resources != null && typeof resources.listResources === 'function'
                ? resources.listResources()
                : undefined;
            const promptList = prompts && typeof prompts.getAllPrompts === 'function'
                ? prompts.getAllPrompts()
                : undefined;
            const card = compileServerCard(cardConfig, registry.getBuilders(), promptList, resourceList);
            serverCardJson = JSON.stringify(card, null, 2);
        }
        // ── Streamable HTTP Transport ────────────────────────
        const sessions = new Map();
        const sessionActivity = new Map();
        // Reverse lookup: transport → sessionId in O(1) for onclose handler.
        // WeakMap ensures entries are GC'd when the transport is disposed.
        const transportToSession = new WeakMap();
        const sessionTtlMs = options.sessionTtlMs ?? 1_800_000; // 30 min
        const reapIntervalMs = options.sessionReapIntervalMs ?? 300_000; // 5 min
        const maxSessions = options.maxSessions ?? 1_000;
        const rateLimiter = new RateLimitBucket(options.rateLimitPerMinute ?? 600);
        // Periodic reaper for abandoned sessions (TCP kill -9, network drops)
        // Also prunes stale rate-limit buckets inline to avoid a redundant Set allocation.
        const reapInterval = setInterval(() => {
            const now = Date.now();
            const activeSessions = new Set();
            for (const [id, lastActive] of sessionActivity) {
                if (now - lastActive > sessionTtlMs) {
                    const t = sessions.get(id);
                    if (t) {
                        void t.close().catch(() => { });
                    }
                    sessions.delete(id);
                    sessionActivity.delete(id);
                }
                else {
                    activeSessions.add(id);
                }
            }
            rateLimiter.prune(activeSessions);
        }, reapIntervalMs);
        if (typeof reapInterval === 'object' && 'unref' in reapInterval) {
            reapInterval.unref();
        }
        // eslint-disable-next-line @typescript-eslint/no-misused-promises -- async HTTP handler is standard Node.js pattern
        const httpServer = createHttpServer(async (req, res) => {
            try {
                // Extract pathname without constructing a full URL object.
                // `new URL()` incurs regex parsing + hostname normalization on every
                // request — unnecessary when we only need the path prefix.
                const rawUrl = req.url ?? '/';
                const qIdx = rawUrl.indexOf('?');
                const pathname = qIdx === -1 ? rawUrl : rawUrl.slice(0, qIdx);
                // ── Server Card: /.well-known/mcp/server-card.json ──────────
                // Serves the auto-discovery endpoint BEFORE the /mcp check.
                // GET-only, publicly cacheable, with security headers.
                if (pathname === SERVER_CARD_PATH && serverCardJson) {
                    if (req.method === 'OPTIONS') {
                        // CORS preflight for server card
                        res.writeHead(204, {
                            'Access-Control-Allow-Origin': '*',
                            'Access-Control-Allow-Methods': 'GET, OPTIONS',
                            'Access-Control-Allow-Headers': 'Content-Type',
                        }).end();
                        return;
                    }
                    if (req.method !== 'GET') {
                        res.writeHead(405).end();
                        return;
                    }
                    res.writeHead(200, {
                        'Content-Type': 'application/json',
                        'Cache-Control': 'public, max-age=300',
                        'X-Content-Type-Options': 'nosniff',
                        'Access-Control-Allow-Origin': '*',
                    }).end(serverCardJson);
                    return;
                }
                if (pathname !== '/mcp') {
                    res.writeHead(404).end();
                    return;
                }
                // Security #6: CORS support
                if (applyCorsHeaders(req, res, options.cors))
                    return; // preflight handled
                if (req.method === 'POST') {
                    // enforce body size limit to prevent DoS/OOM.
                    const maxBytes = options.maxBodyBytes ?? 4_194_304; // 4MB
                    const declaredLength = parseInt(req.headers['content-length'] ?? '', 10);
                    if (declaredLength > maxBytes) {
                        res.writeHead(413).end('Payload too large');
                        return;
                    }
                    const chunks = [];
                    let receivedBytes = 0;
                    for await (const chunk of req) {
                        receivedBytes += chunk.byteLength;
                        if (receivedBytes > maxBytes) {
                            req.destroy();
                            res.writeHead(413).end('Payload too large');
                            return;
                        }
                        chunks.push(chunk);
                    }
                    // Security #3: safe JSON parse — reject malformed bodies with 400
                    let body;
                    try {
                        body = JSON.parse(Buffer.concat(chunks).toString());
                    }
                    catch {
                        res.writeHead(400).end('Invalid JSON');
                        return;
                    }
                    const sessionId = req.headers['mcp-session-id'];
                    // Security #2: validate session ID format (must be UUID v4)
                    if (sessionId && !isValidSessionId(sessionId)) {
                        res.writeHead(400).end('Invalid session ID format');
                        return;
                    }
                    if (sessionId && sessions.has(sessionId)) {
                        // Security #1: rate limiting per session
                        if (!rateLimiter.allow(sessionId)) {
                            res.writeHead(429).end('Too many requests');
                            return;
                        }
                        const t = sessions.get(sessionId);
                        sessionActivity.set(sessionId, Date.now());
                        await t.handleRequest(req, res, body);
                        return;
                    }
                    // Security #4: reject new sessions when at capacity
                    if (sessions.size >= maxSessions) {
                        res.writeHead(503).end('Too many active sessions');
                        return;
                    }
                    const t = new StreamableHTTPServerTransport({
                        sessionIdGenerator: () => crypto.randomUUID(),
                        onsessioninitialized: (id) => {
                            sessions.set(id, t);
                            sessionActivity.set(id, Date.now());
                            transportToSession.set(t, id);
                        },
                    });
                    t.onclose = () => {
                        // O(1) reverse lookup via WeakMap instead of O(n) linear scan
                        const id = transportToSession.get(t);
                        if (id) {
                            sessions.delete(id);
                            sessionActivity.delete(id);
                        }
                    };
                    await server.connect(t);
                    await t.handleRequest(req, res, body);
                }
                else if (req.method === 'GET') {
                    const sessionId = req.headers['mcp-session-id'];
                    if (sessionId && !isValidSessionId(sessionId)) {
                        res.writeHead(400).end('Invalid session ID format');
                        return;
                    }
                    if (sessionId && sessions.has(sessionId)) {
                        if (!rateLimiter.allow(sessionId)) {
                            res.writeHead(429).end('Too many requests');
                            return;
                        }
                        const t = sessions.get(sessionId);
                        sessionActivity.set(sessionId, Date.now());
                        await t.handleRequest(req, res);
                    }
                    else {
                        res.writeHead(400).end('Missing or invalid session');
                    }
                }
                else if (req.method === 'DELETE') {
                    const sessionId = req.headers['mcp-session-id'];
                    if (sessionId && !isValidSessionId(sessionId)) {
                        res.writeHead(400).end('Invalid session ID format');
                        return;
                    }
                    if (sessionId && sessions.has(sessionId)) {
                        const t = sessions.get(sessionId);
                        sessionActivity.set(sessionId, Date.now());
                        await t.handleRequest(req, res);
                    }
                    else {
                        res.writeHead(400).end('Missing or invalid session');
                    }
                }
                else {
                    res.writeHead(405).end();
                }
            }
            catch (err) {
                console.error('[mcpfusion] Unhandled error in HTTP handler:', err);
                if (!res.headersSent)
                    res.writeHead(500).end();
            }
        });
        httpServer.listen(port, () => {
            process.stderr.write(`⚡ ${name} on http://localhost:${port}/mcp\n`);
            if (serverCardJson) {
                process.stderr.write(`🏷️ Server Card at http://localhost:${port}${SERVER_CARD_PATH}\n`);
            }
        });
        async function close() {
            clearInterval(reapInterval);
            if (bus)
                await bus.close();
            for (const t of sessions.values()) {
                try {
                    await t.close();
                }
                catch { /* best effort */ }
            }
            sessions.clear();
            sessionActivity.clear();
            await new Promise((resolve) => httpServer.close(() => resolve()));
            await server.close();
        }
        const result = { server, httpServer, close };
        if (bus)
            result.bus = bus;
        return result;
    }
    // ── Stdio Transport (default) ────────────────────────────
    const stdioTransport = new StdioServerTransport();
    await server.connect(stdioTransport);
    process.stderr.write(`⚡ ${name} running on stdio\n`);
    // 5. Close helper
    async function close() {
        if (bus)
            await bus.close();
        await server.close();
    }
    const result = { server, close };
    if (bus)
        result.bus = bus;
    return result;
}
