// @mcpfusion/core/dist/credentials/credentials.js
/**
 * Credentials — BYOC (Bring Your Own Credentials) System
 *
 * Provides two primitives for marketplace-publishable MCP servers:
 *
 *   defineCredentials() — declare what credentials your server needs.
 *     The Vinkius marketplace reads this at deploy/introspect time and
 *     prompts the buyer to configure credentials before activation.
 *
 *   requireCredential() — read a credential at runtime.
 *     On Vinkius Cloud Edge, secrets are injected into
 *     globalThis.__vinkius_secrets by the runtime before the first tool
 *     call. Locally (stdio/http), populate the same global or use env vars
 *     via a contextFactory.
 *
 * @module
 */
// ============================================================================
// Error
// ============================================================================
/**
 * Thrown by `requireCredential()` when a required credential is missing
 * or empty at tool invocation time.
 */
export class CredentialMissingError extends Error {
    credentialKey;
    constructor(key, hint) {
        const hintText = hint ? ` ${hint}` : '';
        super(`[mcpfusion] Required credential "${key}" is not configured.${hintText}\n` +
            `If running locally, set globalThis.__vinkius_secrets = { "${key}": "..." } ` +
            `before starting the server, or use a contextFactory to read from process.env.`);
        this.name = 'CredentialMissingError';
        this.credentialKey = key;
    }
}
// ============================================================================
// Functions
// ============================================================================
/**
 * Declare the credentials your marketplace server requires.
 *
 * This is a typed identity function — it returns the same map you pass in,
 * providing full TypeScript inference while serving as the introspection
 * anchor read by the Vinkius platform at deploy time.
 *
 * @example
 * ```ts
 * export const credentials = defineCredentials({
 *   REDIS_URL: {
 *     label: 'Redis URL',
 *     description: 'Your Upstash Redis REST URL.',
 *     type: 'url',
 *     required: true,
 *     sensitive: false,
 *   },
 *   REDIS_TOKEN: {
 *     label: 'Redis Token',
 *     description: 'Your Upstash Redis REST Token.',
 *     type: 'api_key',
 *     required: true,
 *     sensitive: true,
 *   },
 * });
 * ```
 */
export function defineCredentials(map) {
    return map;
}
/**
 * Read a credential at runtime.
 *
 * On Vinkius Cloud Edge, the runtime injects secrets into
 * `globalThis.__vinkius_secrets` before the first tool call.
 *
 * @param key   - The credential key as declared in `defineCredentials()`.
 * @param hint  - Optional hint shown in the error message (e.g., where to find the value).
 * @throws {CredentialMissingError} when the credential is absent or empty.
 *
 * @example
 * ```ts
 * function getRedis() {
 *   const url   = requireCredential('REDIS_URL', 'Found in your Upstash console.');
 *   const token = requireCredential('REDIS_TOKEN', 'Found in your Upstash console.');
 *   return new Redis({ url, token });
 * }
 * ```
 */
export function requireCredential(key, hint) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const secrets = globalThis.__vinkius_secrets;
    const value = secrets?.[key];
    if (typeof value !== 'string' || value.trim() === '') {
        throw new CredentialMissingError(key, hint);
    }
    return value;
}
/**
 * Read an optional credential at runtime.
 *
 * Unlike `requireCredential()`, this does NOT throw when the credential
 * is absent or empty — it returns `undefined` instead. Use this for
 * non-mandatory credentials that enhance functionality when present.
 *
 * @param key - The credential key as declared in `defineCredentials()`.
 * @returns The credential value, or `undefined` if absent/empty.
 *
 * @example
 * ```ts
 * const appToken = optionalCredential('BEAGLE_APP_TOKEN');
 * if (appToken) { headers['X-App-Token'] = appToken; }
 * ```
 */
export function optionalCredential(key) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const secrets = globalThis.__vinkius_secrets;
    const value = secrets?.[key];
    return typeof value === 'string' && value.trim() !== '' ? value : undefined;
}
