// @mcpfusion/core/dist/core/schema/SchemaUtils.js
/**
 * SchemaUtils — Shared Schema Inspection & Validation Utilities
 *
 * Extracted helpers for inspecting Zod schema metadata and
 * asserting JSON Schema field compatibility across actions.
 *
 * Used by: SchemaGenerator, DescriptionGenerator, ToonDescriptionGenerator.
 *
 * Pure-function module: no state, no side effects.
 */
import {} from 'zod';
import {} from '../types.js';
// ── Zod Type Guards ──────────────────────────────────────
/**
 * Check if a value is a Zod schema (has `_def` property).
 *
 * Used by `defineTool` and `definePrompt` to distinguish
 * Zod schemas from JSON descriptor objects.
 */
export function isZodSchema(value) {
    return (typeof value === 'object' &&
        value !== null &&
        '_def' in value &&
        typeof value._def === 'object');
}
// ── Schema Inspection ────────────────────────────────────
/**
 * Get the list of required field names from an action's Zod schema.
 * Returns an empty array if the action has no schema.
 */
export function getActionRequiredFields(action) {
    if (!action.schema)
        return [];
    const shape = action.schema.shape;
    const required = [];
    for (const [key, fieldSchema] of Object.entries(shape)) {
        if (!fieldSchema.isOptional()) {
            required.push(key);
        }
    }
    return required;
}
// ── Schema Compatibility ─────────────────────────────────
/**
 * Normalize JSON Schema type for compatibility comparison.
 * "integer" is treated as compatible with "number" (integer IS-A number).
 */
const normalizeType = (t) => t === 'integer' ? 'number' : t;
/** Build a consistently-formatted conflict error. */
function conflictError(field, actionKey, detail) {
    return new Error(`Schema conflict for field "${field}" in action "${actionKey}": ` +
        `${detail}. All actions sharing a field name must use the same type.`);
}
export function assertFieldCompatibility(existing, incoming, field, actionKey) {
    const ex = existing;
    const inc = incoming;
    const exType = ex.type;
    const incType = inc.type;
    const exEnum = ex.enum;
    const incEnum = inc.enum;
    // 1. Base type mismatch (with integer ≈ number normalization)
    if (exType !== undefined && incType !== undefined &&
        normalizeType(exType) !== normalizeType(incType)) {
        throw conflictError(field, actionKey, `type "${incType}" conflicts with previously declared type "${exType}"`);
    }
    // 2. Enum presence mismatch → WIDEN to non-enum (drop constraint)
    //    Plain string is a superset of enum-constrained string.
    //    Essential for OpenAPI imports where the same field name has enum
    //    in one endpoint and plain string in another.
    const exHasEnum = exEnum !== undefined;
    const incHasEnum = incEnum !== undefined;
    if (exHasEnum !== incHasEnum) {
        // Return the non-enum version (superset)
        const nonEnum = exHasEnum ? inc : ex;
        return { ...nonEnum };
    }
    // 3. Enum value-set mismatch → MERGE (union) instead of throwing
    if (exHasEnum && incHasEnum &&
        JSON.stringify(exEnum) !== JSON.stringify(incEnum)) {
        const merged = new Set([...exEnum, ...incEnum]);
        return { ...ex, enum: [...merged] };
    }
    return undefined;
}
