// @mcpfusion/core/dist/core/schema/SchemaGenerator.js
/**
 * SchemaGenerator — JSON Schema Input Schema Strategy
 *
 * Generates MCP-compatible inputSchema from Zod definitions:
 * - Discriminator enum from action keys
 * - Common + per-action schema merging (with omitCommon support)
 * - Per-field annotations (required-for / optional-for)
 *
 * Pure-function module: no state, no side effects.
 */
import {} from 'zod';
import { zodToJson } from './ZodCompat.js';
import {} from '@modelcontextprotocol/sdk/types.js';
import {} from '../types.js';
import { assertFieldCompatibility } from './SchemaUtils.js';
import { isPresenter } from '../../presenter/Presenter.js';
// ── Public API ───────────────────────────────────────────
export function generateInputSchema(actions, discriminator, hasGroup, commonSchema, selectEnabled = false) {
    const actionKeys = actions.map(a => a.key);
    const properties = {};
    const topLevelRequired = [discriminator];
    const fieldActions = new Map();
    addDiscriminatorProperty(properties, discriminator, actionKeys, hasGroup);
    const omitSets = buildOmitSets(actions);
    const commonRequiredFields = collectCommonFields(commonSchema, actionKeys, omitSets, properties, topLevelRequired, fieldActions);
    collectActionFields(actions, properties, fieldActions);
    applyAnnotations(fieldActions, commonRequiredFields, actionKeys, properties);
    // ── _select Reflection (opt-in) ──────────────────────
    if (selectEnabled) {
        injectSelectProperty(actions, properties);
    }
    return {
        type: 'object',
        properties,
        required: topLevelRequired,
    };
}
// ── Internal Steps ───────────────────────────────────────
/** Step 1: Add the discriminator enum property */
function addDiscriminatorProperty(properties, discriminator, actionKeys, hasGroup) {
    properties[discriminator] = {
        type: 'string',
        enum: actionKeys,
        description: hasGroup
            ? `Module and operation (module.${discriminator} format)`
            : 'Which operation to perform',
    };
}
/** Build per-action omit sets for O(1) field exclusion checks */
function buildOmitSets(actions) {
    const sets = new Map();
    for (const action of actions) {
        if ((action.omitCommonFields?.length ?? 0) > 0) {
            sets.set(action.key, new Set(action.omitCommonFields));
        }
    }
    return sets;
}
/**
 * Step 2: Process commonSchema fields.
 *
 * Omitted fields are excluded from per-action tracking.
 * Fields required by ALL actions go to topLevelRequired.
 *
 * @returns The set of common required field names (for annotation logic)
 */
function collectCommonFields(commonSchema, actionKeys, omitSets, properties, topLevelRequired, fieldActions) {
    const commonRequiredFields = new Set();
    if (!commonSchema)
        return commonRequiredFields;
    const jsonSchema = zodToJson(commonSchema);
    const schemaProps = jsonSchema.properties ?? {};
    const schemaRequired = jsonSchema.required ?? [];
    for (const field of schemaRequired) {
        commonRequiredFields.add(field);
    }
    for (const [key, value] of Object.entries(schemaProps)) {
        const actionsUsingField = actionKeys.filter(ak => {
            const omitSet = omitSets.get(ak);
            return !omitSet || !omitSet.has(key);
        });
        // If no action uses the field, skip it entirely
        if (actionsUsingField.length === 0)
            continue;
        properties[key] = value;
        // Only add to topLevelRequired if ALL actions use this field
        if (commonRequiredFields.has(key) && actionsUsingField.length === actionKeys.length) {
            topLevelRequired.push(key);
        }
        fieldActions.set(key, {
            keys: [...actionsUsingField],
            requiredIn: commonRequiredFields.has(key) ? [...actionsUsingField] : [],
        });
    }
    return commonRequiredFields;
}
/** Step 3: Collect per-action schema fields with compatibility checks */
function collectActionFields(actions, properties, fieldActions) {
    for (const action of actions) {
        if (!action.schema)
            continue;
        const jsonSchema = zodToJson(action.schema);
        const schemaProps = jsonSchema.properties ?? {};
        const schemaRequired = jsonSchema.required ?? [];
        const requiredSet = new Set(schemaRequired);
        for (const [key, value] of Object.entries(schemaProps)) {
            const existing = properties[key];
            if (!existing) {
                properties[key] = value;
            }
            else {
                const merged = assertFieldCompatibility(existing, value, key, action.key);
                if (merged) {
                    properties[key] = merged;
                }
            }
            let tracking = fieldActions.get(key);
            if (!tracking) {
                tracking = { keys: [], requiredIn: [] };
                fieldActions.set(key, tracking);
            }
            tracking.keys.push(action.key);
            if (requiredSet.has(key)) {
                tracking.requiredIn.push(action.key);
            }
        }
    }
}
/** Step 4: Apply per-field annotations based on tracking data */
function applyAnnotations(fieldActions, commonRequiredFields, actionKeys, properties) {
    for (const [key, tracking] of fieldActions.entries()) {
        if (commonRequiredFields.has(key) && tracking.keys.length === actionKeys.length) {
            annotateField(properties, key, '(always required)');
        }
        else if (tracking.requiredIn.length > 0 && tracking.requiredIn.length === tracking.keys.length) {
            annotateField(properties, key, `Required for: ${tracking.requiredIn.join(', ')}`);
        }
        else if (tracking.requiredIn.length > 0) {
            const optionalIn = tracking.keys.filter(k => !tracking.requiredIn.includes(k));
            let annotation = `Required for: ${tracking.requiredIn.join(', ')}`;
            if (optionalIn.length > 0) {
                annotation += `. For: ${optionalIn.join(', ')}`;
            }
            annotateField(properties, key, annotation);
        }
        else {
            annotateField(properties, key, `For: ${tracking.keys.join(', ')}`);
        }
    }
}
function annotateField(properties, key, annotation) {
    const field = properties[key];
    if (!field)
        return;
    const existingDesc = field.description ?? '';
    field.description = existingDesc
        ? `${existingDesc}. ${annotation}`
        : annotation;
}
// ── _select Reflection ───────────────────────────────────
/**
 * Inject `_select` property into the input schema.
 *
 * Collects the union of all top-level schema keys from Presenters
 * across all actions. If any action has a Presenter with extractable
 * keys, a `_select` optional array property is added with an enum
 * constraint of those keys.
 *
 * **Top-level only**: The enum lists root-level keys only.
 * Nested objects are returned whole when selected.
 */
function injectSelectProperty(actions, properties) {
    const allKeys = new Set();
    for (const action of actions) {
        if (action.returns && isPresenter(action.returns)) {
            for (const key of action.returns.getSchemaKeys()) {
                allKeys.add(key);
            }
        }
    }
    if (allKeys.size === 0)
        return;
    properties['_select'] = {
        type: 'array',
        description: '⚡ Context optimization: select only the response fields you need. Omit to receive all fields.',
        items: {
            type: 'string',
            enum: [...allKeys].sort(),
        },
    };
}
