// @mcpfusion/core/dist/core/schema/ToonDescriptionGenerator.js
/**
 * ToonDescriptionGenerator — Token-Optimized Description Strategy
 *
 * Generates descriptions using TOON (Token-Oriented Object Notation) format,
 * achieving ~30-50% token reduction compared to the default markdown descriptions.
 *
 * Uses `@toon-format/toon` encode() to serialize action metadata as compact
 * pipe-delimited tabular data inside the description string.
 *
 * Pure-function module: no state, no side effects.
 */
import { encode } from '@toon-format/toon';
import {} from '../types.js';
import { getActionRequiredFields } from './SchemaUtils.js';
// ── Public API ───────────────────────────────────────────
export function generateToonDescription(actions, name, description, hasGroup, discriminator = 'action') {
    const lines = [];
    // Layer 1: Tool summary + dispatch instruction (always human-readable)
    lines.push(`${description || name}. Select operation via the \`${discriminator}\` parameter.`);
    lines.push('');
    // Layer 2: Action metadata in TOON tabular format
    if (hasGroup) {
        lines.push(encodeGroupedActions(actions));
    }
    else {
        lines.push(encodeFlatActions(actions));
    }
    return lines.join('\n');
}
function encodeFlatActions(actions) {
    const rows = actions.map(a => buildActionRow(a.key, a));
    return encode(rows, { delimiter: '|' });
}
function encodeGroupedActions(actions) {
    // Group actions by their groupName
    const groups = new Map();
    for (const action of actions) {
        const key = action.groupName || '_ungrouped';
        let list = groups.get(key);
        if (!list) {
            list = [];
            groups.set(key, list);
        }
        list.push(action);
    }
    // Build a structure that TOON can encode efficiently
    const groupData = {};
    for (const [groupName, groupActions] of groups) {
        groupData[groupName] = groupActions.map(a => buildActionRow(a.actionName, a));
    }
    return encode(groupData, { delimiter: '|' });
}
function buildActionRow(key, action) {
    const required = getActionRequiredFields(action);
    const row = {
        action: key,
        desc: action.description || '',
        required: required.join(','),
    };
    if (action.destructive) {
        row.destructive = true;
    }
    return row;
}
