// @mcpfusion/core/dist/client/MCPFusionClient.js
/**
 * MCPFusionClient — Type-Safe MCP Client (tRPC-style)
 *
 * Provides end-to-end type safety from server to client.
 * The server exports its router type, and the client consumes it
 * with full autocomplete and compile-time validation.
 *
 * @example
 * ```typescript
 * // ── SERVER (mcp-server.ts) ──
 * export const registry = new ToolRegistry<AppContext>();
 * registry.register(projects);
 * registry.register(billing);
 * export type AppRouter = InferRouter<typeof registry>;
 *
 * // ── CLIENT (agent.ts) ──
 * import { createMCPFusionClient } from '@mcpfusion/core/client';
 * import type { AppRouter } from './mcp-server';
 *
 * const client = createMCPFusionClient<AppRouter>(transport);
 * const result = await client.execute('projects.create', { name: 'Vinkius V2' });
 * //                                   ^^^^^^^^^^^^^^^^    ^^^^^^^^^^^^^^^^^
 * //                                   autocomplete!       typed args!
 * ```
 *
 * @module
 */
import {} from '../core/response.js';
// ============================================================================
// Structured Client Error
// ============================================================================
/**
 * Structured error parsed from a `<tool_error>` XML envelope.
 *
 * Provides typed access to self-healing fields so client code
 * can programmatically react to server errors without regex parsing.
 */
export class MCPFusionClientError extends Error {
    /** Error code from the `code` attribute (e.g. `'NOT_FOUND'`). */
    code;
    /** Recovery suggestion from `<recovery>` element. */
    recovery;
    /** Available actions from `<available_actions>` children. */
    availableActions;
    /** Error severity from the `severity` attribute. */
    severity;
    /** Raw ToolResponse that caused the error. */
    raw;
    constructor(message, code, raw, options) {
        super(message);
        this.name = 'MCPFusionClientError';
        this.code = code;
        this.raw = raw;
        this.recovery = options?.recovery;
        this.availableActions = Object.freeze(options?.availableActions ?? []);
        this.severity = options?.severity ?? 'error';
    }
}
// ============================================================================
// XML Error Parser (Internal)
// ============================================================================
/**
 * Decode XML entities back to their original characters.
 *
 * Reverses the escaping applied by `escapeXml()` and `escapeXmlAttr()`
 * so that parsed error messages are human-readable.
 *
 * @internal
 */
function unescapeXml(str) {
    return str
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .replace(/&quot;/g, '"')
        .replace(/&apos;/g, "'")
        .replace(/&amp;/g, '&'); // &amp; must be last to avoid double-decode
}
/**
 * Parse a `<tool_error>` XML envelope into structured fields.
 *
 * Coupled to the output format of `toolError()` from `response.ts`.
 * Uses regex for lightweight parsing — acceptable since the XML is
 * self-produced by the framework, not user-authored.
 *
 * @internal
 */
function parseToolErrorXml(text) {
    const codeMatch = text.match(/<tool_error\s[^>]*code="([^"]+)"/);
    const severityMatch = text.match(/<tool_error\s[^>]*severity="([^"]+)"/);
    const messageMatch = text.match(/<message>([\s\S]*?)<\/message>/);
    const recoveryMatch = text.match(/<recovery>([\s\S]*?)<\/recovery>/);
    if (!messageMatch)
        return null;
    const actions = [];
    const actionMatches = text.matchAll(/<action>([\s\S]*?)<\/action>/g);
    for (const m of actionMatches) {
        if (m[1])
            actions.push(unescapeXml(m[1].trim()));
    }
    // Fallback: legacy comma-separated format
    if (actions.length === 0) {
        const legacyMatch = text.match(/<available_actions>([\s\S]*?)<\/available_actions>/);
        if (legacyMatch?.[1]) {
            actions.push(...legacyMatch[1].split(',').map(a => unescapeXml(a.trim())).filter(Boolean));
        }
    }
    const recovery = recoveryMatch?.[1] != null ? unescapeXml(recoveryMatch[1].trim()) : undefined;
    const result = {
        code: codeMatch?.[1] != null ? unescapeXml(codeMatch[1]) : 'UNKNOWN',
        message: unescapeXml(messageMatch[1].trim()),
        availableActions: actions,
        severity: severityMatch?.[1] ?? 'error',
    };
    if (recovery !== undefined) {
        result.recovery = recovery;
    }
    return result;
}
// ============================================================================
// Middleware Chain Compiler (Internal)
// ============================================================================
/** @internal */
function compileClientMiddleware(middleware, terminal) {
    let chain = terminal;
    // Wrap from right to left: first middleware in array = outermost
    for (let i = middleware.length - 1; i >= 0; i--) {
        const mw = middleware[i];
        const next = chain;
        chain = (action, args) => mw(action, args, next);
    }
    return chain;
}
// ============================================================================
// Factory
// ============================================================================
/**
 * Create a type-safe MCP client.
 *
 * The client provides full autocomplete for action names and
 * compile-time validation for arguments based on the server's
 * router type.
 *
 * @typeParam TRouter - The router map (use `InferRouter<typeof registry>`)
 * @param transport - The MCP transport layer
 * @param options - Client options (middleware, error handling)
 * @returns A typed {@link MCPFusionClient}
 *
 * @example
 * ```typescript
 * import type { AppRouter } from './mcp-server';
 *
 * const client = createMCPFusionClient<AppRouter>(transport);
 *
 * // Full autocomplete + type validation:
 * await client.execute('projects.create', { name: 'Vinkius V2' });
 *
 * // TS error: 'projects.nonexistent' is not a valid action
 * await client.execute('projects.nonexistent', {});
 *
 * // TS error: missing required arg 'name'
 * await client.execute('projects.create', {});
 * ```
 *
 * @example
 * ```typescript
 * // With client middleware and throwOnError
 * const client = createMCPFusionClient<AppRouter>(transport, {
 *     throwOnError: true,
 *     middleware: [
 *         async (action, args, next) => {
 *             console.log(`[Client] calling ${action}`);
 *             const result = await next(action, args);
 *             console.log(`[Client] ${action} done`);
 *             return result;
 *         },
 *     ],
 * });
 * ```
 */
export function createMCPFusionClient(transport, options) {
    const throwOnError = options?.throwOnError ?? false;
    const discriminatorKey = options?.discriminatorKey ?? 'action';
    /** Terminal function: builds the MCP call from the dotted action path */
    function terminalCall(action, args) {
        const dotIndex = action.indexOf('.');
        if (dotIndex === -1) {
            return transport.callTool(action, args);
        }
        const toolName = action.substring(0, dotIndex);
        const actionName = action.substring(dotIndex + 1);
        return transport.callTool(toolName, {
            ...args,
            [discriminatorKey]: actionName,
        });
    }
    // Compile middleware chain once at creation time
    const dispatch = (options?.middleware != null && options.middleware.length > 0)
        ? compileClientMiddleware(options.middleware, terminalCall)
        : terminalCall;
    /** Post-process: handle throwOnError */
    async function executeInternal(action, args) {
        const result = await dispatch(action, args);
        if (throwOnError && result.isError) {
            // Filter defensively — MCP responses may contain ImageContent or
            // EmbeddedResource items at runtime, even though TS infers TextContent only.
            const text = result.content
                .filter(c => c.type === 'text')
                .map(c => c.text ?? '')
                .join('\n');
            const parsed = parseToolErrorXml(text);
            if (parsed) {
                const opts = {
                    availableActions: parsed.availableActions,
                    severity: parsed.severity,
                };
                if (parsed.recovery !== undefined) {
                    opts.recovery = parsed.recovery;
                }
                throw new MCPFusionClientError(parsed.message, parsed.code, result, opts);
            }
            throw new MCPFusionClientError(text || 'Unknown error', 'UNKNOWN', result);
        }
        return result;
    }
    return {
        async execute(action, args) {
            return executeInternal(action, args);
        },
        async executeBatch(calls, batchOptions) {
            const items = calls;
            if (batchOptions?.sequential) {
                const results = [];
                for (const call of items) {
                    results.push(await executeInternal(call.action, call.args));
                }
                return results;
            }
            return Promise.all(items.map(call => executeInternal(call.action, call.args)));
        },
        proxy: buildFluentProxy(executeInternal),
    };
}
// ============================================================================
// Fluent Proxy Builder (Internal)
// ============================================================================
/**
 * Creates a recursive Proxy that accumulates path segments
 * and executes when invoked as a function.
 *
 * ```
 * proxy.projects.create({ name: 'V2' })
 * //   ^^^^^^^^ ^^^^^^  ^^^^^^^^^^^^^^^^
 * //   seg[0]   seg[1]  args → execute('projects.create', args)
 * ```
 *
 * @internal
 */
// Symbols that debugging / inspection tools access on every object.
// Returning undefined for these prevents the proxy from recursing into itself
// when inspected in DevTools, Node REPL, or util.inspect().
const INSPECTION_SYMBOLS = new Set([
    Symbol.toPrimitive,
    Symbol.toStringTag,
    Symbol.iterator,
    Symbol.asyncIterator,
    Symbol.hasInstance,
    ...(typeof Symbol.for === 'function' ? [Symbol.for('nodejs.util.inspect.custom')] : []),
]);
function buildFluentProxy(execute, segments = []) {
    // eslint-disable-next-line @typescript-eslint/no-empty-function -- Proxy requires a callable target for the apply trap
    return new Proxy(function () { }, {
        get(_target, prop) {
            if (typeof prop === 'symbol') {
                // guard inspection symbols
                if (INSPECTION_SYMBOLS.has(prop))
                    return undefined;
                return undefined; // all other symbols → safe no-op
            }
            if (prop === 'then')
                return undefined;
            return buildFluentProxy(execute, [...segments, prop]);
        },
        apply(_target, _thisArg, argsList) {
            const action = segments.join('.');
            const args = (argsList[0] ?? {});
            return execute(action, args);
        },
    });
}
