// @mcpfusion/core/dist/handoff/index.js
/**
 * Federated Handoff Protocol (FHP) — Core Barrel
 *
 * Exporta todas as primitivas FHP que pertencem ao `@mcpfusion/core`:
 *
 * - **Tipos de dados**: `HandoffPayload`, `HandoffResponse`, `HandoffStateStore`
 * - **Guards e factories**: `isHandoffResponse`, `handoff`
 * - **Crypto**: `mintDelegationToken`, `verifyDelegationToken`, `DelegationClaims`, `HandoffAuthError`
 * - **Store**: `InMemoryHandoffStateStore`
 * - **Middleware zero-trust**: `requireGatewayClearance`, `GatewayClearanceContext`
 */
export { isHandoffResponse, handoff } from '../core/response.js';
// ── Store padrão ─────────────────────────────────────────────────────────────
export { InMemoryHandoffStateStore } from './HandoffStateStore.js';
// ── Delegation Token (mint/verify, HMAC-SHA256 puro) ─────────────────────────
export { mintDelegationToken, verifyDelegationToken, HandoffAuthError, } from './DelegationToken.js';
// ── Middleware zero-trust (para micro-servidores upstream) ───────────────────
export { requireGatewayClearance } from './middleware.js';
