// @mcpfusion/core/dist/exposition/ExpositionCompiler.js
/**
 * ExpositionCompiler — Topology Compiler Engine
 *
 * Pure-function transformation module responsible for projecting
 * the tool registry's abstract definitions onto the MCP protocol
 * wire format based on the configured Exposition Strategy.
 *
 * - **flat**: Deterministically expands each action into an independent
 *   atomic MCP tool with isolated schema, description, and annotations.
 *
 * - **grouped**: Passthrough — yields the grouped tool definition unmodified.
 *
 * Also emits an O(1) dispatch map for constant-time request proxying
 * during execution.
 *
 * Pure-function module: no state, no side effects.
 *
 * @module
 */
import {} from '@modelcontextprotocol/sdk/types.js';
import {} from 'zod';
import { zodToJson } from '../core/schema/ZodCompat.js';
import {} from '../core/types.js';
import { isPresenter } from '../presenter/Presenter.js';
import {} from './types.js';
export function compileExposition(builders, exposition = 'flat', separator = '_', onWarn, fsmCompactMode = false) {
    if (exposition === 'grouped') {
        return compileGrouped(builders);
    }
    return compileFlat(builders, separator, onWarn, fsmCompactMode);
}
// ── Flat Strategy ────────────────────────────────────────
/**
 * Flat strategy: expand each action into an independent atomic MCP tool.
 *
 * Iterates over each builder, extracts per-action metadata, and
 * synthesizes one McpTool per action with:
 * - Deterministic name: `{toolName}{separator}{actionKey}`
 * - Isolated schema (action-specific + common, no discriminator)
 * - Isolated annotations (readOnlyHint, destructiveHint, idempotentHint)
 * - Per-action description
 */
function compileFlat(builders, separator, onWarn, fsmCompactMode = false) {
    const tools = [];
    const routingMap = new Map();
    for (const builder of builders) {
        // Ensure the grouped definition is compiled (caches internally)
        builder.buildToolDefinition();
        // Use AST reflection interfaces to access internal structures
        const actions = getBuilderActions(builder);
        const discriminator = getBuilderDiscriminator(builder);
        const commonSchema = getBuilderCommonSchema(builder);
        const toolName = builder.getName();
        const selectEnabled = getBuilderSelectEnabled(builder);
        if (!actions || actions.length === 0) {
            // Fallback: builder has no reflection — pass through as grouped
            tools.push(builder.buildToolDefinition());
            continue;
        }
        const isSingleAction = actions.length === 1;
        for (const action of actions) {
            // Single-action builders with 'default' key use bare tool name
            // (no _default suffix). This covers standalone f.query/f.mutation/f.action
            // tools created via the Fluent API which always use 'default' as the key.
            // Multi-action builders and named single actions retain toolName_actionKey.
            const flatName = (isSingleAction && action.key === 'default')
                ? toolName
                : `${toolName}${separator}${action.key}`;
            // ── Schema Purification ──────────────────────────
            const inputSchema = buildAtomicSchema(action, commonSchema, selectEnabled, onWarn);
            // ── Boundary Isolation (Annotations) ─────────────
            const annotations = buildAtomicAnnotations(action);
            // ── Description ───────────────────────────────────────────
            const description = buildAtomicDescription(action, toolName, fsmCompactMode);
            const tool = {
                name: flatName,
                description,
                inputSchema,
            };
            if (Object.keys(annotations).length > 0) {
                Object.defineProperty(tool, 'annotations', {
                    value: annotations,
                    enumerable: true,
                });
            }
            tools.push(tool);
            // ── O(1) Dispatch Map Entry ──────────────────────
            // Detect flat name collisions — two builders
            // producing the same flat name would silently overwrite
            // the routing entry, making the first tool inaccessible.
            if (routingMap.has(flatName)) {
                const existing = routingMap.get(flatName);
                const msg = `[mcpfusion] Flat tool name collision: "${flatName}" is produced by both ` +
                    `"${existing.builder.getName()}" (action "${existing.actionKey}") and ` +
                    `"${toolName}" (action "${action.key}"). ` +
                    `Rename one of the tools or use a different action separator.`;
                if (onWarn) {
                    onWarn(msg);
                }
                else {
                    throw new Error(msg);
                }
            }
            routingMap.set(flatName, {
                builder,
                actionKey: action.key,
                discriminator,
            });
        }
    }
    return { tools, routingMap, isFlat: true };
}
// ── Grouped Strategy ─────────────────────────────────────
/**
 * Grouped strategy: passthrough — yields the grouped tool definition unmodified.
 * This preserves byte-for-byte identical protocol schema compared to pre-exposition behavior.
 */
function compileGrouped(builders) {
    const tools = [];
    for (const builder of builders) {
        tools.push(builder.buildToolDefinition());
    }
    return { tools, routingMap: new Map(), isFlat: false };
}
// ── Atomic Builders (Internal) ───────────────────────────
/**
 * Build an isolated JSON Schema for a single action.
 *
 * Merges the action's specific schema with the common schema,
 * applies omitCommon fields, and strips the discriminator.
 * Result is a clean object schema without the union wrapper.
 */
function buildAtomicSchema(action, commonSchema, selectEnabled = false, onWarn) {
    const properties = {};
    const required = [];
    // ── Merge common schema fields ───────────────────────
    if (commonSchema) {
        const omitSet = action.omitCommonFields
            ? new Set(action.omitCommonFields)
            : undefined;
        const jsonSchema = zodToJson(commonSchema);
        for (const [key, value] of Object.entries(jsonSchema.properties ?? {})) {
            if (omitSet?.has(key))
                continue;
            properties[key] = value;
        }
        for (const key of jsonSchema.required ?? []) {
            if (omitSet?.has(key))
                continue;
            if (key in properties) {
                required.push(key);
            }
        }
    }
    // ── Merge action-specific schema fields ──────────────
    if (action.schema) {
        const jsonSchema = zodToJson(action.schema);
        for (const [key, value] of Object.entries(jsonSchema.properties ?? {})) {
            // Warn when action schema overwrites a common schema field
            // Route through onWarn callback instead of console.warn
            if (key in properties && onWarn) {
                onWarn(`[mcpfusion] Action schema field '${key}' overwrites common schema field with the same name. ` +
                    `Use omitCommonFields to exclude the common field, or rename the action field.`);
            }
            properties[key] = value;
        }
        for (const key of jsonSchema.required ?? []) {
            if (key in properties && !required.includes(key)) {
                required.push(key);
            }
        }
    }
    // ── _select Reflection (opt-in) ──────────────────────
    if (selectEnabled && action.returns && isPresenter(action.returns)) {
        const schemaKeys = action.returns.getSchemaKeys();
        if (schemaKeys.length > 0) {
            properties['_select'] = {
                type: 'array',
                description: '⚡ Context optimization: select only the response fields you need. Omit to receive all fields.',
                items: {
                    type: 'string',
                    enum: [...schemaKeys].sort(),
                },
            };
        }
    }
    return {
        type: 'object',
        properties,
        ...(required.length > 0 ? { required } : {}),
    };
}
/**
 * Build isolated annotations for a single action.
 *
 * Each action gets its own `readOnlyHint`, `destructiveHint`,
 * and `idempotentHint` — no privilege bleed across actions.
 */
function buildAtomicAnnotations(action) {
    const annotations = {};
    if (action.readOnly === true) {
        // Action is read-only → never destructive.
        // readOnlyHint default is false, so we must explicitly signal true.
        // destructiveHint default is true (MCP assumes dangerous), so we must
        // explicitly signal false to prevent unnecessary confirmation dialogs.
        annotations['readOnlyHint'] = true;
        annotations['destructiveHint'] = false;
    }
    else if (action.destructive === true) {
        // Action is destructive → matches MCP's conservative default.
        // We still emit explicitly for clarity in tools/list responses.
        annotations['destructiveHint'] = true;
    }
    else {
        // Neither read-only nor destructive.
        // MCP spec defaults destructiveHint to true (assume dangerous).
        // We MUST emit false to override this — otherwise clients like
        // Claude Desktop or Cursor will show unnecessary safety warnings.
        annotations['destructiveHint'] = false;
    }
    if (action.idempotent === true) {
        annotations['idempotentHint'] = true;
    }
    return annotations;
}
/**
 * Build a human-readable description for a single atomic action.
 */
function buildAtomicDescription(action, toolName, fsmCompactMode = false) {
    const parts = [];
    // FSM Progressive Disclosure: use compact description during discovery
    if (fsmCompactMode && action.compactDescription) {
        parts.push(action.compactDescription);
    }
    else if (action.description) {
        parts.push(action.description);
    }
    else {
        // Generate a default description from the tool and action name
        const actionLabel = action.groupName
            ? `${action.groupName}.${action.actionName}`
            : action.actionName;
        parts.push(`${toolName} → ${actionLabel}`);
    }
    // Append safety flags for LLM awareness
    if (action.destructive) {
        parts.push('[DESTRUCTIVE]');
    }
    if (action.readOnly) {
        parts.push('[READ-ONLY]');
    }
    return parts.join(' ');
}
// ── AST Reflection Helpers ───────────────────────────────
/**
 * Extract actions from a builder using optional AST reflection interface.
 * Returns undefined if the builder doesn't implement `getActions()`.
 */
function getBuilderActions(builder) {
    return builder.getActions?.();
}
/**
 * Extract discriminator field name from a builder.
 * Defaults to 'action' if the builder doesn't implement `getDiscriminator()`.
 */
function getBuilderDiscriminator(builder) {
    return builder.getDiscriminator?.() ?? 'action';
}
/**
 * Extract common schema from a builder.
 * Returns undefined if the builder doesn't implement `getCommonSchema()`.
 */
function getBuilderCommonSchema(builder) {
    return builder.getCommonSchema?.();
}
/**
 * Check if select reflection is enabled on a builder.
 * Returns false if the builder doesn't implement `getSelectEnabled()`.
 */
function getBuilderSelectEnabled(builder) {
    return builder.getSelectEnabled?.() ?? false;
}
